<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Classification Using Caret • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Classification Using Caret">
<meta property="og:description" content="BS831">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Classification_Caret_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Classification Using Caret</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/Classification_Caret.Rmd"><code>vignettes/docs/Classification_Caret.Rmd</code></a></small>
      <div class="hidden name"><code>Classification_Caret.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/montilab/BS831">BS831</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Biobase</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/">caret</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://expasy.org/tools/pROC/">pROC</a></span><span class="op">)</span></code></pre></div>
<div id="loading-the-drugmatrix-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-the-drugmatrix-data" class="anchor"></a>Loading the DrugMatrix Data</h2>
<p>Loading data for the classification task. We will be using the 10-chemical subset of the DrugMatrix dataset, and aim to predict the chemicals’ carcinogenicity status.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">dm10</span><span class="op">)</span>

<span class="co">## DrugMatrix dataset comprising 10 drugs plus controls</span>
<span class="va">DM</span> <span class="op">&lt;-</span> <span class="va">dm10</span>
<span class="co">## show chemicals and their numbers</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM</span><span class="op">)</span><span class="op">[</span>,<span class="st">"CHEMICAL"</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##                                     2,3,7,8-TETRACHLORODIBENZO-P-DIOXIN 
##                                 225                                   9 
##                       ACETAMINOPHEN                             ASPIRIN 
##                                   3                                  24 
##                            DIAZEPAM                         DOXORUBICIN 
##                                   8                                  21 
##                           IBUPROFEN                          LOVASTATIN 
##                                   9                                  18 
##                       ROSIGLITAZONE                           TAMOXIFEN 
##                                   6                                   9 
##                        TESTOSTERONE 
##                                   6</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## show carcinogenicity status</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Carcinogen_liv"</span><span class="op">]</span>,exclude<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   NON-CARC CARCINOGEN       &lt;NA&gt; 
##         63         50        225</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## exclude the controls (the ones with carcinogenicity status==NA)</span>
<span class="va">DM1</span> <span class="op">&lt;-</span> <span class="va">DM</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Carcinogen_liv"</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>
<span class="va">DM1</span><span class="op">$</span><span class="va">Carcinogen_liv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span><span class="op">(</span><span class="va">DM1</span><span class="op">$</span><span class="va">Carcinogen_liv</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM1</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Carcinogen_liv"</span><span class="op">]</span>,exclude<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   NON-CARC CARCINOGEN 
##         63         50</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## carry out (drastic) variation filtering</span>
<span class="va">DM1000</span> <span class="op">&lt;-</span> <span class="fu">variationFilter</span><span class="op">(</span><span class="va">DM1</span>,ngenes<span class="op">=</span><span class="fl">1000</span>,score<span class="op">=</span><span class="st">"mad"</span>,do.plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## Variation filtering based on mad .. done.
## Selecting top 1000 by mad .. done, 1000 genes selected.</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">)</span></code></pre></div>
<pre><code>## Features  Samples 
##     1000      113</code></pre>
<div id="exploratory-data-visualization" class="section level3">
<h3 class="hasAnchor">
<a href="#exploratory-data-visualization" class="anchor"></a>Exploratory Data Visualization</h3>
<p>After having uploaded and properly processed the data, we start by visualizing them with a heatmap (see <code>Rmodule_heatmap.Rmd</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span>
<span class="va">DMscaled</span> <span class="op">&lt;-</span> <span class="fu">scale_row</span><span class="op">(</span><span class="va">DM1000</span><span class="op">)</span>
<span class="va">hm.annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>
    CARCINOGEN<span class="op">=</span><span class="va">DMscaled</span><span class="op">$</span><span class="va">Carcinogen_liv</span>,
    GTOX<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">DMscaled</span><span class="op">$</span><span class="va">GenTox</span><span class="op">)</span>,
    na_col<span class="op">=</span><span class="st">"ghostwhite"</span>,
    col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>CARCINOGEN<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NON-CARC"</span><span class="op">=</span><span class="st">"green"</span>,CARCINOGEN<span class="op">=</span><span class="st">"red"</span><span class="op">)</span>,
             GTOX<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'0'</span><span class="op">=</span><span class="st">"orange"</span>,<span class="st">'1'</span><span class="op">=</span><span class="st">"purple"</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">DMscaled</span><span class="op">)</span>,
        name<span class="op">=</span><span class="st">"expression"</span>, 
        col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#072448"</span>, <span class="st">"white"</span>, <span class="st">"#ff6150"</span><span class="op">)</span><span class="op">)</span>,
        top_annotation<span class="op">=</span><span class="va">hm.annotation</span>, 
        cluster_rows<span class="op">=</span><span class="cn">TRUE</span>,
        cluster_columns<span class="op">=</span><span class="cn">TRUE</span>,
        clustering_distance_rows<span class="op">=</span><span class="st">"euclidean"</span>,
        clustering_method_rows<span class="op">=</span><span class="st">"ward.D"</span>,    
        clustering_distance_columns<span class="op">=</span><span class="st">"euclidean"</span>,
        clustering_method_columns<span class="op">=</span><span class="st">"ward.D"</span>, 
        column_split<span class="op">=</span><span class="fl">6</span>,
        show_parent_dend_line<span class="op">=</span><span class="cn">TRUE</span>,
        row_title<span class="op">=</span><span class="st">""</span>,
        show_column_names<span class="op">=</span><span class="cn">FALSE</span>,
        show_row_names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="Classification_Caret_files/figure-html/complex.heatmap-1.png" width="672"></p>
</div>
</div>
<div id="classifying-the-carcinogenicity-status-of-compounds" class="section level2">
<h2 class="hasAnchor">
<a href="#classifying-the-carcinogenicity-status-of-compounds" class="anchor"></a>Classifying the Carcinogenicity Status of Compounds</h2>
<div id="data-partitioning" class="section level3">
<h3 class="hasAnchor">
<a href="#data-partitioning" class="anchor"></a>Data partitioning</h3>
<p>Here, we perform the data partition into discovery and validation sets (alternatively, into train and test sets).</p>
<p>We start by simplifying the phenotype labels.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3456</span><span class="op">)</span> <span class="co"># for reproducible results</span>

<span class="co">## some renaming to make data handling simpler</span>
<span class="va">DM1000</span><span class="op">$</span><span class="va">CHEMICAL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">$</span><span class="va">CHEMICAL</span><span class="op">)</span>              <span class="co"># ensuring phenoype is a factor</span>
<span class="va">DM1000</span><span class="op">$</span><span class="va">Carcinogen_liv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">$</span><span class="va">Carcinogen_liv</span><span class="op">)</span>  <span class="co"># ditto</span>
<span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">$</span><span class="va">Carcinogen_liv</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">$</span><span class="va">Carcinogen_liv</span><span class="op">)</span><span class="op">==</span><span class="st">"NON-CARC"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"noncarc"</span>
<span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">$</span><span class="va">Carcinogen_liv</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">$</span><span class="va">Carcinogen_liv</span><span class="op">)</span><span class="op">==</span><span class="st">"CARCINOGEN"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"carc"</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"CHEMICAL"</span>, <span class="st">"Carcinogen_liv"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co"># show composition</span></code></pre></div>
<pre><code>##                                      Carcinogen_liv
## CHEMICAL                              noncarc carc
##   2,3,7,8-TETRACHLORODIBENZO-P-DIOXIN       0    9
##   ACETAMINOPHEN                             3    0
##   ASPIRIN                                  24    0
##   DIAZEPAM                                  0    8
##   DOXORUBICIN                              21    0
##   IBUPROFEN                                 9    0
##   LOVASTATIN                                0   18
##   ROSIGLITAZONE                             6    0
##   TAMOXIFEN                                 0    9
##   TESTOSTERONE                              0    6</code></pre>
<p>We then perform the <em>stratified</em> partition, whereby stratified means that both subsets should have a relatively equal representation of the two classes (‘noncar’ and ‘carc’).</p>
<p>Also, notice that there is an additional level of complexity here, because we have multiple replicates for each chemical. Hence, if we were to carry out the split ‘naively’, we might end up with representative of the same chemicals <em>both</em> in the discovery and validation set, which would not be appropriate (it would be ‘cheating’).</p>
<p>To address this issue, we carry out the partition at the <em>chemical</em> level, rather than the <em>sample</em> level. That is, we partition the chemicals into a discovery and validation set, so that <em>all</em> the replicates of a given chemical will be assigned to either the discovery or the validation set.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## data partitioning into discovery and validation set</span>
<span class="va">CHEMICALGROUP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">DM1000</span><span class="op">)</span><span class="op">$</span><span class="va">CHEMICAL</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CHEMICAL"</span>, <span class="st">"Carcinogen_liv"</span><span class="op">)</span><span class="op">]</span>
<span class="va">discoveryCHEMICALindex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/createDataPartition.html">createDataPartition</a></span><span class="op">(</span><span class="va">CHEMICALGROUP</span><span class="op">$</span><span class="va">Carcinogen_liv</span>, p<span class="op">=</span><span class="fl">0.6</span>, list<span class="op">=</span><span class="cn">FALSE</span>, times<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
<span class="va">discoveryCHEMICAL</span> <span class="op">&lt;-</span> <span class="va">CHEMICALGROUP</span><span class="op">$</span><span class="va">CHEMICAL</span><span class="op">[</span><span class="va">discoveryCHEMICALindex</span><span class="op">]</span>
<span class="va">validationCHEMICAL</span> <span class="op">&lt;-</span> <span class="va">CHEMICALGROUP</span><span class="op">$</span><span class="va">CHEMICAL</span><span class="op">[</span><span class="op">-</span><span class="va">discoveryCHEMICALindex</span><span class="op">]</span>

<span class="co">## split data into discovery and validation set (all replicates belong to one or the other)</span>
<span class="va">DM1000discovery</span> <span class="op">&lt;-</span> <span class="va">DM1000</span><span class="op">[</span>,<span class="va">DM1000</span><span class="op">$</span><span class="va">CHEMICAL</span> <span class="op">%in%</span> <span class="va">discoveryCHEMICAL</span><span class="op">]</span>
<span class="va">DM1000validation</span> <span class="op">&lt;-</span> <span class="va">DM1000</span><span class="op">[</span>,<span class="va">DM1000</span><span class="op">$</span><span class="va">CHEMICAL</span> <span class="op">%in%</span> <span class="va">validationCHEMICAL</span><span class="op">]</span>

<span class="va">discovery</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">DM1000discovery</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">discoveryLab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">DM1000discovery</span><span class="op">$</span><span class="va">Carcinogen_liv</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"noncarc"</span>, <span class="st">"carc"</span><span class="op">)</span><span class="op">)</span>

<span class="va">validation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">DM1000validation</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">validationLab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">DM1000validation</span><span class="op">$</span><span class="va">Carcinogen_liv</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"noncarc"</span>, <span class="st">"carc"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="model-fitting-on-the-training-set" class="section level3">
<h3 class="hasAnchor">
<a href="#model-fitting-on-the-training-set" class="anchor"></a>Model Fitting on the Training Set</h3>
<p>We now fit a KNN classifier. We evaluate several versions, based on different numbers of neighbors, and evaluate the performance of each by 5-fold cross-validation (cv).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## KNN with 5x cross validation</span>
<span class="va">fitControl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/trainControl.html">trainControl</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"cv"</span>,
                           number<span class="op">=</span><span class="fl">5</span>,
                           classProbs<span class="op">=</span><span class="cn">T</span>,
                           summaryFunction<span class="op">=</span><span class="va">twoClassSummary</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span> <span class="co"># for reproducible results</span>

<span class="co">## evaluate on train set based on area under the ROC (AUC)</span>
<span class="va">KNN</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/train.html">train</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">discovery</span>,
             y<span class="op">=</span><span class="va">discoveryLab</span>,
             method<span class="op">=</span><span class="st">"knn"</span>,
             trControl<span class="op">=</span><span class="va">fitControl</span>,
             tuneGrid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span>.k<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">7</span>,<span class="fl">9</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span>,
             metric<span class="op">=</span><span class="st">'ROC'</span><span class="op">)</span>
<span class="co">## summary of performance across each value of tuning parameters</span>
<span class="va">KNN</span></code></pre></div>
<pre><code>## k-Nearest Neighbors 
## 
##   63 samples
## 1000 predictors
##    2 classes: 'noncarc', 'carc' 
## 
## No pre-processing
## Resampling: Cross-Validated (5 fold) 
## Summary of sample sizes: 51, 51, 50, 50, 50 
## Resampling results across tuning parameters:
## 
##   k   ROC        Sens       Spec
##    3  0.9521429  0.9714286  0.92
##    5  0.9717857  0.9464286  0.96
##    7  0.9797321  0.9464286  1.00
##    9  0.9750000  0.9464286  0.88
##   20  0.8258929  0.7964286  0.47
## 
## ROC was used to select the optimal model using the largest value.
## The final value used for the model was k = 7.</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/pROC/man/plot.roc.html">plot</a></span><span class="op">(</span><span class="va">KNN</span>, metric <span class="op">=</span> <span class="st">"ROC"</span><span class="op">)</span></code></pre></div>
<p><img src="Classification_Caret_files/figure-html/classify_3-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## show the parameters that yield the most accurate classifier (as estimated by cv)</span>
<span class="va">KNN</span><span class="op">$</span><span class="va">bestTune</span></code></pre></div>
<pre><code>##   k
## 3 7</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#final model used and performance across test sets of each fold</span>
<span class="va">KNN</span><span class="op">$</span><span class="va">finalModel</span></code></pre></div>
<pre><code>## 7-nearest neighbor model
## Training set outcome distribution:
## 
## noncarc    carc 
##      39      24</code></pre>
<p><strong>WARNING</strong>: It should be noted that the internal cv performed by caret in this case does not take into account the problem of multiple replicates per chemical. Thus the cv estimates of accuracy (based on AUC) will in general be over-optimistic. However, we will hope that all the estimates will be equally over-optimistic, thus allowing us to select the best model parameterization.</p>
</div>
<div id="validation-on-the-test-set" class="section level3">
<h3 class="hasAnchor">
<a href="#validation-on-the-test-set" class="anchor"></a>Validation on the Test Set</h3>
<p>After having selected the “best” classifier, we can now apply it to the validation set.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## predicting the validation data:</span>
<span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">KNN</span>,<span class="va">validation</span><span class="op">)</span>

<span class="co">## or predicting using the probabilities (nice because you can get ROC)</span>
<span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/predict.train.html">extractProb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>model<span class="op">=</span><span class="va">KNN</span><span class="op">)</span>,
                     testX<span class="op">=</span><span class="va">validation</span>,
                     testY<span class="op">=</span><span class="va">validationLab</span><span class="op">)</span>

<span class="co">## removing trainings data</span>
<span class="va">probs</span> <span class="op">&lt;-</span> <span class="va">probs</span><span class="op">[</span><span class="va">probs</span><span class="op">$</span><span class="va">dataType</span><span class="op">!=</span><span class="st">'Training'</span>,<span class="op">]</span>

<span class="co">## Make sure the levels are appropriate for twoClassSummary(), ie case group is first level</span>
<span class="va">levs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"carc"</span>, <span class="st">"noncarc"</span><span class="op">)</span>
<span class="va">probs</span><span class="op">$</span><span class="va">obs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">probs</span><span class="op">$</span><span class="va">obs</span>, levels <span class="op">=</span> <span class="va">levs</span><span class="op">)</span>
<span class="va">probs</span><span class="op">$</span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">probs</span><span class="op">$</span><span class="va">pred</span>, levels <span class="op">=</span> <span class="va">levs</span><span class="op">)</span>

<span class="co">## Calculating Accuracy</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">probs</span><span class="op">$</span><span class="va">obs</span><span class="op">==</span><span class="va">probs</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.48</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## see classification prob for each sample in validation set</span>
<span class="co">## pred column shows model predicted label if cutoff for calling label = 0.5</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">probs</span><span class="op">$</span><span class="va">obs</span>, <span class="va">probs</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></code></pre></div>
<pre><code>##          
##           carc noncarc
##   carc       3      23
##   noncarc    3      21</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## summary of performance result on validation set</span>
<span class="fu"><a href="https://rdrr.io/pkg/caret/man/postResample.html">twoClassSummary</a></span><span class="op">(</span><span class="va">probs</span>, lev <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">probs</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##       ROC      Sens      Spec 
## 0.3421474 0.1153846 0.8750000</code></pre>
</div>
<div id="random-forest" class="section level3">
<h3 class="hasAnchor">
<a href="#random-forest" class="anchor"></a>Random Forest</h3>
<div id="model-fitting-on-the-training-set-1" class="section level4">
<h4 class="hasAnchor">
<a href="#model-fitting-on-the-training-set-1" class="anchor"></a>Model Fitting on the Training Set</h4>
<p>Next we fit a Random Forest classifier. We evaluate several versions, based on different numbers of random features to be used for each tree, and evaluate the performance of each by 5-fold cross-validation (cv). Note, that a commonly features per tree is <span class="math inline">\(\sqrt{total\; features}\)</span>, in this case <span class="math inline">\(\sqrt{total\; features} = \sqrt{1000} \sim 32\)</span>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## RandomForest with 5x cross validation</span>
<span class="va">fitControl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/trainControl.html">trainControl</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"cv"</span>,
                           number<span class="op">=</span><span class="fl">5</span>,
                           classProbs<span class="op">=</span><span class="cn">T</span>,
                           summaryFunction<span class="op">=</span><span class="va">twoClassSummary</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span> <span class="co"># for reproducible results</span>

<span class="co">## evaluate on train set based on area under the ROC (AUC)</span>
<span class="va">RF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/train.html">train</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">discovery</span>,
             y<span class="op">=</span><span class="va">discoveryLab</span>,
             method<span class="op">=</span><span class="st">"rf"</span>,
             trControl<span class="op">=</span><span class="va">fitControl</span>,
             tuneGrid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span>mtry<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">32</span>, <span class="fl">50</span>, <span class="fl">100</span>, <span class="fl">250</span>, <span class="fl">500</span><span class="op">)</span><span class="op">)</span>,
             metric<span class="op">=</span><span class="st">'ROC'</span><span class="op">)</span>
<span class="co">## summary of performance across each value of tuning parameters</span>
<span class="va">RF</span></code></pre></div>
<pre><code>## Random Forest 
## 
##   63 samples
## 1000 predictors
##    2 classes: 'noncarc', 'carc' 
## 
## No pre-processing
## Resampling: Cross-Validated (5 fold) 
## Summary of sample sizes: 51, 51, 50, 50, 50 
## Resampling results across tuning parameters:
## 
##   mtry  ROC        Sens       Spec
##    32   0.9735714  0.9714286  0.88
##    50   0.9785714  0.9714286  0.88
##   100   0.9703571  0.9714286  0.88
##   250   0.9392857  0.9464286  0.84
##   500   0.9400000  0.9214286  0.84
## 
## ROC was used to select the optimal model using the largest value.
## The final value used for the model was mtry = 50.</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/pROC/man/plot.roc.html">plot</a></span><span class="op">(</span><span class="va">RF</span>, metric <span class="op">=</span> <span class="st">"ROC"</span><span class="op">)</span></code></pre></div>
<p><img src="Classification_Caret_files/figure-html/classify_5-1.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## show the parameters that yield the most accurate classifier (as estimated by cv)</span>
<span class="va">RF</span><span class="op">$</span><span class="va">bestTune</span></code></pre></div>
<pre><code>##   mtry
## 2   50</code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#final model used and performance across test sets of each fold</span>
<span class="va">RF</span><span class="op">$</span><span class="va">finalModel</span></code></pre></div>
<pre><code>## 
## Call:
##  randomForest(x = x, y = y, mtry = param$mtry) 
##                Type of random forest: classification
##                      Number of trees: 500
## No. of variables tried at each split: 50
## 
##         OOB estimate of  error rate: 6.35%
## Confusion matrix:
##         noncarc carc class.error
## noncarc      38    1  0.02564103
## carc          3   21  0.12500000</code></pre>
</div>
<div id="validation-on-the-test-set-1" class="section level4">
<h4 class="hasAnchor">
<a href="#validation-on-the-test-set-1" class="anchor"></a>Validation on the Test Set</h4>
<p>All of the classifiers work well on the training data. A nice thing about Random Forests are that they’re performance is fairly robust to different parameter choices We’ll end using use <span class="math inline">\(\sqrt{total\; features} = \sqrt{1000} \sim 32\)</span> , we can now apply it to the validation set.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## predicting the validation data:</span>
<span class="va">predRF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">RF</span>,<span class="va">validation</span><span class="op">)</span>

<span class="co">## or predicting using the probabilities (nice because you can get ROC)</span>
<span class="va">probsRF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/predict.train.html">extractProb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>model<span class="op">=</span><span class="va">RF</span><span class="op">)</span>,
                     testX<span class="op">=</span><span class="va">validation</span>,
                     testY<span class="op">=</span><span class="va">validationLab</span><span class="op">)</span>

<span class="co">## removing trainings data</span>
<span class="va">probsRF</span> <span class="op">&lt;-</span> <span class="va">probsRF</span><span class="op">[</span><span class="va">probsRF</span><span class="op">$</span><span class="va">dataType</span><span class="op">!=</span><span class="st">'Training'</span>,<span class="op">]</span>

<span class="co">## Make sure the levels are appropriate for twoClassSummary(), ie case group is first level</span>
<span class="va">levs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"carc"</span>, <span class="st">"noncarc"</span><span class="op">)</span>
<span class="va">probsRF</span><span class="op">$</span><span class="va">obs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">probsRF</span><span class="op">$</span><span class="va">obs</span>, levels <span class="op">=</span> <span class="va">levs</span><span class="op">)</span>
<span class="va">probsRF</span><span class="op">$</span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">probsRF</span><span class="op">$</span><span class="va">pred</span>, levels <span class="op">=</span> <span class="va">levs</span><span class="op">)</span>

<span class="co">## Calculating Accuracy</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">probsRF</span><span class="op">$</span><span class="va">obs</span><span class="op">==</span><span class="va">probsRF</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.4</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## see classification prob for each sample in validation set</span>
<span class="co">## pred column shows model predicted label if cutoff for calling label = 0.5</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">probsRF</span><span class="op">$</span><span class="va">obs</span>, <span class="va">probsRF</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></code></pre></div>
<pre><code>##          
##           carc noncarc
##   carc       0      26
##   noncarc    4      20</code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## summary of performance result on validation set</span>
<span class="fu"><a href="https://rdrr.io/pkg/caret/man/postResample.html">twoClassSummary</a></span><span class="op">(</span><span class="va">probsRF</span>, lev <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">probsRF</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##       ROC      Sens      Spec 
## 0.1762821 0.0000000 0.8333333</code></pre>
</div>
</div>
<div id="support-vecor-machine-linear-kernel" class="section level3">
<h3 class="hasAnchor">
<a href="#support-vecor-machine-linear-kernel" class="anchor"></a>Support Vecor Machine (Linear Kernel)</h3>
<div id="model-fitting-on-the-training-set-2" class="section level4">
<h4 class="hasAnchor">
<a href="#model-fitting-on-the-training-set-2" class="anchor"></a>Model Fitting on the Training Set</h4>
<p>Next we fit a SVM classifier with a linear. We evaluate several versions, based on different cost parameters, and evaluate the performance of each by 5-fold cross-validation (cv). SVM attempts to separate each group by a hyperplane over all variables by maximizing the distance (or margin) between this hyperplane and the closest points to it. The cost parameter controls the leniency of certain points falling on the wrong side of this hyperplane if it means the margin of the hyperplane to other points to be greater. This is to control overfitting. Large cost parameters don’t allow for any points in the training data to be misclassified.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## SVM with 5x cross validation</span>
<span class="va">fitControl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/trainControl.html">trainControl</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"cv"</span>,
                           number<span class="op">=</span><span class="fl">5</span>,
                           classProbs<span class="op">=</span><span class="cn">T</span>,
                           summaryFunction<span class="op">=</span><span class="va">twoClassSummary</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span> <span class="co"># for reproducible results</span>

<span class="co">## evaluate on train set based on area under the ROC (AUC)</span>
<span class="va">SVM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/train.html">train</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">discovery</span>,
             y<span class="op">=</span><span class="va">discoveryLab</span>,
             method<span class="op">=</span><span class="st">"svmLinear2"</span>,
             trControl<span class="op">=</span><span class="va">fitControl</span>,
             tuneGrid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span>cost<span class="op">=</span><span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4.5</span>, <span class="op">-</span><span class="fl">3</span>, by <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
             metric<span class="op">=</span><span class="st">'ROC'</span><span class="op">)</span>
<span class="co">## summary of performance across each value of tuning parameters</span>
<span class="va">SVM</span></code></pre></div>
<pre><code>## Support Vector Machines with Linear Kernel 
## 
##   63 samples
## 1000 predictors
##    2 classes: 'noncarc', 'carc' 
## 
## No pre-processing
## Resampling: Cross-Validated (5 fold) 
## Summary of sample sizes: 51, 51, 50, 50, 50 
## Resampling results across tuning parameters:
## 
##   cost          ROC        Sens       Spec
##   3.162278e-05  0.9621429  1.0000000  0.00
##   3.548134e-05  0.9621429  1.0000000  0.00
##   3.981072e-05  0.9621429  1.0000000  0.00
##   4.466836e-05  0.9621429  1.0000000  0.00
##   5.011872e-05  0.9621429  1.0000000  0.00
##   5.623413e-05  0.9621429  1.0000000  0.00
##   6.309573e-05  0.9621429  1.0000000  0.00
##   7.079458e-05  0.9621429  1.0000000  0.00
##   7.943282e-05  0.9621429  1.0000000  0.00
##   8.912509e-05  0.9621429  1.0000000  0.00
##   1.000000e-04  0.9621429  1.0000000  0.00
##   1.122018e-04  0.9621429  1.0000000  0.00
##   1.258925e-04  0.9621429  1.0000000  0.00
##   1.412538e-04  0.9621429  1.0000000  0.00
##   1.584893e-04  0.9621429  1.0000000  0.00
##   1.778279e-04  0.9621429  1.0000000  0.00
##   1.995262e-04  0.9621429  1.0000000  0.00
##   2.238721e-04  0.7621429  1.0000000  0.10
##   2.511886e-04  0.9621429  1.0000000  0.15
##   2.818383e-04  0.9621429  1.0000000  0.19
##   3.162278e-04  0.9621429  1.0000000  0.36
##   3.548134e-04  0.9621429  0.9214286  0.60
##   3.981072e-04  0.9621429  0.9214286  0.84
##   4.466836e-04  0.9621429  0.9464286  0.88
##   5.011872e-04  0.9621429  0.9714286  0.88
##   5.623413e-04  0.9621429  0.9714286  0.88
##   6.309573e-04  0.9678571  0.9714286  0.88
##   7.079458e-04  0.9678571  0.9714286  0.88
##   7.943282e-04  0.9678571  0.9714286  0.92
##   8.912509e-04  0.9678571  0.9714286  0.92
##   1.000000e-03  0.9785714  0.9714286  0.92
## 
## ROC was used to select the optimal model using the largest value.
## The final value used for the model was cost = 0.001.</code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/pROC/man/plot.roc.html">plot</a></span><span class="op">(</span><span class="va">SVM</span>, metric <span class="op">=</span> <span class="st">"ROC"</span><span class="op">)</span></code></pre></div>
<p><img src="Classification_Caret_files/figure-html/classify_7-1.png" width="700"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## show the parameters that yield the most accurate classifier (as estimated by cv)</span>
<span class="va">SVM</span><span class="op">$</span><span class="va">bestTune</span></code></pre></div>
<pre><code>##     cost
## 31 0.001</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#final model used and performance across test sets of each fold</span>
<span class="va">SVM</span><span class="op">$</span><span class="va">finalModel</span></code></pre></div>
<pre><code>## 
## Call:
## svm.default(x = as.matrix(x), y = y, kernel = "linear", cost = param$cost, 
##     probability = classProbs)
## 
## 
## Parameters:
##    SVM-Type:  C-classification 
##  SVM-Kernel:  linear 
##        cost:  0.001 
## 
## Number of Support Vectors:  48</code></pre>
</div>
<div id="validation-on-the-test-set-2" class="section level4">
<h4 class="hasAnchor">
<a href="#validation-on-the-test-set-2" class="anchor"></a>Validation on the Test Set</h4>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## predicting the validation data:</span>
<span class="va">predSVM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">SVM</span>,<span class="va">validation</span><span class="op">)</span>

<span class="co">## or predicting using the probabilities (nice because you can get ROC)</span>
<span class="va">probsSVM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/predict.train.html">extractProb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>model<span class="op">=</span><span class="va">SVM</span><span class="op">)</span>,
                     testX<span class="op">=</span><span class="va">validation</span>,
                     testY<span class="op">=</span><span class="va">validationLab</span><span class="op">)</span>

<span class="co">## removing trainings data</span>
<span class="va">probsSVM</span> <span class="op">&lt;-</span> <span class="va">probsSVM</span><span class="op">[</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">dataType</span><span class="op">!=</span><span class="st">'Training'</span>,<span class="op">]</span>

<span class="co">## Make sure the levels are appropriate for twoClassSummary(), ie case group is first level</span>
<span class="va">levs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"carc"</span>, <span class="st">"noncarc"</span><span class="op">)</span>
<span class="va">probsSVM</span><span class="op">$</span><span class="va">obs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">obs</span>, levels <span class="op">=</span> <span class="va">levs</span><span class="op">)</span>
<span class="va">probsSVM</span><span class="op">$</span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">pred</span>, levels <span class="op">=</span> <span class="va">levs</span><span class="op">)</span>

<span class="co">## Calculating Accuracy</span>
<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">obs</span><span class="op">==</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.36</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## see classification prob for each sample in validation set</span>
<span class="co">## pred column shows model predicted label if cutoff for calling label = 0.5</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">obs</span>, <span class="va">probsSVM</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></code></pre></div>
<pre><code>##          
##           carc noncarc
##   carc       6      20
##   noncarc   12      12</code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## summary of performance result on validation set</span>
<span class="fu"><a href="https://rdrr.io/pkg/caret/man/postResample.html">twoClassSummary</a></span><span class="op">(</span><span class="va">probsSVM</span>, lev <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">probsSVM</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##       ROC      Sens      Spec 
## 0.1858974 0.2307692 0.5000000</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/lia978">Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
