<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene Expression Differential Analysis based on Limma • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Gene Expression Differential Analysis based on Limma">
<meta property="og:description" content="BS831">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene Expression Differential Analysis based on Limma</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/Diffanal_LimmaBC.Rmd"><code>vignettes/docs/Diffanal_LimmaBC.Rmd</code></a></small>
      <div class="hidden name"><code>Diffanal_LimmaBC.Rmd</code></div>

    </div>

    
    
<p>Here, we present a couple of simple examples of differential analysis based on <code>limma</code>. In particular, we show how the design matrix can be constructed using different ‘codings’ of the regression variables. We also define a simple wrapper function that can help us remember the different <code>limma</code> steps.</p>
<p>This document also provides examples of how to embed equations into an R markdown document, and how to cross-reference those equations.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">BS831</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">Biobase</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">limma</span>)</pre></body></html></div>
<div id="load-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#load-the-data" class="anchor"></a>Load the Data</h3>
<p>We start by loading and pre-processing the necessary expression dataset.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">## let us load the ExpressionSet, an already simplified version of the</span>
<span class="co">## ..breast cancer dataset</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">renamedBreastDB</span>)

<span class="no">dat</span> <span class="kw">&lt;-</span> <span class="no">renamedBreastDB</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span>(<span class="no">dat</span>))</pre></body></html></div>
<pre><code>##          individual diseaseState
## GSM85500        T44     nonbasal
## GSM85504       T175     nonbasal
## GSM85493        T37     nonbasal
## GSM85496       T161     nonbasal
## GSM85510       T162     nonbasal
## GSM85509        T74     nonbasal</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">dat</span>$<span class="no">diseaseState</span>)</pre></body></html></div>
<pre><code>## 
## nonbasal    basal 
##       20       18</code></pre>
</div>
<div id="differential-analysis-based-on-limma" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-analysis-based-on-limma" class="anchor"></a>Differential Analysis based on <code>Limma</code>
</h3>
<p>When the regression variable is categorical (binary in this case), we can choose different (yet equivalent) ‘codings’. In particular, we can fit a standard model</p>
<p><span class="math display">\[
\begin{equation}
y = \beta_0 + \beta_1 X_{group},
\label{eq:lm1}
\tag{1}
\end{equation}
\]</span></p>
<p>where <span class="math inline">\(X_{group} = 0,1,\)</span> if the observation is from a nonbasal- or a basal-type tumor, respectively. Alternatively, we can fit the following model</p>
<p><span class="math display">\[
\begin{equation}
y = \beta_{\not{basal}} X_{\not{basal}} + \beta_{basal}X_{basal},
\label{eq:lm2}
\tag{2}
\end{equation}
\]</span></p>
<p>where <span class="math inline">\(X_{\not{basal}} = 1,0\)</span>, if the observation is from a nonbasal- or a basal-type tumor, respectively. Conversely, <span class="math inline">\(X_{basal} = 0,1\)</span>, if the observation is from a nonbasal- or a basal-type tumor, respectively.</p>
<p>The corresponding models in <code>limma</code> are defined based on the two design matrices <code>design1</code>, for model <span class="math inline">\(\eqref{eq:lm1}\)</span>, and <code>design2</code>, for model <span class="math inline">\(\eqref{eq:lm2}\)</span>, as follows.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">design1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>( ~ <span class="no">diseaseState</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span>(<span class="no">dat</span>) )
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">design1</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"baseline"</span>,<span class="st">"basal"</span>)      <span class="co"># let's simplify column names  </span>
<span class="no">design2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>( ~ <span class="fl">0</span> + <span class="no">dat</span>$<span class="no">diseaseState</span> )
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">design2</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>( <span class="no">dat</span>$<span class="no">diseaseState</span> ) <span class="co"># ditto</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">design1</span>)) <span class="co"># showing only one instance of each class</span></pre></body></html></div>
<pre><code>##          baseline basal
## GSM85500        1     0
## GSM85480        1     1</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">design2</span>)) <span class="co"># ditto</span></pre></body></html></div>
<pre><code>##    nonbasal basal
## 1         1     0
## 21        0     1</code></pre>
<p>The different designs require slightly different approaches to model fitting. In model <span class="math inline">\(\eqref{eq:lm1}\)</span>, we want to extract the p-value corresponding to the <span class="math inline">\(\beta_{group}\)</span> parameter. In model <span class="math inline">\(\eqref{eq:lm2}\)</span>, we care about the <em>difference</em> <span class="math inline">\(\beta_{basal} - \beta_{\not{basal}}\)</span>.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co">## model (1)</span>
<span class="no">fit1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span>(<span class="no">dat</span>,<span class="no">design1</span>) <span class="co"># fitting of linear model</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fit1</span>$<span class="no">coefficients</span>)</pre></body></html></div>
<pre><code>##          baseline     basal
## AGR3     9.241425 -5.476069
## SCGB2A2  9.049974 -4.020716
## AGR2    10.349870 -5.139503
## CALML5   6.314069  2.564793
## KRT23    6.348392  3.193484
## FOXA1    7.215394 -3.997576</code></pre>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">fit1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span>(<span class="no">fit1</span>)       <span class="co"># pooling of variance across like-genes</span>
<span class="no">fit1table</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span>(<span class="no">fit1</span>,<span class="kw">coef</span><span class="kw">=</span><span class="st">"basal"</span>,<span class="kw">adjust</span><span class="kw">=</span><span class="st">"BH"</span>,<span class="kw">number</span><span class="kw">=</span><span class="fl">Inf</span>,<span class="kw">sort.by</span><span class="kw">=</span><span class="st">"P"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fit1table</span>)</pre></body></html></div>
<pre><code>##       affy_hg_u133_plus_2 hgnc_symbol
## MLPH          218211_s_at        MLPH
## FOXA1           237086_at       FOXA1
## FOXC1        1553613_s_at       FOXC1
## AGR2            209173_at        AGR2
## SFRP1         202035_s_at       SFRP1
## SIDT1           219734_at       SIDT1
##                                                                   description
## MLPH                         melanophilin [Source:HGNC Symbol;Acc:HGNC:29643]
## FOXA1                      forkhead box A1 [Source:HGNC Symbol;Acc:HGNC:5021]
## FOXC1                      forkhead box C1 [Source:HGNC Symbol;Acc:HGNC:3800]
## AGR2                    anterior gradient 2 [Source:HGNC Symbol;Acc:HGNC:328]
## SFRP1 secreted frizzled-related protein 1 [Source:HGNC Symbol;Acc:HGNC:10776]
## SIDT1 SID1 transmembrane family, member 1 [Source:HGNC Symbol;Acc:HGNC:25967]
##           logFC  AveExpr         t      P.Value    adj.P.Val        B
## MLPH  -4.539281 9.487324 -15.77794 1.198422e-19 2.539912e-16 34.35447
## FOXA1 -3.997576 5.321805 -15.63213 1.693275e-19 2.539912e-16 34.02400
## FOXC1  3.774299 7.847670  14.27359 4.745041e-18 4.745041e-15 30.82304
## AGR2  -5.139503 7.915368 -12.95288 1.488409e-16 1.116307e-13 27.48939
## SFRP1  4.144489 8.377033  12.32914 8.162432e-16 4.897459e-13 25.83515
## SIDT1 -2.605517 6.013343 -12.23990 1.045491e-15 5.227453e-13 25.59416</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co">## model (2)</span>
<span class="no">fit2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span>(<span class="no">dat</span>,<span class="no">design2</span>) <span class="co"># fitting of linear model</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fit2</span>$<span class="no">coefficients</span>)</pre></body></html></div>
<pre><code>##          nonbasal    basal
## AGR3     9.241425 3.765357
## SCGB2A2  9.049974 5.029258
## AGR2    10.349870 5.210367
## CALML5   6.314069 8.878863
## KRT23    6.348392 9.541876
## FOXA1    7.215394 3.217818</code></pre>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co">## let's define the contrast matrix we need</span>
<span class="no">contrast.matrix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html">makeContrasts</a></span>((<span class="no">basal</span>-<span class="no">nonbasal</span>),<span class="kw">levels</span><span class="kw">=</span><span class="no">design2</span>)
<span class="no">fit2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span>(<span class="no">fit2</span>,<span class="no">contrast.matrix</span>)
<span class="co">## notice that the contrast coefficient takes the same values as the basal coefficient in design1</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fit2</span>$<span class="no">coefficients</span>)</pre></body></html></div>
<pre><code>##          Contrasts
##           (basal - nonbasal)
##   AGR3             -5.476069
##   SCGB2A2          -4.020716
##   AGR2             -5.139503
##   CALML5            2.564793
##   KRT23             3.193484
##   FOXA1            -3.997576</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">fit2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span>(<span class="no">fit2</span>)       <span class="co"># pooling of variance across like-genes</span>
<span class="no">fit2table</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span>(<span class="no">fit2</span>,<span class="kw">coef</span><span class="kw">=</span><span class="fl">1</span>,<span class="kw">adjust</span><span class="kw">=</span><span class="st">"BH"</span>,<span class="kw">number</span><span class="kw">=</span><span class="fl">Inf</span>,<span class="kw">sort.by</span><span class="kw">=</span><span class="st">"P"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fit2table</span>)</pre></body></html></div>
<pre><code>##       affy_hg_u133_plus_2 hgnc_symbol
## MLPH          218211_s_at        MLPH
## FOXA1           237086_at       FOXA1
## FOXC1        1553613_s_at       FOXC1
## AGR2            209173_at        AGR2
## SFRP1         202035_s_at       SFRP1
## SIDT1           219734_at       SIDT1
##                                                                   description
## MLPH                         melanophilin [Source:HGNC Symbol;Acc:HGNC:29643]
## FOXA1                      forkhead box A1 [Source:HGNC Symbol;Acc:HGNC:5021]
## FOXC1                      forkhead box C1 [Source:HGNC Symbol;Acc:HGNC:3800]
## AGR2                    anterior gradient 2 [Source:HGNC Symbol;Acc:HGNC:328]
## SFRP1 secreted frizzled-related protein 1 [Source:HGNC Symbol;Acc:HGNC:10776]
## SIDT1 SID1 transmembrane family, member 1 [Source:HGNC Symbol;Acc:HGNC:25967]
##           logFC  AveExpr         t      P.Value    adj.P.Val        B
## MLPH  -4.539281 9.487324 -15.77794 1.198422e-19 2.539912e-16 34.35447
## FOXA1 -3.997576 5.321805 -15.63213 1.693275e-19 2.539912e-16 34.02400
## FOXC1  3.774299 7.847670  14.27359 4.745041e-18 4.745041e-15 30.82304
## AGR2  -5.139503 7.915368 -12.95288 1.488409e-16 1.116307e-13 27.48939
## SFRP1  4.144489 8.377033  12.32914 8.162432e-16 4.897459e-13 25.83515
## SIDT1 -2.605517 6.013343 -12.23990 1.045491e-15 5.227453e-13 25.59416</code></pre>
<p>Regardless of the ‘coding’, the results are the same.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">## compare the results from the two 'codings'</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(<span class="no">fit2table</span>[,<span class="st">"t"</span>],<span class="no">fit1table</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">fit2table</span>),<span class="st">"t"</span>])</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">fit2table</span>[,<span class="st">"t"</span>],<span class="no">fit1table</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">fit2table</span>),<span class="st">"t"</span>],<span class="kw">xlab</span><span class="kw">=</span><span class="st">"design2"</span>,<span class="kw">ylab</span><span class="kw">=</span><span class="st">"design1"</span>)</pre></body></html></div>
<p><img src="Diffanal_LimmaBC_files/figure-html/limma.compare-1.png" width="700"></p>
</div>
<div id="defining-a-wrapper-function" class="section level3">
<h3 class="hasAnchor">
<a href="#defining-a-wrapper-function" class="anchor"></a>Defining a wrapper function</h3>
<p>Since running <code>limma</code> requires memorizing several commands that are not easy to remember, here we define a simple wrapper function that will relief us from remembering them all. Incidentally, this is also (a simplified version of) the wrapper function defined in <code>CBMRtools</code>, and it corresponds to the <code>design2</code>, model <span class="math inline">\(\eqref{eq:lm2}\)</span>, shown above.</p>
<p>It should be noted that this simple wrapper function does not allow you to specify additional covariates (confounders). In that case, you will have to either define a more advanced wrapper (any volunteer?), or revert back to the direct use of limma commands.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">run_limma</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(
    <span class="no">eset</span>,                <span class="co"># the expression set</span>
    <span class="no">treatment</span>,           <span class="co"># pData variable name of the phenotype</span>
    <span class="no">cond</span>,                <span class="co"># label of the 'treatment' group</span>
    <span class="no">control</span>,             <span class="co"># label of the 'control' group</span>
    <span class="no">sort.by</span><span class="kw">=</span><span class="st">"P"</span>,         <span class="co"># sort output by given criterion</span>
    <span class="no">decreasing</span> <span class="kw">=</span> <span class="fl">FALSE</span>,  <span class="co"># sort order</span>
    <span class="no">cutoff</span> <span class="kw">=</span> <span class="fl">NA</span>          <span class="co"># extract only genes above/below 'sort.by' cutoff</span>
)
{
    <span class="kw">if</span> ( <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(<span class="no">eset</span>))<span class="kw">&lt;</span><span class="fl">1</span> )
        <span class="fu"><a href="https://rdrr.io/r/base/stop.html">stop</a></span>( <span class="st">"empty fData"</span> )

    <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>( <span class="st">"running limma\n"</span> )
    <span class="no">test_name</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">control</span>, <span class="st">"_vs_"</span>, <span class="no">cond</span>, <span class="kw">sep</span> <span class="kw">=</span> <span class="st">""</span>)
    <span class="no">treatmentvec</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span>(<span class="no">eset</span>)[, <span class="no">treatment</span>]
    <span class="no">eset</span> <span class="kw">&lt;-</span> <span class="no">eset</span>[, <span class="no">treatmentvec</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">cond</span>, <span class="no">control</span>)]
    <span class="no">treatmentvec</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>( <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span>(<span class="no">eset</span>)[, <span class="no">treatment</span>] )
    <span class="no">design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>( ~ <span class="fl">0</span> + <span class="no">treatmentvec</span> )
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">design</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>( <span class="no">treatmentvec</span> )
    <span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span>(<span class="no">eset</span>, <span class="no">design</span>)
    <span class="no">command_str</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"makeContrasts("</span>, <span class="st">"("</span>, <span class="no">cond</span> , <span class="st">"-"</span>, <span class="no">control</span>, <span class="st">")"</span>,
                         <span class="st">",levels = design)"</span>, <span class="kw">sep</span> <span class="kw">=</span> <span class="st">""</span>)
    <span class="no">contrast.matrix</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/eval.html">eval</a></span>( <span class="fu"><a href="https://rdrr.io/r/base/parse.html">parse</a></span>(<span class="kw">text</span><span class="kw">=</span><span class="no">command_str</span>) )
    <span class="no">fit2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span>( <span class="no">fit</span>, <span class="no">contrast.matrix</span> )
    <span class="no">fit2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span>(<span class="no">fit2</span>)

    <span class="co">## extract full table</span>
    <span class="no">fit2.table</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span>(<span class="no">fit2</span>, <span class="kw">coef</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">adjust</span><span class="kw">=</span><span class="st">"BH"</span>, <span class="kw">number</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">fit2</span>), <span class="kw">sort.by</span><span class="kw">=</span><span class="no">sort.by</span>)

    <span class="co">## extract genes above/below cutoff</span>
    <span class="kw">if</span>( !<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">cutoff</span>) ) {
        <span class="kw">if</span> (<span class="no">decreasing</span>) {
            <span class="no">fit2.table</span> <span class="kw">&lt;-</span> <span class="no">fit2.table</span>[<span class="no">fit2.table</span>[,<span class="no">sort.by</span>] <span class="kw">&gt;</span> <span class="no">cutoff</span>,]
        }
        <span class="kw">else</span> {
            <span class="no">fit2.table</span> <span class="kw">&lt;-</span> <span class="no">fit2.table</span>[<span class="no">fit2.table</span>[,<span class="no">sort.by</span>] <span class="kw">&lt;</span> <span class="no">cutoff</span>,]
        }
    }
    <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">fit2.table</span>)
}</pre></body></html></div>
<p>Now, let us test the newly defined function on our data.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">fit3table</span> <span class="kw">&lt;-</span> <span class="fu">run_limma</span>(<span class="no">dat</span>,<span class="kw">treatment</span><span class="kw">=</span><span class="st">"diseaseState"</span>,<span class="kw">cond</span><span class="kw">=</span><span class="st">"basal"</span>,<span class="kw">control</span><span class="kw">=</span><span class="st">"nonbasal"</span>)</pre></body></html></div>
<pre><code>## running limma</code></pre>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">fit3table</span>)</pre></body></html></div>
<pre><code>##       affy_hg_u133_plus_2 hgnc_symbol
## MLPH          218211_s_at        MLPH
## FOXA1           237086_at       FOXA1
## FOXC1        1553613_s_at       FOXC1
## AGR2            209173_at        AGR2
## SFRP1         202035_s_at       SFRP1
## SIDT1           219734_at       SIDT1
##                                                                   description
## MLPH                         melanophilin [Source:HGNC Symbol;Acc:HGNC:29643]
## FOXA1                      forkhead box A1 [Source:HGNC Symbol;Acc:HGNC:5021]
## FOXC1                      forkhead box C1 [Source:HGNC Symbol;Acc:HGNC:3800]
## AGR2                    anterior gradient 2 [Source:HGNC Symbol;Acc:HGNC:328]
## SFRP1 secreted frizzled-related protein 1 [Source:HGNC Symbol;Acc:HGNC:10776]
## SIDT1 SID1 transmembrane family, member 1 [Source:HGNC Symbol;Acc:HGNC:25967]
##           logFC  AveExpr         t      P.Value    adj.P.Val        B
## MLPH  -4.539281 9.487324 -15.77794 1.198422e-19 2.539912e-16 34.35447
## FOXA1 -3.997576 5.321805 -15.63213 1.693275e-19 2.539912e-16 34.02400
## FOXC1  3.774299 7.847670  14.27359 4.745041e-18 4.745041e-15 30.82304
## AGR2  -5.139503 7.915368 -12.95288 1.488409e-16 1.116307e-13 27.48939
## SFRP1  4.144489 8.377033  12.32914 8.162432e-16 4.897459e-13 25.83515
## SIDT1 -2.605517 6.013343 -12.23990 1.045491e-15 5.227453e-13 25.59416</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(<span class="no">fit2table</span>[,<span class="st">"t"</span>],<span class="no">fit3table</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">fit2table</span>),<span class="st">"t"</span>])</pre></body></html></div>
<pre><code>## [1] "'is.NA' value mismatch: 2974 in current 0 in target"</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/lia978">Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
