<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dimensionality Reduction • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Dimensionality Reduction">
<meta property="og:description" content="BS831">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Dimensionality_Reduction_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="Dimensionality_Reduction_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="Dimensionality_Reduction_files/typedarray-0.1/typedarray.min.js"></script><script src="Dimensionality_Reduction_files/jquery-1.11.3/jquery.min.js"></script><link href="Dimensionality_Reduction_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="Dimensionality_Reduction_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="Dimensionality_Reduction_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="Dimensionality_Reduction_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Dimensionality Reduction</h1>
                        <h4 class="author">Stefano Monti, Anthony Federico</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/Dimensionality_Reduction.Rmd"><code>vignettes/docs/Dimensionality_Reduction.Rmd</code></a></small>
      <div class="hidden name"><code>Dimensionality_Reduction.Rmd</code></div>

    </div>

    
    
<p>We illustrate the use of dimensionality reduction methods to visualize high-dimensional data into low-dimensional (2D or 3D) space, so as to gain some intuition about their structure. A typical purpose of these visualizations is to carry out quick QC, to evaluate wether the data behaves as expected, or to identify (possibly unexpected) outliers or anomalies.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">BS831</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">Biobase</span>)</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">DMsubset</span>)

<span class="no">subset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TESTOSTERONE"</span>,<span class="st">"DOXORUBICIN"</span>,<span class="st">"DIAZEPAM"</span>)
<span class="no">dm3</span> <span class="kw">&lt;-</span> <span class="no">dm10</span>[,<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm10</span>)$<span class="no">CHEMICAL</span> <span class="kw">%in%</span> <span class="no">subset</span>]             <span class="co"># subset data on the 3 chemicals</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>) <span class="co"># eliminate unused levels from factor</span>
<span class="no">col_palette</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>,<span class="st">"dark green"</span>,<span class="st">"blue"</span>)
<span class="no">dm3_col</span> <span class="kw">&lt;-</span> <span class="no">col_palette</span>[<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>]</pre></body></html></div>
<p>In this example, we show three commonly used dimensionality reduction algorithm with our toy dataset.</p>
<div id="principal-component-analysis-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#principal-component-analysis-pca" class="anchor"></a>Principal Component Analysis (PCA)</h2>
<div id="pca-with-svd" class="section level3">
<h3 class="hasAnchor">
<a href="#pca-with-svd" class="anchor"></a>PCA with SVD</h3>
<p>We start with Principle Component Analysis ‘PCA’, using the function <code>prcomp</code>. This will perform PCA using Singular Value Decomposition. We then print the summary statistics, which tell us the variance explained by each component. We then visualize the dataprojected on the first two components, and then projected on the 1st and 3rd component.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">dm3_svd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="kw pkg">Biobase</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(<span class="no">dm3</span>), <span class="kw">scale</span> <span class="kw">=</span> <span class="fl">TRUE</span>)  <span class="co">## perform PCA</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">dm3_svd</span>)$<span class="no">importance</span>[,<span class="fl">1</span>:<span class="fl">4</span>]        <span class="co">## show variance explained by each component</span></pre></body></html></div>
<pre><code>##                             PC1       PC2      PC3       PC4
## Standard deviation     5.870621 0.3358374 0.221574 0.2197043
## Proportion of Variance 0.984690 0.0032200 0.001400 0.0013800
## Cumulative Proportion  0.984690 0.9879100 0.989320 0.9907000</code></pre>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">## let's plot the first two components</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="st">"PC1"</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="st">"PC2"</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,
     <span class="kw">main</span><span class="kw">=</span><span class="st">"PCA"</span>,<span class="kw">xlab</span><span class="kw">=</span><span class="st">"1st PC"</span>, <span class="kw">ylab</span><span class="kw">=</span><span class="st">"2nd PC"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"topleft"</span>,<span class="kw">col</span><span class="kw">=</span><span class="no">col_palette</span>,<span class="kw">legend</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>),<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/PCA-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co">## let's plot the first and third component</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="st">"PC1"</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="st">"PC3"</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,
     <span class="kw">main</span><span class="kw">=</span><span class="st">"PCA"</span>,<span class="kw">xlab</span><span class="kw">=</span><span class="st">"1st PC"</span>, <span class="kw">ylab</span><span class="kw">=</span><span class="st">"3rd PC"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"topright"</span>,<span class="kw">col</span><span class="kw">=</span><span class="no">col_palette</span>,<span class="kw">legend</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>),<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/PCA-2.png" width="700"></p>
<p>Let us use the “fancier” plotting functions in the <code>ggplot</code> package.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">ggplot2</span>)
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="st">"PC1"</span>], <span class="kw">y</span><span class="kw">=</span><span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="st">"PC2"</span>],
                 <span class="kw">id</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>)

<span class="no">p2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">x</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">y</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">id</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"PC1"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"PC2"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"PCA 1 vs. PCA2"</span>)
<span class="no">p2</span></pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/pca.ggplot-1.png" width="700"></p>
<p>It appears that the first two components do a better job at separating the three chemicals than the 1st and 3rd components. We can bring it altogether in an interactive 3D plot.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">plotly</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)

<span class="no">df.pca</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">dm3_svd</span>$<span class="no">rotation</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>:<span class="fl">3</span>)], <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>)) <span class="kw">%&gt;%</span>
          <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
          <span class="kw pkg">magrittr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/magrittr/man/aliases.html">set_colnames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC3"</span>, <span class="st">"id"</span>))

<span class="fu"><a href="https://rdrr.io/pkg/plotly/man/plot_ly.html">plot_ly</a></span>(<span class="no">df.pca</span>,
        <span class="kw">x</span><span class="kw">=</span>~<span class="no">PC1</span>,
        <span class="kw">y</span><span class="kw">=</span>~<span class="no">PC2</span>,
        <span class="kw">z</span><span class="kw">=</span>~<span class="no">PC3</span>,
        <span class="kw">type</span><span class="kw">=</span><span class="st">"scatter3d"</span>,
        <span class="kw">mode</span><span class="kw">=</span><span class="st">"markers"</span>,
        <span class="kw">color</span><span class="kw">=</span>~<span class="no">id</span>,
        <span class="kw">marker</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">size</span><span class="kw">=</span><span class="fl">5</span>))</pre></body></html></div>
<div id="htmlwidget-ba3d2b98e10af6376269" style="width:700px;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-ba3d2b98e10af6376269">{"x":{"visdat":{"a14767be5b70":["function () ","plotlyVisDat"]},"cur_data":"a14767be5b70","attrs":{"a14767be5b70":{"x":{},"y":{},"z":{},"mode":"markers","marker":{"size":5},"color":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d"}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"scene":{"xaxis":{"title":"PC1"},"yaxis":{"title":"PC2"},"zaxis":{"title":"PC3"}},"xaxis":{"type":"category","categoryorder":"array","categoryarray":["-0.167832653009481","-0.168181610717582","-0.16820707312098","-0.16826481879353","-0.168319232967258","-0.168495754437454","-0.168609279488958","-0.168650326804236","-0.168813684958572","-0.168818441440201","-0.168819974664174","-0.168845339608456","-0.168880197035185","-0.169012964968936","-0.169019866779075","-0.169187247556933","-0.16919745868216","-0.169212396514186","-0.169232018461883","-0.169247971264098","-0.169280257246067","-0.16929314646094","-0.169331618992758","-0.169355711552205","-0.169377353368693","-0.169379180589538","-0.16939827747243","-0.169399148151932","-0.169412455496752","-0.169434172386874","-0.169450328544544","-0.169496753973336","-0.169501999342541","-0.169544842626687","-0.169554910569557"]},"yaxis":{"type":"category","categoryorder":"array","categoryarray":["-0.0487981602545538","-0.0770767028905376","-0.0832805929226953","-0.0885793057298395","-0.0893395366686505","-0.0902939950979821","-0.0965039391175194","-0.0995487411652108","-0.106311079550195","-0.109740920062283","-0.112514500687231","-0.11704648621414","-0.117836340547625","-0.129993244726842","-0.135081310270705","-0.143672122600428","-0.160085206495906","-0.160168464948768","-0.163282630985286","-0.166097535426109","-0.1844690297699","0.0141102917017338","0.0386915934452821","0.0435287272895618","0.0476248168924737","0.0592422344454145","0.132351082409941","0.135851336318954","0.159833647354632","0.189006813282695","0.263234953845906","0.288807383026446","0.320243120077975","0.381953244084084","0.418143888847834"]},"hovermode":"closest","showlegend":true},"source":"A","config":{"showSendToCloud":false},"data":[{"x":["-0.168181610717582","-0.168609279488958","-0.168495754437454","-0.168819974664174","-0.16820707312098","-0.168880197035185","-0.167832653009481","-0.168818441440201"],"y":["0.320243120077975","0.288807383026446","0.135851336318954","0.189006813282695","0.381953244084084","0.263234953845906","0.418143888847834","0.132351082409941"],"z":["-0.286512968166151","-0.149715191878358","0.126424865689338","0.0597117048112359","-0.058856667795272","-0.0478889475287211","-0.186710329503986","0.128120807766432"],"mode":"markers","marker":{"color":"rgba(102,194,165,1)","size":5,"line":{"color":"rgba(102,194,165,1)"}},"type":"scatter3d","name":"DIAZEPAM","textfont":{"color":"rgba(102,194,165,1)"},"error_y":{"color":"rgba(102,194,165,1)"},"error_x":{"color":"rgba(102,194,165,1)"},"line":{"color":"rgba(102,194,165,1)"},"frame":null},{"x":["-0.169280257246067","-0.169379180589538","-0.169450328544544","-0.169412455496752","-0.169331618992758","-0.169355711552205","-0.16929314646094","-0.168845339608456","-0.169399148151932","-0.169544842626687","-0.16939827747243","-0.169554910569557","-0.169232018461883","-0.169501999342541","-0.169247971264098","-0.169187247556933","-0.169496753973336","-0.169434172386874","-0.169377353368693","-0.169019866779075","-0.168813684958572"],"y":["-0.112514500687231","-0.11704648621414","-0.0995487411652108","-0.0893395366686505","-0.0885793057298395","-0.0770767028905376","-0.160168464948768","-0.135081310270705","-0.106311079550195","-0.129993244726842","-0.0487981602545538","-0.109740920062283","-0.160085206495906","-0.0965039391175194","-0.117836340547625","-0.0832805929226953","-0.143672122600428","-0.0902939950979821","-0.163282630985286","-0.166097535426109","-0.1844690297699"],"z":["-0.248626988479288","-0.0402735112555943","0.107991014735247","-0.0970409085389985","-0.106777159261717","0.0217531102807362","-0.218795892165819","-0.323669960332648","0.215454729691652","0.0105103303412429","-0.0991530826309389","0.0608319226431818","-0.255255422577347","0.0219325631873207","0.0744530027980156","0.250024973464592","0.0300134276145922","0.11299613588706","-0.106592101599145","-0.0845656810799535","-0.133260709251184"],"mode":"markers","marker":{"color":"rgba(252,141,98,1)","size":5,"line":{"color":"rgba(252,141,98,1)"}},"type":"scatter3d","name":"DOXORUBICIN","textfont":{"color":"rgba(252,141,98,1)"},"error_y":{"color":"rgba(252,141,98,1)"},"error_x":{"color":"rgba(252,141,98,1)"},"line":{"color":"rgba(252,141,98,1)"},"frame":null},{"x":["-0.168650326804236","-0.169012964968936","-0.16919745868216","-0.168319232967258","-0.16826481879353","-0.169212396514186"],"y":["0.0386915934452821","0.0592422344454145","0.0435287272895618","0.0476248168924737","0.159833647354632","0.0141102917017338"],"z":["0.099692175500763","0.329833877371796","0.311658216270117","0.0681842259346062","0.102960771507427","0.308356130073517"],"mode":"markers","marker":{"color":"rgba(141,160,203,1)","size":5,"line":{"color":"rgba(141,160,203,1)"}},"type":"scatter3d","name":"TESTOSTERONE","textfont":{"color":"rgba(141,160,203,1)"},"error_y":{"color":"rgba(141,160,203,1)"},"error_x":{"color":"rgba(141,160,203,1)"},"line":{"color":"rgba(141,160,203,1)"},"frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="pca-with-eigen-decomposition" class="section level3">
<h3 class="hasAnchor">
<a href="#pca-with-eigen-decomposition" class="anchor"></a>PCA with Eigen Decomposition</h3>
<p>Next, we perform Principle Component Analysis ‘PCA’, using the function <code>princomp</code>. This will perform PCA using eigen decomposition. We then visualize the data projected on the first two components, and then projected on the 1st and 3rd component.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">dm3_eg</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/princomp.html">princomp</a></span>(<span class="kw pkg">Biobase</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(<span class="no">dm3</span>))  <span class="co">## perform PCA</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dm3_eg</span>$<span class="no">loadings</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"PC"</span>, <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dm3_eg</span>$<span class="no">loadings</span>)))</pre></body></html></div>
<p>Again with <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">PC1</span><span class="kw">=</span><span class="no">dm3_eg</span>$<span class="no">loadings</span>[,<span class="st">"PC1"</span>], <span class="kw">PC2</span><span class="kw">=</span><span class="no">dm3_eg</span>$<span class="no">loadings</span>[,<span class="st">"PC2"</span>], <span class="kw">PC3</span><span class="kw">=</span><span class="no">dm3_eg</span>$<span class="no">loadings</span>[,<span class="st">"PC3"</span>],
                 <span class="kw">id</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>)

<span class="no">p2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">PC1</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">PC2</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">id</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"PC1"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"PC2"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"PCA 1 vs. PCA2"</span>)
<span class="no">p2</span></pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/pca.ggplot_eg-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">p3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">PC1</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">PC3</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">id</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"PC1"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"PC3"</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"PCA 1 vs. PCA3"</span>)
<span class="no">p3</span></pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/pca.ggplot_eg-2.png" width="700"></p>
<p>Again, it appears that the first two components do a better job at separating the three chemicals than the 1st and 3rd components. Debatable whether separation with is better with <code>prcomp</code> (SVD) than <code>princomp</code> (eigen decomposition).</p>
</div>
<div id="t-distributed-stochastic-neighbor-embedding-t-sne" class="section level3">
<h3 class="hasAnchor">
<a href="#t-distributed-stochastic-neighbor-embedding-t-sne" class="anchor"></a>t-Distributed Stochastic Neighbor Embedding (T-SNE)</h3>
<p>Lastly, we carry out projection and visualization based on the <a href="http://www.jmlr.org/papers/v9/vandermaaten08a.html">t-SNE method</a> (see this nice <a href="http://distill.pub/2016/misread-tsne/%0A">tutorial</a>).</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">tsne</span>)

<span class="co"># Create distance matrix (Euclidean)</span>
<span class="no">distDM3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw pkg">Biobase</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(<span class="no">dm3</span>)))

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## t-SNE with perplexity=10</span>
<span class="no">subset_tsne</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tsne/man/tsne.html">tsne</a></span>(<span class="no">distDM3</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">2</span>,<span class="kw">initial_dims</span><span class="kw">=</span><span class="fl">300</span>,<span class="kw">max_iter</span><span class="kw">=</span><span class="fl">3000</span>,<span class="kw">perplexity</span><span class="kw">=</span><span class="fl">3</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">1</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">2</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"perplexity=3"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/t-SNE-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## t-SNE with perplexity=10</span>
<span class="no">subset_tsne</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tsne/man/tsne.html">tsne</a></span>(<span class="no">distDM3</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">2</span>,<span class="kw">initial_dims</span><span class="kw">=</span><span class="fl">300</span>,<span class="kw">max_iter</span><span class="kw">=</span><span class="fl">3000</span>,<span class="kw">perplexity</span><span class="kw">=</span><span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">1</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">2</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"perplexity=10"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/t-SNE-2.png" width="700"></p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## t-SNE with perplexity=10</span>
<span class="no">subset_tsne</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tsne/man/tsne.html">tsne</a></span>(<span class="no">distDM3</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">2</span>,<span class="kw">initial_dims</span><span class="kw">=</span><span class="fl">300</span>,<span class="kw">max_iter</span><span class="kw">=</span><span class="fl">3000</span>,<span class="kw">perplexity</span><span class="kw">=</span><span class="fl">20</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">1</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">2</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"perplexity=20"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/t-SNE-3.png" width="700"></p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co">## here we show lack of "convergence" if not enough iterations are run</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## t-SNE with perplexity=40</span>
<span class="no">subset_tsne</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tsne/man/tsne.html">tsne</a></span>(<span class="no">distDM3</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">2</span>,<span class="kw">initial_dims</span><span class="kw">=</span><span class="fl">300</span>,<span class="kw">max_iter</span><span class="kw">=</span><span class="fl">2000</span>,<span class="kw">perplexity</span><span class="kw">=</span><span class="fl">40</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">1</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">2</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"perplexity=30, 2K iterations"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/t-SNE-4.png" width="700"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co">## convergence reached after 5000 iterations</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## t-SNE with perplexity=30</span>
<span class="no">subset_tsne</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tsne/man/tsne.html">tsne</a></span>(<span class="no">distDM3</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">2</span>,<span class="kw">initial_dims</span><span class="kw">=</span><span class="fl">300</span>,<span class="kw">max_iter</span><span class="kw">=</span><span class="fl">5000</span>,<span class="kw">perplexity</span><span class="kw">=</span><span class="fl">30</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">1</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">2</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"perplexity=30, 5K iterations"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/t-SNE-5.png" width="700"></p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="co">## here, the perplexity is too large to yield good separation (also </span>
<span class="co">## ..tried w/ 10K iterations, not shown)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## t-SNE with perplexity=40</span>
<span class="no">subset_tsne</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tsne/man/tsne.html">tsne</a></span>(<span class="no">distDM3</span>,<span class="kw">k</span><span class="kw">=</span><span class="fl">2</span>,<span class="kw">initial_dims</span><span class="kw">=</span><span class="fl">300</span>,<span class="kw">max_iter</span><span class="kw">=</span><span class="fl">3000</span>,<span class="kw">perplexity</span><span class="kw">=</span><span class="fl">40</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">1</span>],<span class="kw">y</span><span class="kw">=</span><span class="no">subset_tsne</span>[,<span class="fl">2</span>],<span class="kw">col</span><span class="kw">=</span><span class="no">dm3_col</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"perplexity=40"</span>,<span class="kw">pch</span><span class="kw">=</span><span class="fl">20</span>)</pre></body></html></div>
<p><img src="Dimensionality_Reduction_files/figure-html/t-SNE-6.png" width="700"></p>
</div>
<div id="uniform-manifold-approximation-and-projection-umap" class="section level3">
<h3 class="hasAnchor">
<a href="#uniform-manifold-approximation-and-projection-umap" class="anchor"></a>Uniform Manifold Approximation and Projection (UMAP)</h3>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">umap</span>)

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">umap.defaults</span>)

<span class="no">umap.custom</span> <span class="kw">&lt;-</span> <span class="no">umap.defaults</span>
<span class="no">umap.custom</span>$<span class="no">n_neighbors</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">umap.custom</span>$<span class="no">n_components</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">umap.custom</span>$<span class="no">metric</span> <span class="kw">&lt;-</span> <span class="st">"euclidean"</span>

<span class="no">umap.data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/umap/man/umap.html">umap</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw pkg">Biobase</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(<span class="no">dm3</span>)), <span class="kw">config</span><span class="kw">=</span><span class="no">umap.custom</span>, <span class="kw">method</span><span class="kw">=</span><span class="st">"naive"</span>)
<span class="no">df.umap</span> <span class="kw">&lt;-</span> <span class="no">umap.data</span>$<span class="no">layout</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">df.umap</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"C"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">df.umap</span>)), <span class="kw">sep</span><span class="kw">=</span><span class="st">""</span>)
<span class="no">df.umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">df.umap</span>)
<span class="no">df.umap</span>$<span class="no">id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">phenoData</a></span>(<span class="no">dm3</span>)$<span class="no">CHEMICAL</span>)

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">df.umap</span>)</pre></body></html></div>
<pre><code>##                   C1        C2        C3       id
## 94151.CEL -10.525063 22.885314 -7.579127 DIAZEPAM
## 96366.CEL -10.455134 22.815423 -7.509185 DIAZEPAM
## 97769.CEL  -9.929507 -6.004171 24.847695 DIAZEPAM
## 94334.CEL  -9.760381 -5.835043 24.678580 DIAZEPAM
## 96375.CEL -10.158072 22.518477 -7.212086 DIAZEPAM
## 96881.CEL -10.724380 23.085014 -7.778317 DIAZEPAM</code></pre>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/plot_ly.html">plot_ly</a></span>(<span class="no">df.umap</span>,
        <span class="kw">x</span><span class="kw">=</span>~<span class="no">C1</span>,
        <span class="kw">y</span><span class="kw">=</span>~<span class="no">C2</span>,
        <span class="kw">z</span><span class="kw">=</span>~<span class="no">C3</span>,
        <span class="kw">type</span><span class="kw">=</span><span class="st">"scatter3d"</span>,
        <span class="kw">mode</span><span class="kw">=</span><span class="st">"markers"</span>,
        <span class="kw">color</span><span class="kw">=</span>~<span class="no">id</span>,
        <span class="kw">marker</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">size</span><span class="kw">=</span><span class="fl">5</span>))</pre></body></html></div>
<div id="htmlwidget-a6ec0ba3ad0703a63760" style="width:700px;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-a6ec0ba3ad0703a63760">{"x":{"visdat":{"a14739bceb0e":["function () ","plotlyVisDat"]},"cur_data":"a14739bceb0e","attrs":{"a14739bceb0e":{"x":{},"y":{},"z":{},"mode":"markers","marker":{"size":5},"color":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d"}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"scene":{"xaxis":{"title":"C1"},"yaxis":{"title":"C2"},"zaxis":{"title":"C3"}},"hovermode":"closest","showlegend":true},"source":"A","config":{"showSendToCloud":false},"data":[{"x":[-10.5250629557214,-10.4551341757195,-9.9295065914073,-9.76038083065859,-10.1580724669841,-10.7243796810019,-10.2353874534127,-9.95899063893305],"y":[22.885314141908,22.8154229374154,-6.00417105859984,-5.83504276351477,22.518477198784,23.0850141009352,22.5957882403799,-6.03365300856914],"z":[-7.57912727551035,-7.5091845925392,24.8476954200056,24.6785797996726,-7.21208584855707,-7.77831732923285,-7.28940333157881,24.8771869387163],"mode":"markers","marker":{"color":"rgba(102,194,165,1)","size":5,"line":{"color":"rgba(102,194,165,1)"}},"type":"scatter3d","name":"DIAZEPAM","textfont":{"color":"rgba(102,194,165,1)"},"error_y":{"color":"rgba(102,194,165,1)"},"error_x":{"color":"rgba(102,194,165,1)"},"line":{"color":"rgba(102,194,165,1)"},"frame":null},{"x":[-1.04508555913665,0.228258685297222,0.206113793021445,-0.0652075717851552,0.0385139766913909,-1.41277572076681,0.681611684561605,0.992660548894904,-1.08029959795704,0.225015760817843,-1.26567237145792,-0.541801050123247,0.862511317562356,0.0580630600359422,0.293049025140331,-0.855320803457528,0.241092620040097,-1.03133466910435,0.405448902385253,0.246504369194488,0.354515922340853],"y":[-4.01117473189118,-2.99995509096038,-3.70713832790622,-2.91894828804636,-2.69203145577146,-4.04152440159321,-3.67402126573203,-4.02134069699894,-3.06656852284784,-2.7747467409081,-4.03821270631444,-2.90625913374582,-3.88142916922322,-3.23267354155271,-3.84123871407142,-2.83359080238632,-2.58885902300312,-3.32457270174973,-2.89876469557749,-2.3451957273127,-2.33919801161878],"z":[-1.46305244561384,5.20907082311597,5.79092295941705,5.62908138649543,5.1970553949315,-0.90943763164659,-3.12821890156158,-3.0875847750917,-6.32191419971099,-4.66022408730627,-1.13539976103992,-5.68192510092761,-2.9795081765543,5.72712478831189,5.90654558991132,-6.03494100780597,-4.30958242251702,-6.28188822793144,-4.03406731613455,-4.9061551553585,-4.96646455427147],"mode":"markers","marker":{"color":"rgba(252,141,98,1)","size":5,"line":{"color":"rgba(252,141,98,1)"}},"type":"scatter3d","name":"DOXORUBICIN","textfont":{"color":"rgba(252,141,98,1)"},"error_y":{"color":"rgba(252,141,98,1)"},"error_x":{"color":"rgba(252,141,98,1)"},"line":{"color":"rgba(252,141,98,1)"},"frame":null},{"x":[14.2793865677841,13.7526960118894,14.0022123658466,14.1521201586591,13.7568857682942,14.26775159917],"y":[-4.86071571631349,-4.60258615450033,-4.73781042788328,-4.25920403940344,-4.90724642744595,-4.5221432739809],"z":[-1.91227881026735,-1.73747385569367,-1.88192665700058,-1.43068169081995,-1.99214946709262,-1.64027047881353],"mode":"markers","marker":{"color":"rgba(141,160,203,1)","size":5,"line":{"color":"rgba(141,160,203,1)"}},"type":"scatter3d","name":"TESTOSTERONE","textfont":{"color":"rgba(141,160,203,1)"},"error_y":{"color":"rgba(141,160,203,1)"},"error_x":{"color":"rgba(141,160,203,1)"},"line":{"color":"rgba(141,160,203,1)"},"frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="summary" class="section level3">
<h3 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h3>
<ul>
<li>Use SVD (<code>prcomp</code>) to perform PCA over (<code>princomp</code>).</li>
<li>tSNE is a useful way to visualize data.
<ul>
<li>Drawbacks: Need to set a seed to get reprocuible results.</li>
<li>Need to play with perplexity parameter in order to get good separation.</li>
</ul>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/lia978">Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
