<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Geneset projection based on ASSIGN â€¢ BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Geneset projection based on ASSIGN">
<meta property="og:description" content="BS831">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Geneset projection based on ASSIGN</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/GeneSetProjectionASSIGN.Rmd"><code>vignettes/docs/GeneSetProjectionASSIGN.Rmd</code></a></small>
      <div class="hidden name"><code>GeneSetProjectionASSIGN.Rmd</code></div>

    </div>

    
    
In this module, we illustrate the use of a method of pathway (or geneset, or signature) projection. In particular, we show an example of use of Adaptive Signature Selection and InteGratioN (<a href="http://www.ncbi.nlm.nih.gov/pubmed/25617415">ASSIGN</a>). See <a href="Rmodule_GeneSetProjectionKS.html">Rmodule_GeneSetProjectionGSVA </a> and <a href="Rmodule_KSenrichment.html">Rmodule_KSenrichment</a> for an example of alternative method of pathway projection based on KS enrichment.

<p>We start by loading the ASSIGN package, available through <a href="http://www.bioconductor.org/packages/release/bioc/html/ASSIGN.html">Bioconductor</a>, and by sourcing a wrapper script for its easier handling.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BS831</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">Biobase</span>)      <span class="co"># the package with the ExpressionSet definition, among others</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">ASSIGN</span>)       <span class="co"># the assign package</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">RColorBrewer</span>) <span class="co"># for heatmap visualization</span></pre></body></html></div>
<p>We next load the necessary data, consisting of gene expression profiles from the TCGA head and neck squamous carcinoma (HNSC) dataset, and a signature corresponding to combined shRNA knockdown of TAZ and YAP, two transcriptional effectors of the <a href="http://www.cellsignal.com/contents/science-cst-pathways-stem-cell-markers/hippo-signaling-pathway/pathways-hippo">hippo pathway</a>.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">HNSC_RNASeq_ES</span>)

<span class="co">## load the GEP data</span>
<span class="no">hnsc</span> <span class="kw">&lt;-</span> <span class="no">HNSC_RNASeq_ES</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">hnsc</span>)</pre></body></html></div>
<pre><code>## Features  Samples 
##    20263      566</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">## load the signature</span>
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"BS831"</span>), <span class="st">"yap_taz.gmt"</span>), <span class="kw">header</span><span class="kw">=</span><span class="no">F</span>,<span class="kw">stringsAsFactors</span><span class="kw">=</span><span class="no">F</span>,<span class="kw">row.names</span><span class="kw">=</span><span class="fl">1</span> )
<span class="co">## show the signature (one per row)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sig</span>)</pre></body></html></div>
<pre><code>## [1] "TAZ.YAP.UP" "TAZ.YAP.DN"</code></pre>
<p>We will only project genes that go down upon YAP/TAZ knockdown for this example.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">gene_set</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">sig</span>[<span class="st">"TAZ.YAP.DN"</span>,])
<span class="no">gene_set</span> <span class="kw">&lt;-</span> <span class="no">gene_set</span>[<span class="no">gene_set</span><span class="kw">!=</span><span class="st">""</span>] <span class="co"># Strip off any empty entries</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">gene_set</span>))            <span class="co"># Show the size of the gene signature</span></pre></body></html></div>
<pre><code>## [1] 381</code></pre>
<p>We are now ready to run ASSIGN. We specify the output directory where all ASSIGN output will be stored. This folder will automatically be created if it does not already exist.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">479</span>) <span class="co"># for reproducible results</span>
<span class="no">assign_out_dir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>( <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(), <span class="st">"yap_taz_activated_HNSC"</span> )
<span class="fu">run_assign</span>(<span class="kw">ES</span> <span class="kw">=</span> <span class="no">hnsc</span>,<span class="kw">eSig</span> <span class="kw">=</span> <span class="no">gene_set</span>, <span class="kw">oDir</span> <span class="kw">=</span> <span class="no">assign_out_dir</span>, <span class="kw">iter</span><span class="kw">=</span><span class="fl">3000</span>)</pre></body></html></div>
<pre><code>## Running with signature:                   gene_set 
## Number of genes in signature:             381 
## Overlap between signature and ESet genes: 358 
## Run ASSIGN with mixture modelling beta:   TRUE  ..
## Saving output to directory:               /var/folders/7j/bg701dss38lc2xq26vt_by240000gn/T//RtmpLclhYQ/yap_taz_activated_HNSC</code></pre>
<pre><code>## Done!</code></pre>
<p>Load the generated ASSIGN output object and generate heatmaps to visualize the signature projection</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>( <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">assign_out_dir</span>,<span class="st">"output.rda"</span>))
<span class="fu">plotAll</span>(<span class="kw">eSet</span> <span class="kw">=</span> <span class="no">hnsc</span>,<span class="kw">output.data</span> <span class="kw">=</span><span class="no">output.data</span>,<span class="kw">title</span> <span class="kw">=</span> <span class="st">"YAP/TAZ activation"</span> ,<span class="kw">gene_list</span> <span class="kw">=</span>  <span class="no">gene_set</span>)</pre></body></html></div>
<p><img src="GeneSetProjectionASSIGN_files/figure-html/plot%20projection-1.png" width="576"><img src="GeneSetProjectionASSIGN_files/figure-html/plot%20projection-2.png" width="576"></p>
<p>Geneset projection is particularly useful when we need to determine the <em>activity</em> of important proteins, like transcription factors (or effectors, more in general), which often is not properly captured by their transcript expression. In these cases, looking at the coordinated expression of their downstream targets is often more effective.</p>
<p>Letâ€™s take a look at the gene expression profiles for YAP and TAZ with respect to tumor grade. Below, we show that while the expression of neither transcripts is significantly associated with grade progression, their activity as captured by the ASSIGN scores (i.e., by the coordinated expression of their downstream targets) is significantly associated in the epxected direction, with higher grades associated to higher TAZ/YAP activity.</p>
<p>We first fetch the expression values for the two genes. Note their specific gene symbols (YAP1 and WWTR1)</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co">## YAP1 is the gene symbol for YAP</span>
<span class="no">yap1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(<span class="no">hnsc</span>)[<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(<span class="no">hnsc</span>)$<span class="no">gene_symbol</span> <span class="kw">%in%</span> <span class="st">"YAP1"</span>]))
<span class="co">## WWTR1 is the gene symbol for TAZ</span>
<span class="no">wwtr1</span><span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(<span class="no">hnsc</span>)[<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(<span class="no">hnsc</span>)$<span class="no">gene_symbol</span> <span class="kw">%in%</span> <span class="st">"WWTR1"</span>]))</pre></body></html></div>
<p>We next extract the generated ASSIGN YAP/TAZ projection scores for each sample.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">scores_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">assign_out_dir</span>,<span class="st">"pathway_activity_testset.csv"</span>)
<span class="no">scores</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span>(<span class="no">scores_file</span>,<span class="kw">stringsAsFactors</span><span class="kw">=</span><span class="no">F</span>,<span class="kw">header</span><span class="kw">=</span><span class="no">T</span>)
<span class="no">yap.taz.scores</span> <span class="kw">&lt;-</span> <span class="no">scores</span>[,<span class="fl">2</span>]
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">yap.taz.scores</span>)</pre></body></html></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.0000  0.5850  0.7084  0.6881  0.8169  0.9973</code></pre>
<p>We then make a data frame combinding the yap/taz ASSIGN scores, yap and taz gene expression and tumor grade. We will visualize these to see the effect of geneset projection vs single gene expression analysis</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">d</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="st">"sample_ID"</span><span class="kw">=</span><span class="no">scores</span>[,<span class="fl">1</span>],
                <span class="st">"tumor_grade"</span><span class="kw">=</span><span class="no">hnsc</span>$<span class="no">my_grade</span>,
                <span class="st">"assign_scores"</span><span class="kw">=</span><span class="no">yap.taz.scores</span>,
                <span class="st">"yap1"</span><span class="kw">=</span><span class="no">yap1</span>,
                <span class="st">"wwtr1"</span><span class="kw">=</span><span class="no">wwtr1</span>)

<span class="no">d</span> <span class="kw">&lt;-</span> <span class="no">d</span>[!(<span class="no">d</span>$<span class="no">tumor_grade</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">NA</span>,<span class="st">"gx"</span>)),] <span class="co"># Remove samples with missing grade information (NA or "gx")</span>
<span class="no">d</span>$<span class="no">tumor_grade</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">d</span>$<span class="no">tumor_grade</span>)) <span class="co"># Remove eliminated entries</span></pre></body></html></div>
<p>Plot individual gene expression values, as well as YAP/TAZ ASSIGN scores with respect to tumor grade</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">f</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gray75"</span>,<span class="st">"darkgreen"</span>))
<span class="co">## Make a gradient of colours for different tumor grade</span>
<span class="no">grade.cols</span> <span class="kw">&lt;-</span> <span class="fu">f</span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">d</span>$<span class="no">tumor_grade</span>))))

<span class="co">#par(mfrow=c(3,1)) #Arrange plots in single column</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(<span class="no">yap1</span>~<span class="no">tumor_grade</span>,<span class="kw">col</span><span class="kw">=</span><span class="no">grade.cols</span>,<span class="kw">data</span> <span class="kw">=</span> <span class="no">d</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"YAP gene expression \nwrt tumor grade"</span>)</pre></body></html></div>
<p><img src="GeneSetProjectionASSIGN_files/figure-html/plot%20assign%20scores%20and%20gene%20expression-1.png" width="576"></p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(<span class="no">wwtr1</span>~<span class="no">tumor_grade</span>,<span class="kw">col</span><span class="kw">=</span><span class="no">grade.cols</span>,<span class="kw">data</span> <span class="kw">=</span> <span class="no">d</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"TAZ gene expression \nwrt tumor grade"</span>)</pre></body></html></div>
<p><img src="GeneSetProjectionASSIGN_files/figure-html/plot%20assign%20scores%20and%20gene%20expression-2.png" width="576"></p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(<span class="no">assign_scores</span>~<span class="no">tumor_grade</span>,<span class="kw">col</span><span class="kw">=</span><span class="no">grade.cols</span>,<span class="kw">data</span><span class="kw">=</span><span class="no">d</span>,<span class="kw">main</span><span class="kw">=</span><span class="st">"YAP/TAZ ASSIGN scores \nwrt tumor grade"</span>)</pre></body></html></div>
<p><img src="GeneSetProjectionASSIGN_files/figure-html/plot%20assign%20scores%20and%20gene%20expression-3.png" width="576"></p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="co">## association with TAZ/YAP "activity" highly significant</span>
<span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span>(<span class="no">assign_scores</span>~<span class="no">tumor_grade</span>,<span class="kw">data</span><span class="kw">=</span><span class="no">d</span>))</pre></body></html></div>
<pre><code>## Analysis of Variance Table
## 
## Response: assign_scores
##              Df  Sum Sq Mean Sq F value    Pr(&gt;F)    
## tumor_grade   4  5.5282 1.38205  62.333 &lt; 2.2e-16 ***
## Residuals   539 11.9508 0.02217                      
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/lia978">Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
