<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Hierarchical Clustering • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Hierarchical Clustering">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Hierarchical Clustering</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/hclust.Rmd"><code>vignettes/docs/hclust.Rmd</code></a></small>
      <div class="hidden name"><code>hclust.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(BS831)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(heatmap.plus)</span></code></pre></div>
<div id="clustering-of-real-data-the-drug-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-of-real-data-the-drug-matrix" class="anchor"></a>Clustering of Real Data: the Drug Matrix</h2>
<p>We will use the DrugMatrix subset previously described for our examples. This data corresponds to chemical perturbation experiments, whereby rats are exposed to different chemicals, and their liver’s RNA is profiled.</p>
<p>We upload the data and perform a drastic variation filtering to obtain a small and easy to manage dataset.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(dm10)</span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a>DM &lt;-<span class="st"> </span><span class="kw">variationFilter</span>(dm10,<span class="dt">ngenes=</span><span class="dv">1000</span>,<span class="dt">do.plot=</span><span class="ot">TRUE</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/drugMatrix-1.png" width="700"></p>
</div>
<div id="hierarchical-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#hierarchical-clustering" class="anchor"></a>Hierarchical Clustering</h2>
<p>Let us analyze the data by carrying out hierarchical clustering. We’ll use <code>heatmap.plus</code> to visualize the data. Let us first define a simple function to create a color gradient to be used for coloring the gene expression heatmaps.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>colGradient &lt;-<span class="st"> </span><span class="cf">function</span>( cols, length, <span class="dt">cmax=</span><span class="dv">255</span> )</span>
<span id="cb3-2"><a href="#cb3-2"></a>{</span>
<span id="cb3-3"><a href="#cb3-3"></a>  <span class="co">## e.g., to create a white-to-red gradient with 10 levels</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>  <span class="co">##</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>  <span class="co">##   colGradient(cols=c('white','red'),length=10)</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>  <span class="co">##</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>  <span class="co">## or, to create a blue-to-white-to-red gradients with 9 colors (4 blue's, white, 4 red's)</span></span>
<span id="cb3-8"><a href="#cb3-8"></a>  <span class="co">##</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>  <span class="co">##   colGradient(cols=c('blue','white','red'),length=9)</span></span>
<span id="cb3-10"><a href="#cb3-10"></a>  <span class="co">##</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>  ramp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRamp</a></span>(cols)</span>
<span id="cb3-12"><a href="#cb3-12"></a>  <span class="kw"><a href="https://rdrr.io/r/grDevices/rgb.html">rgb</a></span>( <span class="kw">ramp</span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dt">length=</span>length)), <span class="dt">max=</span>cmax )</span>
<span id="cb3-13"><a href="#cb3-13"></a>}</span></code></pre></div>
<p>Next, let us perform the actual clustering. While the function <code>heatmap.plus</code> can carry out hierarchical clustering internally, we explicitely call <code>hclust</code> outside the function call to illustrate its use (this will also save computation time should one want to call heatmap.plus multiple times with different color coding or other changes that would not affect the clustering).</p>
<p>Notice that the heatmap color is somewhat saturated. In the module <code>Rmodule_heatmap.Rmd</code> we present additional visualization functions that do a better job at color-coding the heatmap.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">## color coding of the samples indicating carcinogenicity status and chemical</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>chemicals &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">pData</span>(dm10)[,<span class="st">"CHEMICAL"</span>])</span>
<span id="cb4-3"><a href="#cb4-3"></a>CSC &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">CARC=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>,<span class="st">"green"</span>,<span class="st">"pink"</span>)[<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">pData</span>(dm10)[,<span class="st">"Carcinogen_liv"</span>],<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="ot">NA</span>,<span class="st">"NON-CARC"</span>,<span class="st">"CARCINOGEN"</span>))],</span>
<span id="cb4-4"><a href="#cb4-4"></a>             <span class="dt">CHEM=</span><span class="kw"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(chemicals))[<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">pData</span>(dm10)[,<span class="st">"CHEMICAL"</span>],chemicals)])</span>
<span id="cb4-5"><a href="#cb4-5"></a></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">## color gradient for the expression levels (blue=down-regulated; white=neutral; red=up-regulated)</span></span>
<span id="cb4-7"><a href="#cb4-7"></a>bwrPalette &lt;-<span class="st"> </span><span class="kw">colGradient</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"blue"</span>,<span class="st">"white"</span>,<span class="st">"red"</span>),<span class="dt">length=</span><span class="dv">13</span>)</span>
<span id="cb4-8"><a href="#cb4-8"></a></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">## cluster rows (genes) and columns (samples)</span></span>
<span id="cb4-10"><a href="#cb4-10"></a>hc.row &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(<span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM)))),<span class="dt">method=</span><span class="st">"ward.D2"</span>) <span class="co"># genes by correlation</span></span>
<span id="cb4-11"><a href="#cb4-11"></a>hc.col &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">method=</span><span class="st">"ward.D2"</span>)           <span class="co"># samples by euclidean distance (default)</span></span>
<span id="cb4-12"><a href="#cb4-12"></a></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co">## draw the heatmap (hide row and col labels since they'd be unreadable)</span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(<span class="kw">exprs</span>(DM),<span class="dt">Rowv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.row),<span class="dt">Colv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.col),<span class="dt">col=</span>bwrPalette,<span class="dt">ColSideColors=</span>CSC,<span class="dt">labCol=</span><span class="ot">NA</span>,<span class="dt">labRow=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/hclust-1.png" width="700"></p>
<div id="optimal-leaf-ordering" class="section level3">
<h3 class="hasAnchor">
<a href="#optimal-leaf-ordering" class="anchor"></a>Optimal Leaf Ordering</h3>
<p>As discussed in class, hierarchical clustering induces a <em>partial</em> ordering of the dendogram leaves (i.e., of the clustered items), modulo the ‘flipping’ of any of the sub-trees. However, one can obtain a <em>total</em> ordering by using the leaf-ordering algorithm developed by <a href="http://bioinformatics.oxfordjournals.org/content/19/9/1070">Bar-Joseph et al. (2001)</a>, which minimizes the distance betwees adjacent items in two distinct sub-trees.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(cba) <span class="co"># load the necessary pacakge</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">## First, call hclust, then call the leaf-ordering algorithm and re-order the dendrogram</span></span>
<span id="cb5-4"><a href="#cb5-4"></a></span>
<span id="cb5-5"><a href="#cb5-5"></a>hc.col &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">method=</span><span class="st">"ward.D2"</span>)</span>
<span id="cb5-6"><a href="#cb5-6"></a>ord &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/cba/man/order.optimal.html">order.optimal</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">merge=</span>hc.col<span class="op">$</span>merge)</span>
<span id="cb5-7"><a href="#cb5-7"></a>hc.col<span class="op">$</span>merge &lt;-<span class="st"> </span>ord<span class="op">$</span>merge</span>
<span id="cb5-8"><a href="#cb5-8"></a>hc.col<span class="op">$</span>order &lt;-<span class="st"> </span>ord<span class="op">$</span>order</span>
<span id="cb5-9"><a href="#cb5-9"></a></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="co">## visualize the results (and compare, visually, to the results w/o optimal leaf ordering above)</span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(<span class="kw">exprs</span>(DM),<span class="dt">Rowv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.row),<span class="dt">Colv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.col),<span class="dt">col=</span>bwrPalette,<span class="dt">ColSideColors=</span>CSC,<span class="dt">labCol=</span><span class="ot">NA</span>,<span class="dt">labRow=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/hcopt-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co">## since this is a step always worth performing, let's define a simple function that implements the necessary steps</span></span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a>hcopt &lt;-<span class="st"> </span><span class="cf">function</span>(d, <span class="dt">HC=</span><span class="ot">NULL</span>, <span class="dt">method =</span> <span class="st">"ward.D"</span>, <span class="dt">members =</span> <span class="ot">NULL</span>)</span>
<span id="cb6-4"><a href="#cb6-4"></a>{</span>
<span id="cb6-5"><a href="#cb6-5"></a>  <span class="cf">if</span> ( <span class="kw"><a href="https://rdrr.io/r/base/NULL.html">is.null</a></span>(HC) ) { </span>
<span id="cb6-6"><a href="#cb6-6"></a>    HC &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(d,<span class="dt">method=</span>method,<span class="dt">members=</span>members)</span>
<span id="cb6-7"><a href="#cb6-7"></a>  }</span>
<span id="cb6-8"><a href="#cb6-8"></a>  ORD &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/cba/man/order.optimal.html">order.optimal</a></span>(d,<span class="dt">merge=</span>HC<span class="op">$</span>merge)</span>
<span id="cb6-9"><a href="#cb6-9"></a>  HC<span class="op">$</span>merge &lt;-<span class="st"> </span>ORD<span class="op">$</span>merge</span>
<span id="cb6-10"><a href="#cb6-10"></a>  HC<span class="op">$</span>order &lt;-<span class="st"> </span>ORD<span class="op">$</span>order</span>
<span id="cb6-11"><a href="#cb6-11"></a>  HC</span>
<span id="cb6-12"><a href="#cb6-12"></a>}</span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">## ok, now we can easily use the function in place of hclust</span></span>
<span id="cb6-14"><a href="#cb6-14"></a></span>
<span id="cb6-15"><a href="#cb6-15"></a>hc.row &lt;-<span class="st"> </span><span class="kw">hcopt</span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(<span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM)))),<span class="dt">method=</span><span class="st">"ward.D2"</span>) <span class="co"># genes by correlation</span></span>
<span id="cb6-16"><a href="#cb6-16"></a>hc.col &lt;-<span class="st"> </span><span class="kw">hcopt</span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">method=</span><span class="st">"ward.D2"</span>)            <span class="co"># samples by euclidean distance (default)</span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(<span class="kw">exprs</span>(DM),<span class="dt">Rowv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.row),<span class="dt">Colv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.col),<span class="dt">col=</span>bwrPalette,<span class="dt">ColSideColors=</span>CSC,<span class="dt">labCol=</span><span class="ot">NA</span>,<span class="dt">labRow=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/hcopt-2.png" width="700"></p>
</div>
<div id="alternative-agglomeration-rules" class="section level3">
<h3 class="hasAnchor">
<a href="#alternative-agglomeration-rules" class="anchor"></a>Alternative Agglomeration Rules</h3>
<p>In the examples above, we used the <a href="https://en.wikipedia.org/wiki/Ward's_method">Ward method</a> as our agglomeration rule. One of the nice properties of Ward is that it tends to yield well-balanced trees. Let us not look at alternative agglomeration rules.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co">## complete linkage</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>hc.row1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(<span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM)))),<span class="dt">method=</span><span class="st">"complete"</span>) <span class="co"># genes by correlation</span></span>
<span id="cb7-3"><a href="#cb7-3"></a>hc.col1 &lt;-<span class="st"> </span><span class="kw">hcopt</span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">method=</span><span class="st">"complete"</span>)            <span class="co"># samples by euclidean distance (default)</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(<span class="kw">exprs</span>(DM),<span class="dt">Rowv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.row1),<span class="dt">Colv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.col1),<span class="dt">col=</span>bwrPalette,<span class="dt">ColSideColors=</span>CSC,<span class="dt">labCol=</span><span class="ot">NA</span>,<span class="dt">labRow=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/linkage-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co">## single linkage</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>hc.row2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(<span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM)))),<span class="dt">method=</span><span class="st">"single"</span>) <span class="co"># genes by correlation</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>hc.col2 &lt;-<span class="st"> </span><span class="kw">hcopt</span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">method=</span><span class="st">"single"</span>)            <span class="co"># samples by euclidean distance (default)</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(<span class="kw">exprs</span>(DM),<span class="dt">Rowv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.row2),<span class="dt">Colv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.col2),<span class="dt">col=</span>bwrPalette,<span class="dt">ColSideColors=</span>CSC,<span class="dt">labCol=</span><span class="ot">NA</span>,<span class="dt">labRow=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/linkage-2.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">## average linkage</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>hc.row3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span>(<span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM)))),<span class="dt">method=</span><span class="st">"average"</span>) <span class="co"># genes by correlation</span></span>
<span id="cb9-3"><a href="#cb9-3"></a>hc.col3 &lt;-<span class="st"> </span><span class="kw">hcopt</span>(<span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">exprs</span>(DM))),<span class="dt">method=</span><span class="st">"average"</span>)            <span class="co"># samples by euclidean distance (default)</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(<span class="kw">exprs</span>(DM),<span class="dt">Rowv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.row3),<span class="dt">Colv=</span><span class="kw"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span>(hc.col3),<span class="dt">col=</span>bwrPalette,<span class="dt">ColSideColors=</span>CSC,<span class="dt">labCol=</span><span class="ot">NA</span>,<span class="dt">labRow=</span><span class="ot">NA</span>)</span></code></pre></div>
<p><img src="hclust_files/figure-html/linkage-3.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#clustering-of-real-data-the-drug-matrix">Clustering of Real Data: the Drug Matrix</a></li>
      <li><a href="#hierarchical-clustering">Hierarchical Clustering</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/lia978">Amy Li</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/anfederico">Anthony Federico</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
