<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>HyperGeometric Enrichment • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="HyperGeometric Enrichment">
<meta property="og:description" content="BS831">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="HyperEnrichment_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>HyperGeometric Enrichment</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/HyperEnrichment.Rmd"><code>vignettes/docs/HyperEnrichment.Rmd</code></a></small>
      <div class="hidden name"><code>HyperEnrichment.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/montilab/BS831">BS831</a></span><span class="op">)</span></code></pre></div>
<p>Here, we show the use of the <a href="https://en.wikipedia.org/wiki/Hypergeometric_distribution"> hyper-geometric distribution</a> to test for enrichment of a (biologically relevant) category (e.g., a pathway) in a differential gene expression signature. We will show the use of the functions <code>phyper</code> and <code>fisher.test</code>. We will then show the definition of a simple script to perform “hyper-enrichment” across multiple categories/pathways.</p>
<p>Recall the set-up from the slides (<code>BS831_class05_ComparativeEnrichment.pptx</code>). <img src="../figs/hyper.enrichment.png" alt="enrichment"></p>
<div id="loading-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-the-data" class="anchor"></a>Loading the Data</h3>
<p>We start by loading the package containing the hyperEnrichment script.</p>
<p>We start by loading the code containing the hyperEnrichment scripts.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">hyper</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">hyperGsets</span><span class="op">)</span></code></pre></div>
<pre><code>## GeneSet Object containing:  1320 gene sets</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># show the size of the first 10 genesets</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu">getGeneSet</span><span class="op">(</span><span class="va">hyperGsets</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,<span class="va">length</span><span class="op">)</span></code></pre></div>
<pre><code>##               KEGG_GLYCOLYSIS_GLUCONEOGENESIS 
##                                            62 
##                  KEGG_CITRATE_CYCLE_TCA_CYCLE 
##                                            32 
##                KEGG_PENTOSE_PHOSPHATE_PATHWAY 
##                                            27 
## KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS 
##                                            28 
##          KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM 
##                                            34 
##                     KEGG_GALACTOSE_METABOLISM 
##                                            26 
##        KEGG_ASCORBATE_AND_ALDARATE_METABOLISM 
##                                            25 
##                    KEGG_FATTY_ACID_METABOLISM 
##                                            42 
##                     KEGG_STEROID_BIOSYNTHESIS 
##                                            17 
##           KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS 
##                                            16</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># show the size of the signatures</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">hyperSig</span>,<span class="va">length</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## control.UP control.DN     TAZ.UP     TAZ.DN     YAP.UP     YAP.DN TAZ+YAP.UP 
##         11         23         41         35        356        470        166 
## TAZ+YAP.DN  DPAGT1.UP  DPAGT1.DN 
##        381         70        199</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># let's rename the signatures</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hyperSig</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"UP"</span>,<span class="st">"REPRESSED"</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"DN"</span>,<span class="st">"ACTIVATED"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hyperSig</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The genesets (the categories) represent a subset of the genesets contained in the MSgigDB’s c2.cp compendium. The signatures represent the up- and down-regulated genes in oral cancer cell lines where one of several regulators (TAZ, YAP, TAZ/YAP, DPAGT1) was knocked down (KD).</p>
<p>For example, the “YAP” signature was obtained by performing differential analysis on a 6-sample datasets corresponding to “knockdown <em>vs.</em> control” (3 <em>vs.</em> 3) experiments performed on the HSC3 oral cancer cell line. The genes significantly down-regulated (up-regulated) upon knockdown represent the genes assumed to be <em>activated</em> (<em>repressed</em>) by TAZ.</p>
</div>
<div id="enrichment-test-by-phyper" class="section level3">
<h3 class="hasAnchor">
<a href="#enrichment-test-by-phyper" class="anchor"></a>Enrichment test by <code>phyper</code>
</h3>
<p>We first test for enrichment of a single geneset in a single signature using the function <code>phyper</code> (the cumulative function of the hyper-geometric distribution). The background population is set to 23,467, which represents the number of annotated genes in the dataset used to derive the differential signature.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## background population</span>
<span class="va">backpop</span> <span class="op">&lt;-</span> <span class="fl">23467</span>

<span class="co">## let us extract the signature of YAP-activated genes</span>
<span class="va">signature</span> <span class="op">&lt;-</span> <span class="va">hyperSig</span><span class="op">$</span><span class="va">YAP.ACTIVATED</span>

<span class="co">## and a REACTOME pathway</span>
<span class="va">pathway</span> <span class="op">&lt;-</span> <span class="fu">getGeneSet</span><span class="op">(</span><span class="va">hyperGsets</span><span class="op">)</span><span class="op">[[</span><span class="st">"REACTOME_TRANSCRIPTION"</span><span class="op">]</span><span class="op">]</span>

<span class="co">## let's check the overlap (if any)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span> <span class="va">overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span><span class="op">(</span><span class="va">signature</span>,<span class="va">pathway</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></code></pre></div>
<pre><code>## [1] 17</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## let us take a look at the phyper arguments (call ?phyper for more details)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/args.html">args</a></span><span class="op">(</span><span class="va">phyper</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## function (q, m, n, k, lower.tail = TRUE, log.p = FALSE) 
## NULL</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## let us apply the function to our data (see figure)</span>
<span class="fu"><a href="https://rdrr.io/r/stats/Hypergeometric.html">phyper</a></span><span class="op">(</span> q<span class="op">=</span><span class="va">overlap</span><span class="op">-</span><span class="fl">1</span>,               <span class="co"># number of red marbles in the draw - 1 (see below)</span>
        m<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">pathway</span><span class="op">)</span>,         <span class="co"># number of red marbles in urn</span>
        n<span class="op">=</span><span class="va">backpop</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">pathway</span><span class="op">)</span>, <span class="co"># number of green marbles in urn </span>
        k<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span>,       <span class="co"># Number of drawn marbles</span>
        lower.tail<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>          <span class="co"># compute P( X &gt; overlap ), hence the '-1' above</span></code></pre></div>
<pre><code>## [1] 1.262929e-06</code></pre>
<p>Clearly, the category (pathway) “REACTOME_TRANSCRIPTOME” is highly significantly enriched in the signature of YAP-activated genes.</p>
</div>
<div id="enrichment-test-by-fisher-test" class="section level3">
<h3 class="hasAnchor">
<a href="#enrichment-test-by-fisher-test" class="anchor"></a>Enrichment test by <code>fisher.test</code>
</h3>
<p>We then show how we can obtain equivalent results using the <code>fisher.test</code> function. We first need to properly fill-in a contingency table, and then apply the function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## we need to define the contingency table</span>
<span class="co">##</span>
<span class="co">##       | DRAWN   !DRAWN | TOT</span>
<span class="co">## ------+----------------+----</span>
<span class="co">## GREEN |  k-q   n-m-k+q | n-m</span>
<span class="co">##   RED |   q      m-q   |  m</span>
<span class="co">## ------+----------------+----</span>
<span class="co">##   TOT |   k      n-k   |  n</span>

<span class="va">contable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
    dg<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">-</span><span class="va">overlap</span>,
    dr<span class="op">=</span><span class="va">overlap</span>,
    ng<span class="op">=</span><span class="va">backpop</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">pathway</span><span class="op">)</span><span class="op">+</span><span class="va">overlap</span>,
    nr<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">pathway</span><span class="op">)</span><span class="op">-</span><span class="va">overlap</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">2</span>,dimnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GREEN"</span>,<span class="st">"RED"</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DRAWN"</span>,<span class="st">"not DRAWN"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">contable</span><span class="op">)</span></code></pre></div>
<pre><code>##       DRAWN not DRAWN
## GREEN   453     22804
## RED      17       193</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test</a></span><span class="op">(</span><span class="va">contable</span>,alt<span class="op">=</span><span class="st">"less"</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  Fisher's Exact Test for Count Data
## 
## data:  contable
## p-value = 1.263e-06
## alternative hypothesis: true odds ratio is less than 1
## 95 percent confidence interval:
##  0.0000000 0.3632734
## sample estimates:
## odds ratio 
##  0.2255642</code></pre>
<p>As you can see, the p-value is the same (p=1.263e-06).</p>
</div>
<div id="defining-and-applying-a-hyper-enrichment-function" class="section level3">
<h3 class="hasAnchor">
<a href="#defining-and-applying-a-hyper-enrichment-function" class="anchor"></a>Defining and applying a “Hyper-Enrichment” function</h3>
<p>We here define a (relatively) simple function, <code>hyperEnrichment</code>, to run hyper-geometric-based enrichment tests on multiple signatures and mutiple genesets.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## defined in CBMRtools</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">hyperEnrichment</span><span class="op">)</span></code></pre></div>
<pre><code>## function
## (
##    drawn,          # one or more sets of 'drawn' items (e.g., genes). Basically, a list of signatures.
##    categories,     # gene sets (list of gene sets)
##    ntotal=length(unique(unlist(categories))),
##                    # background population, i.e., the total no. of items from which
##                    # ..items are supposed to have been drawn
##    min.drawsize=4, # min no. of drawn items that must be among categories' items
##    mht=TRUE,       # correct for MHT across multiple 'draws'
##    verbose=TRUE
## )
## {
##    ## checks on inputs
##    ##
##    if (!is(categories, "list") ) {
##      stop( "categories expected to be a list of gene sets" )
##    }
##    gene.names&lt;-unique(unlist(categories))
##    if ( is.list(drawn) &amp;&amp; is.null(names(drawn)) ) {
##      stop( "drawn must have non-null names when a list" )
##    }
##    if ( ntotal&lt;length(unique(unlist(categories)))) {
##      warning( "background population's size less than unique categories' items: ", ntotal,"&lt;",length(gene.names))
##    }
##    ##
##    ## end checks
##    
##    cnames &lt;-
##       c("pval","fdr","set annotated","set size","category annotated","total annotated","category","hits")
##    
##    ## handling of multiple 'draws'
##    ##
##    if ( is.list(drawn) ) 
##    {
##      ncat &lt;- length(categories)
##      enrich &lt;- matrix(NA,ncat*length(drawn),length(cnames)+1)
##      
##      VERBOSE(verbose,"Testing",length(drawn),"drawsets on",ncat,"categories and",
##              length(gene.names),"total items ..\n")
##      
##      percent &lt;- 0.1
##      base &lt;- 0
##      ntst &lt;- 0
##      for ( i in 1:length(drawn) )
##      {
##        VERBOSE(verbose,"*** Testing", names(drawn)[i], ".. " )
##        dset &lt;- drawn[[i]]
##        tmp &lt;- hyperEnrichment(dset,categories,ntotal=ntotal,verbose=verbose)
##        if (is.null(tmp)) {
##          VERBOSE(verbose,"not enough items drawn\n")
##          next
##        }
##        ntst &lt;- ntst+1
##        rng &lt;- (base+1):(base+ncat)
##        if (any(!is.na(enrich[rng,]))) stop( "something wrong")
##        
##        enrich[rng,] &lt;- cbind(set=rep(names(drawn)[i],ncat),tmp)
##        base &lt;- base+ncat
##        if (F &amp;&amp; i&gt;=round(length(drawn)*percent)) {
##          VERBOSE(verbose, round(100*percent),"% ",sep="")
##          percent &lt;- percent+0.1
##        }
##        VERBOSE(verbose," (min fdr: ", signif(min(as.numeric(tmp[,"fdr"])),2),")\n",sep="")
##      }
##      VERBOSE(verbose,"done.\n")
##      colnames(enrich) &lt;- c("set",cnames)
##       
##      enrich &lt;- enrich[1:base,,drop=F]
##      if (mht) {
##        VERBOSE(verbose,"MHT-correction across multiple draws ..")
##        enrich[,"fdr"] &lt;- pval2fdr(as.numeric(enrich[,"pval"]))
##        VERBOSE(verbose,"done.\n")
##      }    
##      VERBOSE(verbose,
##              "Categories tested: ",rjust(length(categories),4),"\n",
##              "Candidate sets:    ",rjust(length(drawn),4),"\n",
##              "Sets tested:       ",rjust(ntst,4),"\n",
##              "Items tested:      ",rjust(sum(sapply(drawn,length)),4)," (min,med,max: ",
##              paste(quantile(sapply(drawn,length),probs=c(0,.5,1)),collapse=","),")\n",
##              "N(FDR&lt;=0.25):      ",rjust(sum(enrich[,"fdr"]&lt;=.25),4),"\n",
##              "N(FDR&lt;=0.05):      ",rjust(sum(enrich[,"fdr"]&lt;=.05),4),"\n",
##              "N(FDR&lt;=0.01):      ",rjust(sum(enrich[,"fdr"]&lt;=.01),4),"\n",
##              sep="")
##      return(enrich)
##    }
##    ## handling of a single draw
##    ##
##    m.idx&lt;-drawn[drawn %in% gene.names] 
##    
##    if ( length(m.idx)&lt;min.drawsize ) {
##      VERBOSE(verbose,"insufficient annotated genes in the drawn set: ",
##              paste(gene.names[m.idx],collapse=","),"\n")
##      return(NULL)
##    }
##    VERBOSE(verbose,length(m.idx),"/",length(drawn), " annotated genes found",sep="")
##    
##    nhits &lt;-sapply(categories, function(x,y) length(intersect(x,y)), m.idx)
##    ndrawn &lt;- length(drawn) # length(m.idx)
##    ncats &lt;- sapply(categories,length)
##    nleft &lt;- ntotal-ncats
##    
##    ## compute P[X&gt;=nhits]
##    enrich &lt;- phyper(q=nhits-1,m=ncats,n=nleft,k=ndrawn,lower.tail=F)
##    enrich &lt;- cbind(pval=enrich,
##                    fdr=pval2fdr(enrich),
##                    nhits=nhits,
##                    ndrawn=ndrawn,
##                    ncats=ncats,
##                    ntot=ntotal,
##                    category=names(categories))
##    
##    enrich &lt;- cbind(enrich,
##                    hits=sapply(categories,function(x,y)paste(intersect(x,y),collapse=','),m.idx))
##    ord &lt;- order(as.numeric(enrich[,"pval"]))
##    enrich &lt;- enrich[ord,,drop=F]
##    enrich[,"pval"] &lt;- signif(as.numeric(enrich[,"pval"]),2)
##    enrich[,"fdr"] &lt;- signif(as.numeric(enrich[,"fdr"]),2)
##    
##    colnames(enrich) &lt;- cnames
##    rownames(enrich) &lt;- names(categories)[ord]
##    
##    return(enrich)
## }
## &lt;environment: namespace:BS831&gt;</code></pre>
<p>We then apply it to the list of oral cancer knockdown signatures and two MSigDB geneset compendia, the set of canonical pathways <a href="http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=CP"> c2.cp</a> and a set of hallmark genesets <a href="http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=H"> h.all</a> [<a href="http://www.ncbi.nlm.nih.gov/pubmed/26771021">Liberzon et al., 2015</a>].</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html">openxlsx</a></span><span class="op">)</span> 
<span class="va">hyperE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/hyperEnrichment.html">hyperEnrichment</a></span><span class="op">(</span>drawn<span class="op">=</span><span class="va">hyperSig</span>,categories<span class="op">=</span><span class="fu">getGeneSet</span><span class="op">(</span><span class="va">hyperGsets</span><span class="op">)</span>,ntotal<span class="op">=</span><span class="va">backpop</span><span class="op">)</span> </code></pre></div>
<pre><code>## Testing 10 drawsets on 1320 categories and 8428 total items ..
## *** Testing control.REPRESSED .. 4/11 annotated genes found (min fdr: 1)
## *** Testing control.ACTIVATED .. 8/23 annotated genes found (min fdr: 0.057)
## *** Testing TAZ.REPRESSED .. 15/41 annotated genes found (min fdr: 1)
## *** Testing TAZ.ACTIVATED .. 17/35 annotated genes found (min fdr: 1)
## *** Testing YAP.REPRESSED .. 164/356 annotated genes found (min fdr: 1.5e-15)
## *** Testing YAP.ACTIVATED .. 261/470 annotated genes found (min fdr: 4.4e-86)
## *** Testing TAZ+YAP.REPRESSED .. 60/166 annotated genes found (min fdr: 0.24)
## *** Testing TAZ+YAP.ACTIVATED .. 186/381 annotated genes found (min fdr: 2.4e-26)
## *** Testing DPAGT1.REPRESSED .. 15/70 annotated genes found (min fdr: 1)
## *** Testing DPAGT1.ACTIVATED .. 67/199 annotated genes found (min fdr: 0.32)
## done.
## MHT-correction across multiple draws ..done.
## Categories tested: 1320
## Candidate sets:      10
## Sets tested:         10
## Items tested:      1752 (min,med,max: 11,118,470)
## N(FDR&lt;=0.25):       130
## N(FDR&lt;=0.05):        67
## N(FDR&lt;=0.01):        41</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hyperE</span><span class="op">)</span></code></pre></div>
<pre><code>##      set                 pval     fdr                 set annotated set size
## [1,] "control.REPRESSED" "0.0051" "0.278181818181818" "1"           "11"    
## [2,] "control.REPRESSED" "0.0056" "0.29568"           "1"           "11"    
## [3,] "control.REPRESSED" "0.0098" "0.45993031358885"  "1"           "11"    
## [4,] "control.REPRESSED" "0.012"  "0.52624584717608"  "1"           "11"    
## [5,] "control.REPRESSED" "0.012"  "0.52624584717608"  "1"           "11"    
## [6,] "control.REPRESSED" "0.014"  "0.586666666666667" "1"           "11"    
##      category annotated total annotated
## [1,] "11"               "23467"        
## [2,] "12"               "23467"        
## [3,] "21"               "23467"        
## [4,] "25"               "23467"        
## [5,] "25"               "23467"        
## [6,] "29"               "23467"        
##      category                                                                  
## [1,] "PID_ALK2PATHWAY"                                                         
## [2,] "REACTOME_CS_DS_DEGRADATION"                                              
## [3,] "REACTOME_CHONDROITIN_SULFATE_BIOSYNTHESIS"                               
## [4,] "PID_TOLL_ENDOGENOUS_PATHWAY"                                             
## [5,] "REACTOME_A_TETRASACCHARIDE_LINKER_SEQUENCE_IS_REQUIRED_FOR_GAG_SYNTHESIS"
## [6,] "PID_INTEGRIN2_PATHWAY"                                                   
##      hits   
## [1,] "BMP7" 
## [2,] "BGN"  
## [3,] "BGN"  
## [4,] "BGN"  
## [5,] "BGN"  
## [6,] "SPON2"</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hyperE.fdr25</span> <span class="op">&lt;-</span> <span class="va">hyperE</span><span class="op">[</span><span class="va">hyperE</span><span class="op">[</span>,<span class="st">'fdr'</span><span class="op">]</span><span class="op">&lt;=</span><span class="fl">0.25</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hyperE.fdr25</span><span class="op">)</span></code></pre></div>
<pre><code>##      set                 pval      fdr                   set annotated set size
## [1,] "control.ACTIVATED" "4.3e-05" "0.00423582089552239" "3"           "23"    
## [2,] "control.ACTIVATED" "0.00037" "0.0299631901840491"  "2"           "23"    
## [3,] "control.ACTIVATED" "0.00097" "0.0692108108108108"  "2"           "23"    
## [4,] "control.ACTIVATED" "0.0033"  "0.195336322869955"   "2"           "23"    
## [5,] "control.ACTIVATED" "0.0034"  "0.199466666666667"   "2"           "23"    
## [6,] "TAZ.REPRESSED"     "0.0012"  "0.0812307692307692"  "2"           "41"    
##      category annotated total annotated
## [1,] "70"               "23467"        
## [2,] "29"               "23467"        
## [3,] "47"               "23467"        
## [4,] "88"               "23467"        
## [5,] "89"               "23467"        
## [6,] "29"               "23467"        
##      category                                           hits             
## [1,] "PID_AP1_PATHWAY"                                  "FOSL1,ETS1,CSF2"
## [2,] "PID_TCRCALCIUMPATHWAY"                            "FOSL1,CSF2"     
## [3,] "PID_NFAT_TFPATHWAY"                               "FOSL1,CSF2"     
## [4,] "KEGG_HEMATOPOIETIC_CELL_LINEAGE"                  "IL1R2,CSF2"     
## [5,] "WNT_SIGNALING"                                    "FOSL1,WNT5A"    
## [6,] "REACTOME_DEGRADATION_OF_THE_EXTRACELLULAR_MATRIX" "MMP1,MMP13"</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## let us save it as a '.xlsx' object for you to inspect</span>
<span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/write.xlsx.html">write.xlsx</a></span><span class="op">(</span><span class="va">hyperE.fdr25</span>, file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"BS831"</span><span class="op">)</span>, <span class="st">"hyperE.fdr25.xls"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Let’s load a different geneset compendium (hallmark genesets)</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">HALL</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">"GeneSet"</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"BS831"</span><span class="op">)</span>, <span class="st">"h.all.v6.1.symbols.gmt"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># show the size of the first 10 genesets</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu">getGeneSet</span><span class="op">(</span><span class="va">HALL</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,<span class="va">length</span><span class="op">)</span></code></pre></div>
<pre><code>##    HALLMARK_TNFA_SIGNALING_VIA_NFKB                    HALLMARK_HYPOXIA 
##                                 200                                 200 
##    HALLMARK_CHOLESTEROL_HOMEOSTASIS            HALLMARK_MITOTIC_SPINDLE 
##                                  74                                 200 
## HALLMARK_WNT_BETA_CATENIN_SIGNALING         HALLMARK_TGF_BETA_SIGNALING 
##                                  42                                  54 
##    HALLMARK_IL6_JAK_STAT3_SIGNALING                 HALLMARK_DNA_REPAIR 
##                                  87                                 150 
##             HALLMARK_G2M_CHECKPOINT                  HALLMARK_APOPTOSIS 
##                                 200                                 161</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hyperHALL</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/hyperEnrichment.html">hyperEnrichment</a></span><span class="op">(</span>drawn<span class="op">=</span><span class="va">hyperSig</span>,categories<span class="op">=</span><span class="fu">getGeneSet</span><span class="op">(</span><span class="va">HALL</span><span class="op">)</span>,ntotal<span class="op">=</span><span class="va">backpop</span><span class="op">)</span></code></pre></div>
<pre><code>## Testing 10 drawsets on 50 categories and 4386 total items ..
## *** Testing control.REPRESSED .. insufficient annotated genes in the drawn set:  NA 
## not enough items drawn
## *** Testing control.ACTIVATED .. 12/23 annotated genes found (min fdr: 0.048)
## *** Testing TAZ.REPRESSED .. 14/41 annotated genes found (min fdr: 0.25)
## *** Testing TAZ.ACTIVATED .. 15/35 annotated genes found (min fdr: 0.032)
## *** Testing YAP.REPRESSED .. 135/356 annotated genes found (min fdr: 2.5e-37)
## *** Testing YAP.ACTIVATED .. 199/470 annotated genes found (min fdr: 7.4e-77)
## *** Testing TAZ+YAP.REPRESSED .. 45/166 annotated genes found (min fdr: 0.0023)
## *** Testing TAZ+YAP.ACTIVATED .. 151/381 annotated genes found (min fdr: 1.4e-26)
## *** Testing DPAGT1.REPRESSED .. 16/70 annotated genes found (min fdr: 0.076)
## *** Testing DPAGT1.ACTIVATED .. 51/199 annotated genes found (min fdr: 0.0028)
## done.
## MHT-correction across multiple draws ..done.
## Categories tested:   50
## Candidate sets:      10
## Sets tested:          9
## Items tested:      1752 (min,med,max: 11,118,470)
## N(FDR&lt;=0.25):        80
## N(FDR&lt;=0.05):        34
## N(FDR&lt;=0.01):        15</code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hyperHALL.fdr25</span> <span class="op">&lt;-</span> <span class="va">hyperHALL</span><span class="op">[</span><span class="va">hyperHALL</span><span class="op">[</span>,<span class="st">'fdr'</span><span class="op">]</span><span class="op">&lt;=</span><span class="fl">0.25</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hyperHALL.fdr25</span><span class="op">)</span></code></pre></div>
<pre><code>##      set                 pval      fdr                  set annotated set size
## [1,] "control.ACTIVATED" "0.00095" "0.01425"            "3"           "23"    
## [2,] "control.ACTIVATED" "0.0033"  "0.0353571428571429" "2"           "23"    
## [3,] "control.ACTIVATED" "0.0054"  "0.0514285714285714" "2"           "23"    
## [4,] "control.ACTIVATED" "0.016"   "0.10625"            "2"           "23"    
## [5,] "control.ACTIVATED" "0.016"   "0.10625"            "2"           "23"    
## [6,] "control.ACTIVATED" "0.016"   "0.10625"            "2"           "23"    
##      category annotated total annotated category                            
## [1,] "200"              "23467"         "HALLMARK_TNFA_SIGNALING_VIA_NFKB"  
## [2,] "87"               "23467"         "HALLMARK_IL6_JAK_STAT3_SIGNALING"  
## [3,] "113"              "23467"         "HALLMARK_UNFOLDED_PROTEIN_RESPONSE"
## [4,] "200"              "23467"         "HALLMARK_HYPOXIA"                  
## [5,] "200"              "23467"         "HALLMARK_MYOGENESIS"               
## [6,] "200"              "23467"         "HALLMARK_IL2_STAT5_SIGNALING"      
##      hits             
## [1,] "FOSL1,BTG3,CSF2"
## [2,] "IL1R2,CSF2"     
## [3,] "CHAC1,STC2"     
## [4,] "STC2,ETS1"      
## [5,] "CHRNB1,STC2"    
## [6,] "CSF2,IL1R2"</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/lia978">Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
