<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>KS-based Enrichment • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="KS-based Enrichment">
<meta property="og:description" content="BS831">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="KS_Enrichment_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>KS-based Enrichment</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/KS_Enrichment.Rmd"><code>vignettes/docs/KS_Enrichment.Rmd</code></a></small>
      <div class="hidden name"><code>KS_Enrichment.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/montilab/BS831">BS831</a></span><span class="op">)</span></code></pre></div>
<div id="introduction" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h3>
<p>Here, we show some simple examples of use of the function <code>ks.test</code>, and of a simple function wrapper, <code>ksGenescore</code>, tailored for gene set enrichment analysis, which adds plotting capabilities and returns a signed score.</p>
<p>The <code>ks.test</code> function implements a Kolmogorov-Smirnov (KS) non-parametric test for the comparison of two distributions. That is, given two random samples, <em>x</em> and <em>y</em>, from unknown distributions, <code>ks.test</code> tests whether the two generating distributions are the same or not.</p>
</div>
<div id="comparing-two-distributions" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-two-distributions" class="anchor"></a>Comparing two distributions</h3>
<p>Here are few examples of use of the function. Comparing a random sample (from a Gaussian) to an actual distribution (uniform):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>                   <span class="co"># for reproducible results</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,mean<span class="op">=</span><span class="fl">0</span>,sd<span class="op">=</span><span class="fl">1</span><span class="op">)</span>     <span class="co"># sample 100 random observations from a standard Gaussian</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">"punif"</span>,min<span class="op">=</span><span class="op">-</span><span class="fl">3</span>,max<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="co"># test for difference from a uniform distribution U[-3;3]</span></code></pre></div>
<pre><code>## 
##  One-sample Kolmogorov-Smirnov test
## 
## data:  x
## D = 0.2491, p-value = 8.153e-06
## alternative hypothesis: two-sided</code></pre>
<p>The function returns a score, the D statistic (between 0 and 1, with larger values indicating larger differences), and a p-value.</p>
<p>Next, we compare a random sample (from a Gaussian) to another random sample (from a Uniform):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>                   <span class="co"># for reproducible results</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">100</span>,mean<span class="op">=</span><span class="fl">0</span>,sd<span class="op">=</span><span class="fl">1</span><span class="op">)</span>     <span class="co"># sample 100 random observations from a standard Gaussian</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">100</span>,min<span class="op">=</span><span class="op">-</span><span class="fl">3</span>,max<span class="op">=</span><span class="fl">3</span><span class="op">)</span>    <span class="co"># sample 100 random observations from a Uniform in [-3,3]</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>                    <span class="co"># test for difference between the two samples</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and y
## D = 0.29, p-value = 0.0004453
## alternative hypothesis: two-sided</code></pre>
</div>
<div id="using-ks-test-for-geneset-enrichment" class="section level3">
<h3 class="hasAnchor">
<a href="#using-ks-test-for-geneset-enrichment" class="anchor"></a>Using <code>ks.test</code> for geneset enrichment</h3>
<p>We want to use the <code>ks.test</code> to determine the skewdness of the ranks of a given set of genes. The general problem is as follows: we have a gene expression dataset (say a 1000-gene x 100-sample dataset), and an associated phenotype of interest (say, 50 tumors vs. 50 normals). We want to test whether a given geneset (say, the geneset containing the genes in the <a href="http://cgap.nci.nih.gov/Pathways/BioCarta/h_arenrf2Pathway"> Oxidative Stress</a> pathway) is <em>associated</em> with the phenotype. By “associated”, we mean that the genes in the geneset are coordinatedly up-regulated in one of the phenotype’s classes.</p>
<p>This task can be addressed with the <code>ks.test</code> function:</p>
<ol style="list-style-type: lower-roman">
<li>by first sorting in descending order the (1,000) genes in the dataset according to a differential score (e.g., a t-statistic of ‘tumor vs. normal’);</li>
<li>by then finding the ranks of the genes in the geneset within the (1K-long) sorted list;</li>
<li>by finally testing for the skewdness of the genes’ ranks.</li>
</ol>
<p>If the geneset (Oxidative Stress) were <em>not</em> enriched with respect to the phenotype of interest (tumor vs. normal), we would expect the ranks of the geneset’s genes to be uniformly distributed between 1 and 1,000. If they are skewed to the left (beginning of the sorted list, large positive t-score), then we would conclude that they are enriched in the ‘tumor’ class. If they are skewed to the right (end of the sorted list, large negative t-score), we would conclude that they are enriched in the ‘normal’ class.</p>
<p>Let us assume that our geneset contains 10 genes, and let us assume we have already established their ranks with respect to a given phenotype, and let us denote with <em>x</em> the vector of their ranks. Here is how we can use <code>ks.test</code> to assess “skewedness”:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>     <span class="co"># the genes in the genesets have rank 1 to 10 (i.e., they have the 10 largest t-scores)</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">"punif"</span>,min<span class="op">=</span><span class="fl">1</span>,max<span class="op">=</span><span class="fl">1000</span><span class="op">)</span> <span class="co"># test for their skewdness against a U[1;1000]</span></code></pre></div>
<pre><code>## 
##  One-sample Kolmogorov-Smirnov test
## 
## data:  x
## D = 0.99099, p-value &lt; 2.2e-16
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span>                 <span class="co"># test them against an actual vector of 1000</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:1000
## D = 0.99, p-value = 7.452e-09
## alternative hypothesis: two-sided</code></pre>
<p>In the first call to <code>ks.test</code> above, we used the one-sample test against a ‘punif’ distribution. In the 2nd call, we used the two-sample test with the second sample represented by the sorted list 1-to-1000. The latter formulation is more appropriate to take into account the ‘sample size’, i.e., the lenght of the sorted gene list (in these examples, our dataset has 1000 genes (more realistically, if using a whole-transcriptome profiling platform, such as Affymetrix or RNAseq, the length of the list will be closer to 20,000).</p>
<p>The sample size affects the extent of the statistical significance. Notice how the p-value becomes progressively smaller as the sample size increases:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>          <span class="co"># the genes to be tested have rank 1 to 10 (10 largest t-scores)</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span>   <span class="co"># test them against a sample size of 100</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:100
## D = 0.9, p-value = 8.036e-07
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span>  <span class="co"># test them against a sample size of 1000</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:1000
## D = 0.99, p-value = 7.452e-09
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10000</span><span class="op">)</span> <span class="co"># test them against a sample size of 10000</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:10000
## D = 0.999, p-value = 4.377e-09
## alternative hypothesis: two-sided</code></pre>
<p>Henceforth, we will only use the two-sample option. Below, we test a set of 10 genes ranked toward the end of the sorted list, toward the middle, and a set actually randomly distributed.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">991</span><span class="op">:</span><span class="fl">1000</span> <span class="co"># the genes are last in the ranking (10 most negative t-scores)</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:1000
## D = 0.99, p-value = 7.452e-09
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">496</span><span class="op">:</span><span class="fl">505</span>  <span class="co"># the genes are smack in the middle of the ranking</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">0000</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:0
## D = 1, p-value = 0.0303
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>,min<span class="op">=</span><span class="fl">1</span>,max<span class="op">=</span><span class="fl">1000</span><span class="op">)</span> <span class="co"># the genes' ranks are actually uniformly distributed</span>
<span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">x</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  x and 1:1000
## D = 0.209, p-value = 0.7801
## alternative hypothesis: two-sided</code></pre>
<p>Notice that the KS score <em>D</em> is always positive, so whether the gene ranks are skewed to the left or the right, the returned <em>D</em> statistic is the same. Also, the function doesn’t give us insights into how the distribution of ranks actually looks like. For this, we defined a simple wrapper.</p>
</div>
<div id="the-ksgenescore-script" class="section level3">
<h3 class="hasAnchor">
<a href="#the-ksgenescore-script" class="anchor"></a>The <code>ksGenescore</code> script</h3>
<p>We defined a simple script (in <code>BS831/R/ksGenescore.R</code>) that builds upon <code>ks.test</code> but is specifically tailored to testing for the skewdness of the distribution of a set of ‘genes’ toward the beginning or the end of a sorted list. The script also plots the enrichment score, and returns a signed statistic indicating whether the set of genes is skewed to the left (&gt;0) or to the right (&lt;0). Since we are always testing one geneset against a given sample size, the input arguments are modified, with the first argument denoting the sample size (i.e., the length of the sorted list) and the second argument denoting the vector of genes’ ranks.</p>
<p>Few examples of use of <code>ks.genescore</code> follow. For visualization purposes, we use a reduced sample size of 100 (to better see the tickmarks corresponding to the genes’ ranks).</p>
<p>Testing whether 10 genes fall randomly in the 1:100 range.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># skewed to the left</span>
<span class="fu">ksGenescore</span><span class="op">(</span><span class="fl">100</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,do.plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>    </code></pre></div>
<p><img src="KS_Enrichment_files/figure-html/ksGenescore-1.png" width="700"></p>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  1:n.x and y
## D = 0.9, p-value = 8.036e-07
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># skewed to the right</span>
<span class="fu">ksGenescore</span><span class="op">(</span><span class="fl">100</span>,<span class="fl">91</span><span class="op">:</span><span class="fl">100</span>,do.plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="KS_Enrichment_files/figure-html/ksGenescore-2.png" width="700"></p>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  1:n.x and y
## D = -0.9, p-value = 8.036e-07
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># skewed to the center</span>
<span class="fu">ksGenescore</span><span class="op">(</span><span class="fl">100</span>,<span class="fl">46</span><span class="op">:</span><span class="fl">55</span>,do.plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="KS_Enrichment_files/figure-html/ksGenescore-3.png" width="700"></p>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  1:n.x and y
## D = -0.45, p-value = 0.05035
## alternative hypothesis: two-sided</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># randomly distributed</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">6719</span><span class="op">)</span>; <span class="fu">ksGenescore</span><span class="op">(</span><span class="fl">100</span>,<span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10</span>,min<span class="op">=</span><span class="fl">1</span>,max<span class="op">=</span><span class="fl">100</span><span class="op">)</span>,do.plot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="KS_Enrichment_files/figure-html/ksGenescore-4.png" width="700"></p>
<pre><code>## 
##  Two-sample Kolmogorov-Smirnov test
## 
## data:  1:n.x and y
## D = 0.16497, p-value = 0.9274
## alternative hypothesis: two-sided</code></pre>
<p>Notice that the KS test, when used against a uniform distribution, simply tests deviation of an observed sample from a uniform distribution. Thus, in the examples above, the “geneset” <code>46:55</code> (of ten genes ranked in the middle and adjacent to each other) is also unlikely to come from a uniform distribution, hence the significant p-value (although weakly so).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/anfederico">Anthony Federico</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/lia978">Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
