<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multiple-Hypothesis Correction • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Multiple-Hypothesis Correction">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Multiple-Hypothesis Correction</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/MHTcorrection.Rmd"><code>vignettes/docs/MHTcorrection.Rmd</code></a></small>
      <div class="hidden name"><code>MHTcorrection.Rmd</code></div>

    </div>

    
    
<p>In this module, we show how testing for multiple hypotheses (genes) can increase the chance of false positives, especially for small sample sizes.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(BS831)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(heatmap.plus)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(tidyr)</span></code></pre></div>
<div id="multiple-hypothesis-correction-by-fdr-and-fwer" class="section level1">
<h1 class="hasAnchor">
<a href="#multiple-hypothesis-correction-by-fdr-and-fwer" class="anchor"></a>Multiple-Hypothesis Correction by FDR and FWER</h1>
<p>In the remainder of this module we will explore the effects of multiple hypothesis corrections on family-wise error rate, false positive rate, false negative rate, and power when using <a href="https://en.wikipedia.org/wiki/Bonferroni_correction">Bonferroni</a> and <a href="https://www.jstor.org/stable/2346101">Benjamini Hochberg</a> corrections.</p>
<div id="define-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#define-functions" class="anchor"></a>Define functions</h2>
<p>The following function is just a wrapper for runnning two-sample t-tests.</p>
<div id="wrapper-of-students-t-test" class="section level3">
<h3 class="hasAnchor">
<a href="#wrapper-of-students-t-test" class="anchor"></a>Wrapper of Student’s T-Test</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">## wrapper function extracting t-statistic and p-value from a call to function t.test</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>tscore &lt;-<span class="st"> </span><span class="cf">function</span>(y,x) { </span>
<span id="cb2-3"><a href="#cb2-3"></a>  tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span>(y<span class="op">~</span>x)</span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">score=</span>tmp<span class="op">$</span>statistic,<span class="dt">pval=</span>tmp<span class="op">$</span>p.value)</span>
<span id="cb2-5"><a href="#cb2-5"></a>}</span></code></pre></div>
<p>The next function is a wrapper to define a data matrix, X, and outcome vector, Y, for different conditions. By default we can generate a complete set of data from the null hypthesis to check the FWER. We can also generate data with a subset of features which have different means between groups to check, false positive rate, false negative rate, and power by changing <code>propDiff</code> and <code>Diff</code>.</p>
</div>
<div id="wrapper-of-data-generation-and-differential-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#wrapper-of-data-generation-and-differential-analysis" class="anchor"></a>Wrapper of data generation and differential analysis</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>wrapDA &lt;-<span class="st"> </span><span class="cf">function</span>(Nsamps, <span class="co"># Number of samples per pseudogroup</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>                   <span class="dt">Nrow =</span> <span class="dv">200</span>, <span class="co"># Number of parameters</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>                   <span class="dt">mean =</span> <span class="dv">0</span>, <span class="co"># Mean of normal</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>                   <span class="dt">sd =</span> <span class="fl">0.5</span>, <span class="co"># SD of normal</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>                   <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="co"># Signficance threshold</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>                   <span class="dt">propDiff =</span> <span class="dv">0</span>, <span class="co"># Proportion to make Ha</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>                   <span class="dt">Diff =</span> <span class="dv">0</span>, <span class="co"># Difference for Ha features</span></span>
<span id="cb3-8"><a href="#cb3-8"></a>                   <span class="dt">reportAll =</span> <span class="ot">FALSE</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>                   )</span>
<span id="cb3-10"><a href="#cb3-10"></a>  {</span>
<span id="cb3-11"><a href="#cb3-11"></a>  </span>
<span id="cb3-12"><a href="#cb3-12"></a>  <span class="co"># Genreate a data of random data from null distribution</span></span>
<span id="cb3-13"><a href="#cb3-13"></a>  DAT &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>( <span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>( Nrow <span class="op">*</span><span class="st"> </span>Nsamps <span class="op">*</span><span class="st"> </span><span class="dv">2</span>, <span class="dt">mean=</span><span class="dv">0</span>, <span class="dt">sd=</span><span class="fl">0.5</span>), <span class="dt">nrow=</span>Nrow, <span class="dt">ncol=</span>( Nsamps <span class="op">*</span><span class="dv">2</span>) )</span>
<span id="cb3-14"><a href="#cb3-14"></a>  </span>
<span id="cb3-15"><a href="#cb3-15"></a>  <span class="co"># Add difference to one group</span></span>
<span id="cb3-16"><a href="#cb3-16"></a>  <span class="cf">if</span>(propDiff <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){</span>
<span id="cb3-17"><a href="#cb3-17"></a>    nChange &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">floor</a></span>(propDiff<span class="op">*</span>Nrow)</span>
<span id="cb3-18"><a href="#cb3-18"></a>    DAT[<span class="dv">1</span><span class="op">:</span>(nChange), <span class="dv">1</span><span class="op">:</span>Nsamps] &lt;-<span class="st"> </span>DAT[<span class="dv">1</span><span class="op">:</span>(nChange), <span class="dv">1</span><span class="op">:</span>Nsamps] <span class="op">+</span><span class="st"> </span>Diff</span>
<span id="cb3-19"><a href="#cb3-19"></a>    <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(DAT) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"Ha"</span>, nChange), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"H0"</span>, (Nrow <span class="op">-</span><span class="st"> </span>nChange))), <span class="dv">1</span><span class="op">:</span>Nrow)</span>
<span id="cb3-20"><a href="#cb3-20"></a>  }</span>
<span id="cb3-21"><a href="#cb3-21"></a>  </span>
<span id="cb3-22"><a href="#cb3-22"></a>  <span class="co">## generate a (head/tail) phenotype of proper size</span></span>
<span id="cb3-23"><a href="#cb3-23"></a>  pheno &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'head'</span>,<span class="st">'tail'</span>),<span class="dt">each=</span>Nsamps))</span>
<span id="cb3-24"><a href="#cb3-24"></a></span>
<span id="cb3-25"><a href="#cb3-25"></a>  <span class="co">## perform t.test on each data row with respect to the random phenotype</span></span>
<span id="cb3-26"><a href="#cb3-26"></a>  DIF &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(DAT,<span class="dv">1</span>,tscore,<span class="dt">x=</span>pheno)))</span>
<span id="cb3-27"><a href="#cb3-27"></a>  </span>
<span id="cb3-28"><a href="#cb3-28"></a>  <span class="co">## Sort by p-value</span></span>
<span id="cb3-29"><a href="#cb3-29"></a>  DIF &lt;-<span class="st"> </span>DIF[<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(DIF<span class="op">$</span>pval),]</span>
<span id="cb3-30"><a href="#cb3-30"></a>  </span>
<span id="cb3-31"><a href="#cb3-31"></a>  <span class="co">## Perform Bonferroni and BH corrections</span></span>
<span id="cb3-32"><a href="#cb3-32"></a>  DIF<span class="op">$</span>BFp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span>(DIF<span class="op">$</span>pval, <span class="dt">method =</span> <span class="st">"bonferroni"</span>)</span>
<span id="cb3-33"><a href="#cb3-33"></a>  DIF<span class="op">$</span>BHp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span>(DIF<span class="op">$</span>pval, <span class="dt">method =</span> <span class="st">"BH"</span>)</span>
<span id="cb3-34"><a href="#cb3-34"></a>  </span>
<span id="cb3-35"><a href="#cb3-35"></a>  <span class="co">## Get BH Significant Rows</span></span>
<span id="cb3-36"><a href="#cb3-36"></a>  DIFsig &lt;-<span class="st"> </span>DIF[DIF<span class="op">$</span>BHp <span class="op">&lt;</span><span class="st"> </span>alpha,]</span>
<span id="cb3-37"><a href="#cb3-37"></a>  </span>
<span id="cb3-38"><a href="#cb3-38"></a>  <span class="co">## Return All results or only features with small p-values</span></span>
<span id="cb3-39"><a href="#cb3-39"></a>  <span class="cf">if</span>(reportAll){</span>
<span id="cb3-40"><a href="#cb3-40"></a>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(DIF)</span>
<span id="cb3-41"><a href="#cb3-41"></a>  } <span class="cf">else</span> {</span>
<span id="cb3-42"><a href="#cb3-42"></a>    <span class="co">## If any genes BH significant return these rows else just return row with smallest p-value</span></span>
<span id="cb3-43"><a href="#cb3-43"></a>    <span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(DIFsig)<span class="op">&gt;</span><span class="dv">0</span>) <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(DIFsig) <span class="cf">else</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(DIF[<span class="dv">1</span>,])</span>
<span id="cb3-44"><a href="#cb3-44"></a>  }</span>
<span id="cb3-45"><a href="#cb3-45"></a>}  </span></code></pre></div>
</div>
</div>
<div id="run-differential-analysis-from-null-distributions-for-different-sample-sizes" class="section level2">
<h2 class="hasAnchor">
<a href="#run-differential-analysis-from-null-distributions-for-different-sample-sizes" class="anchor"></a>Run differential analysis from null distributions for different sample sizes</h2>
<p>First, we will check the FWER between Bonferroni and BH corrections for two different sample sizes: 10 samples per group and 50 samples per group. Each feature will have mean=0 and SD=0.5. For each run, the result will either be the feature with the smallest p-value, or a data.frame of features that passed BH significance threshold.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>sGroups &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">10</span>, <span class="dv">25</span>, <span class="dv">50</span>) <span class="co"># Vector of group sizes</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># Run simulation for each sample size 100 times</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>outList &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(sGroups, <span class="cf">function</span>(Nsamps){</span>
<span id="cb4-5"><a href="#cb4-5"></a>    <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">replicate</a></span>(<span class="dv">100</span>, <span class="kw">wrapDA</span>(Nsamps), <span class="dt">simplify =</span> <span class="ot">FALSE</span>)</span>
<span id="cb4-6"><a href="#cb4-6"></a>})</span>
<span id="cb4-7"><a href="#cb4-7"></a></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co"># Name list by sample size</span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(outList) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"Samps"</span>, sGroups, <span class="dt">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb4-10"><a href="#cb4-10"></a></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co"># What does the output look like?</span></span>
<span id="cb4-12"><a href="#cb4-12"></a>outList<span class="op">$</span>Samps_<span class="dv">10</span>[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>##       score.t         pval        BFp        BHp
## 185 -4.477631 0.0002993474 0.05986947 0.05986947</code></pre>
<div id="get-family-wise-error-rates-from-bonferroni-alpha-0-05" class="section level3">
<h3 class="hasAnchor">
<a href="#get-family-wise-error-rates-from-bonferroni-alpha-0-05" class="anchor"></a>Get family wise error rates from Bonferroni (alpha = 0.05)</h3>
<div id="function-to-calculate-fwer-from-list-of-experiment-results" class="section level4">
<h4 class="hasAnchor">
<a href="#function-to-calculate-fwer-from-list-of-experiment-results" class="anchor"></a>Function to calculate FWER from list of experiment results</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>getFWER &lt;-<span class="st"> </span><span class="cf">function</span>(resList, <span class="co"># List of lowest p-values</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>                    <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="co"># P-value threshold</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>                    <span class="dt">pCol =</span> <span class="st">"BFp"</span> <span class="co"># Column with MHT P-value of interest</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>                    )</span>
<span id="cb6-5"><a href="#cb6-5"></a>{</span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="co">## Which experiments had at least one positive hit?</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>  FWERvec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(resList, <span class="cf">function</span>(y) <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(y[,<span class="st">"BFp"</span>] <span class="op">&lt;</span><span class="st"> </span>alpha)<span class="op">&gt;</span><span class="dv">0</span>))</span>
<span id="cb6-8"><a href="#cb6-8"></a>  </span>
<span id="cb6-9"><a href="#cb6-9"></a>  <span class="co">## What proportion of experiments had at least one positive hit?</span></span>
<span id="cb6-10"><a href="#cb6-10"></a>  FWERval &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(FWERvec)</span>
<span id="cb6-11"><a href="#cb6-11"></a>  </span>
<span id="cb6-12"><a href="#cb6-12"></a>  <span class="co">## Return FWER for this set of experiments</span></span>
<span id="cb6-13"><a href="#cb6-13"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(FWERval)</span>
<span id="cb6-14"><a href="#cb6-14"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co">## Get FWER for bonferonni corrected p-values</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>BonVec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(outList, getFWER, <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="dt">pCol =</span> <span class="st">"BFp"</span>)) <span class="co"># Note pCol</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(BonVec)</span></code></pre></div>
<pre><code>## Samps_10 Samps_25 Samps_50 
##     0.06     0.08     0.07</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">## Get FWER for Benjamini-Hochberg corrected p-values</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>BhVec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(outList, getFWER, <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="dt">pCol =</span> <span class="st">"BHp"</span>)) <span class="co"># Note pCol</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(BhVec)</span></code></pre></div>
<pre><code>## Samps_10 Samps_25 Samps_50 
##     0.06     0.08     0.07</code></pre>
<p>Note, for repeated experiments with multiple hypotheses the FWER returned from Bonferroni and Benjamini-Hochberg is the same</p>
</div>
</div>
</div>
<div id="run-differential-analysis-from-alternative-distriubtions-for-different-sample-sizes" class="section level2">
<h2 class="hasAnchor">
<a href="#run-differential-analysis-from-alternative-distriubtions-for-different-sample-sizes" class="anchor"></a>Run differential analysis from alternative distriubtions for different sample sizes</h2>
<p>Next, we will create data where 25% of the genes differ between groups by 0.5. In the output, true positive will have row names starting with “Ha”.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>haList &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(sGroups, <span class="cf">function</span>(Nsamps){</span>
<span id="cb11-2"><a href="#cb11-2"></a>    <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">replicate</a></span>(<span class="dv">100</span>, <span class="kw">wrapDA</span>(Nsamps, <span class="dt">propDiff =</span> <span class="fl">0.25</span>, <span class="dt">Diff =</span> <span class="fl">0.5</span>), <span class="dt">simplify =</span> <span class="ot">FALSE</span>)</span>
<span id="cb11-3"><a href="#cb11-3"></a>})</span>
<span id="cb11-4"><a href="#cb11-4"></a></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co">## Name list by sample size</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(haList) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"Samps"</span>, sGroups, <span class="dt">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb11-7"><a href="#cb11-7"></a></span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="co">## What does the output look like?</span></span>
<span id="cb11-9"><a href="#cb11-9"></a>haList<span class="op">$</span>Samps_<span class="dv">10</span>[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>##       score.t         pval         BFp         BHp
## Ha27 6.102593 9.367465e-06 0.001873493 0.001873493
## Ha30 5.355881 5.922261e-05 0.011844522 0.005922261
## Ha29 4.693650 2.068936e-04 0.041378726 0.011369955
## Ha15 4.876862 2.273991e-04 0.045479818 0.011369955
## Ha45 4.171281 5.739726e-04 0.114794515 0.020704197
## Ha22 4.142938 6.211259e-04 0.124225182 0.020704197
## Ha2  4.047819 7.614228e-04 0.152284551 0.021754936
## Ha21 3.951397 1.026725e-03 0.205345045 0.025668131
## Ha9  3.955238 1.262681e-03 0.252536233 0.028059581
## Ha10 3.668688 1.758112e-03 0.351622443 0.035162244</code></pre>
<p>The rows with the prefix, <em>Ha</em>, come from the distribution for which the alternative hypthesis is true.</p>
<div id="get-false-positive-and-false-negative-rates" class="section level3">
<h3 class="hasAnchor">
<a href="#get-false-positive-and-false-negative-rates" class="anchor"></a>Get False Positive and False Negative Rates</h3>
<div id="function-to-calculate-the-false-positive-rate-and-false-negative-rates-from-list-of-experiment-results" class="section level4">
<h4 class="hasAnchor">
<a href="#function-to-calculate-the-false-positive-rate-and-false-negative-rates-from-list-of-experiment-results" class="anchor"></a>Function to calculate the false positive rate and false negative rates from list of experiment results</h4>
<p>The a false positive rate will be calculated for each experiment individually</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>getFPR_FNR &lt;-<span class="st"> </span><span class="cf">function</span>(resList, <span class="co"># List of lowest p-values</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>                       <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="co"># P-value threshold</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>                       <span class="dt">pCol =</span> <span class="st">"BFp"</span>, <span class="co"># Column with MHT P-value of interest</span></span>
<span id="cb13-4"><a href="#cb13-4"></a>                       <span class="dt">nMH =</span> <span class="dv">200</span>, <span class="co"># Number of hypothesis tested</span></span>
<span id="cb13-5"><a href="#cb13-5"></a>                       <span class="dt">nHa =</span> <span class="dv">50</span> <span class="co"># Number of true positives</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>                       )</span>
<span id="cb13-7"><a href="#cb13-7"></a>{  </span>
<span id="cb13-8"><a href="#cb13-8"></a>  <span class="co">## Extract instances of null hypotheses with small p-values from each experiment</span></span>
<span id="cb13-9"><a href="#cb13-9"></a>  H0list &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(resList, <span class="cf">function</span>(y) y[<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"Ha"</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(y)),])</span>
<span id="cb13-10"><a href="#cb13-10"></a>  </span>
<span id="cb13-11"><a href="#cb13-11"></a>  <span class="co"># Extract vector of false postive rates from each experiment</span></span>
<span id="cb13-12"><a href="#cb13-12"></a>  FPvec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(H0list, <span class="cf">function</span>(y) <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(y[,pCol]<span class="op">&lt;</span>alpha)<span class="op">/</span>(nMH<span class="op">-</span>nHa)))</span>
<span id="cb13-13"><a href="#cb13-13"></a>  </span>
<span id="cb13-14"><a href="#cb13-14"></a>  <span class="co"># Extract instances of alternative hypotheses with small p-values from each experiment</span></span>
<span id="cb13-15"><a href="#cb13-15"></a>  Halist &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(resList, <span class="cf">function</span>(y) y[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"Ha"</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(y)),])</span>
<span id="cb13-16"><a href="#cb13-16"></a>  </span>
<span id="cb13-17"><a href="#cb13-17"></a>  <span class="co"># Extract vector of false negative rates from each experiment</span></span>
<span id="cb13-18"><a href="#cb13-18"></a>  FNvec &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(Halist, <span class="cf">function</span>(y) <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(y[,pCol]<span class="op">&lt;</span>alpha)<span class="op">/</span>(nHa)))</span>
<span id="cb13-19"><a href="#cb13-19"></a>  </span>
<span id="cb13-20"><a href="#cb13-20"></a>  <span class="co"># Create data.frame of results</span></span>
<span id="cb13-21"><a href="#cb13-21"></a>  FP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">FalsePosi =</span> FPvec, <span class="dt">FalseNeg =</span> FNvec)</span>
<span id="cb13-22"><a href="#cb13-22"></a>  </span>
<span id="cb13-23"><a href="#cb13-23"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(FP_FN)</span>
<span id="cb13-24"><a href="#cb13-24"></a>}</span></code></pre></div>
</div>
</div>
<div id="bonferroni-corrected-p-values" class="section level3">
<h3 class="hasAnchor">
<a href="#bonferroni-corrected-p-values" class="anchor"></a>Bonferroni Corrected P-values</h3>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co">## Get FP and FN</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>BonFP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(haList, getFPR_FNR, <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="dt">pCol =</span> <span class="st">"BFp"</span>, <span class="dt">nMH =</span> <span class="dv">200</span>, <span class="dt">nHa =</span> <span class="dv">50</span>)</span>
<span id="cb14-3"><a href="#cb14-3"></a></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">## Concatenate the results for each sample size</span></span>
<span id="cb14-5"><a href="#cb14-5"></a>BonFP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(rbind, BonFP_FN)</span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(BonFP_FN)</span></code></pre></div>
<pre><code>##            FalsePosi FalseNeg
## Samps_10.1         0     0.92
## Samps_10.2         0     0.94
## Samps_10.3         0     0.98
## Samps_10.4         0     1.00
## Samps_10.5         0     0.98
## Samps_10.6         0     0.96</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co">## Add column for sample size identifier</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>BonFP_FN<span class="op">$</span>SampleSize &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"[.][[:digit:]]*"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(BonFP_FN))</span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(BonFP_FN)</span></code></pre></div>
<pre><code>##            FalsePosi FalseNeg SampleSize
## Samps_10.1         0     0.92   Samps_10
## Samps_10.2         0     0.94   Samps_10
## Samps_10.3         0     0.98   Samps_10
## Samps_10.4         0     1.00   Samps_10
## Samps_10.5         0     0.98   Samps_10
## Samps_10.6         0     0.96   Samps_10</code></pre>
</div>
<div id="benjamini-hochberg-corrected-p-values" class="section level3">
<h3 class="hasAnchor">
<a href="#benjamini-hochberg-corrected-p-values" class="anchor"></a>Benjamini Hochberg Corrected P-values</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co">## Get FP and FN</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>BhFP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(haList, getFPR_FNR, <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="dt">pCol =</span> <span class="st">"BHp"</span>, <span class="dt">nMH =</span> <span class="dv">200</span>, <span class="dt">nHa =</span> <span class="dv">50</span>)</span>
<span id="cb18-3"><a href="#cb18-3"></a></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co">## Concatenate the results for each sample size</span></span>
<span id="cb18-5"><a href="#cb18-5"></a>BhFP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(rbind, BhFP_FN)</span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(BhFP_FN)</span></code></pre></div>
<pre><code>##              FalsePosi FalseNeg
## Samps_10.1 0.000000000     0.80
## Samps_10.2 0.006666667     0.64
## Samps_10.3 0.000000000     0.82
## Samps_10.4 0.000000000     1.00
## Samps_10.5 0.000000000     0.86
## Samps_10.6 0.013333333     0.84</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="co">## Add column for sample size identifier</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>BhFP_FN<span class="op">$</span>SampleSize &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"[.][[:digit:]]*"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(BhFP_FN))</span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(BhFP_FN)</span></code></pre></div>
<pre><code>##              FalsePosi FalseNeg SampleSize
## Samps_10.1 0.000000000     0.80   Samps_10
## Samps_10.2 0.006666667     0.64   Samps_10
## Samps_10.3 0.000000000     0.82   Samps_10
## Samps_10.4 0.000000000     1.00   Samps_10
## Samps_10.5 0.000000000     0.86   Samps_10
## Samps_10.6 0.013333333     0.84   Samps_10</code></pre>
</div>
<div id="plot-results" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-results" class="anchor"></a>Plot Results</h3>
<p>We will create boxplots of the results with <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co">## Concatenate results between Bonferroni and Benjamini-Hochberg</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>FP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(BonFP_FN, BhFP_FN)</span>
<span id="cb22-3"><a href="#cb22-3"></a>FP_FN<span class="op">$</span>correction &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Bonferroni"</span>, <span class="st">"BH"</span>), <span class="dt">each =</span> <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(BonFP_FN)) <span class="co"># Add correction type</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(FP_FN)</span></code></pre></div>
<pre><code>##            FalsePosi FalseNeg SampleSize correction
## Samps_10.1         0     0.92   Samps_10 Bonferroni
## Samps_10.2         0     0.94   Samps_10 Bonferroni
## Samps_10.3         0     0.98   Samps_10 Bonferroni
## Samps_10.4         0     1.00   Samps_10 Bonferroni
## Samps_10.5         0     0.98   Samps_10 Bonferroni
## Samps_10.6         0     0.96   Samps_10 Bonferroni</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co">## Create one column of False Positive and False Negative Rates</span></span>
<span id="cb24-2"><a href="#cb24-2"></a>FP_FN &lt;-<span class="st"> </span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span>(FP_FN, <span class="dt">key =</span> <span class="st">"statistic"</span>, <span class="dt">value =</span> <span class="st">"measure"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)</span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(FP_FN)</span></code></pre></div>
<pre><code>##   SampleSize correction statistic measure
## 1   Samps_10 Bonferroni FalsePosi       0
## 2   Samps_10 Bonferroni FalsePosi       0
## 3   Samps_10 Bonferroni FalsePosi       0
## 4   Samps_10 Bonferroni FalsePosi       0
## 5   Samps_10 Bonferroni FalsePosi       0
## 6   Samps_10 Bonferroni FalsePosi       0</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co"># Create plot of results</span></span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(FP_FN, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> SampleSize, </span>
<span id="cb26-3"><a href="#cb26-3"></a>                        <span class="dt">y =</span> measure, </span>
<span id="cb26-4"><a href="#cb26-4"></a>                        <span class="dt">fill =</span> correction, </span>
<span id="cb26-5"><a href="#cb26-5"></a>                        <span class="dt">colour =</span> correction)) <span class="op">+</span></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() <span class="op">+</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>statistic, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">scales =</span> <span class="st">"free_y"</span>)</span></code></pre></div>
<p><img src="MHTcorrection_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>What do you notice about the nature of the performance of the FWER-based Bonferroni correction, and FDR-based Benjamini-Hochberg correction.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#multiple-hypothesis-correction-by-fdr-and-fwer">Multiple-Hypothesis Correction by FDR and FWER</a><ul class="nav nav-pills nav-stacked">
<li><a href="#define-functions">Define functions</a></li>
      <li><a href="#run-differential-analysis-from-null-distributions-for-different-sample-sizes">Run differential analysis from null distributions for different sample sizes</a></li>
      <li><a href="#run-differential-analysis-from-alternative-distriubtions-for-different-sample-sizes">Run differential analysis from alternative distriubtions for different sample sizes</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/lia978">Amy Li</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/anfederico">Anthony Federico</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
