<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data Acquisition and ExpressionSet Basics • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Data Acquisition and ExpressionSet Basics">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Data Acquisition and ExpressionSet Basics</h1>
                        <h4 class="author">Anthony Federico, Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/Microarray_Processing.Rmd"><code>vignettes/docs/Microarray_Processing.Rmd</code></a></small>
      <div class="hidden name"><code>Microarray_Processing.Rmd</code></div>

    </div>

    
    
<p>In this module, we illustrate how to download gene expression datasets from <a href="http://www.ncbi.nlm.nih.gov/geo">GEO</a>.</p>
<p>To this end, we will download and pre-process the gene expression data from the LOAD (Late Onset Alzheimer Disease) study in <a href="http://www.cell.com/fulltext/S0092-8674(13)00387-5"> [Zhang et al., Cell 2013] </a>. The dataset consists of 690 samples corresponding to three brain regions (cerebellum, prefrontal cortex, ) from 230 subjects, including 121 LOAD patients and 109 controls.</p>
<p>We start by loading some required packages and functions.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(BS831)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(GEOquery)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(Biobase)</span>
<span id="cb1-4"><a href="#cb1-4"></a>overwrite &lt;-<span class="st"> </span><span class="ot">FALSE</span></span></code></pre></div>
<div id="download-data-from-geo" class="section level1">
<h1 class="hasAnchor">
<a href="#download-data-from-geo" class="anchor"></a>Download data from GEO</h1>
<p>We start by downloading the data from GEO. We will then show two approaches to adding gene annotation to the expression matrix, one based on the use of already available annotation available on GEO, and another one based on “de-novo” annotation based on the <code>biomaRt</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">## Download expression data from GEO and store in temporary directory</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>tmp_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>()</span>
<span id="cb2-3"><a href="#cb2-3"></a>LOAD &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GEOquery/man/getGEO.html">getGEO</a></span>(<span class="dt">GEO=</span><span class="st">"GSE44772"</span>, <span class="dt">GSEMatrix=</span><span class="ot">TRUE</span>, <span class="dt">destdir=</span>tmp_dir)</span>
<span id="cb2-4"><a href="#cb2-4"></a>LOAD &lt;-<span class="st"> </span>LOAD[[<span class="dv">1</span>]] <span class="co"># getGEO returns a list, so we extract the first element</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(LOAD)       <span class="co"># display summary info for the object</span></span></code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 39280 features, 690 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM1090267 GSM1090268 ... GSM1090960 (690 total)
##   varLabels: title geo_accession ... tissue:ch2 (66 total)
##   varMetadata: labelDescription
## featureData
##   featureNames: 10019475365 10019481149 ... 10033669049 (39280 total)
##   fvarLabels: ID EntrezGeneID ... RosettaGeneModelID (7 total)
##   fvarMetadata: Column Description labelDescription
## experimentData: use 'experimentData(object)'
##   pubMedIds: 23622250 
## Annotation: GPL4372</code></pre>
<div id="use-pre-existing-gene-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#use-pre-existing-gene-annotation" class="anchor"></a>Use pre-existing Gene Annotation</h2>
<p>We first show how to use already available gene annotation information. To this end, we manually downloaded to the data/ subfolder the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL4372">GPL4372.annot</a> file from geo, then ran the following command.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">## read.delim can read compressed files</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>fdata &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package=</span><span class="st">"BS831"</span>), <span class="st">"GPL4372.annot"</span>), <span class="dt">skip=</span><span class="dv">27</span>,<span class="dt">row.names=</span><span class="dv">1</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata))</span></code></pre></div>
<pre><code>##  [1] "Gene.title"            "Gene.symbol"           "Gene.ID"              
##  [4] "UniGene.title"         "UniGene.symbol"        "UniGene.ID"           
##  [7] "Nucleotide.Title"      "GI"                    "GenBank.Accession"    
## [10] "Platform_CLONEID"      "Platform_ORF"          "Platform_SPOTID"      
## [13] "Chromosome.location"   "Chromosome.annotation" "GO.Function"          
## [16] "GO.Process"            "GO.Component"          "GO.Function.ID"       
## [19] "GO.Process.ID"         "GO.Component.ID"</code></pre>
<p>There are many more annotation columns than needeed, thus we extract some relevant columns, and since we are at it, we also properly sort the rows, and rename the columns, for easier handling.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>fdata &lt;-<span class="st"> </span>fdata[<span class="kw">match.nona</span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html">featureNames</a></span>(LOAD),<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(fdata)),<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Gene.symbol"</span>,<span class="st">"Gene.title"</span>,<span class="st">"Gene.ID"</span>)]</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="co">## modifying fdata colnames for ease of handling</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"Gene.symbol"</span>,<span class="st">"gene_symbol"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata))</span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"Gene.title"</span>,<span class="st">"gene_title"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata))</span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"Gene.ID"</span>,<span class="st">"GeneID"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(fdata)) </span>
<span id="cb6-6"><a href="#cb6-6"></a></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">## always be super-cautious, double- and triple-check</span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="cf">if</span> ( <span class="kw"><a href="https://rdrr.io/r/base/any.html">any</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html">featureNames</a></span>(LOAD)<span class="op">!=</span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(fdata)) ) <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span>( <span class="st">"row mismatch"</span> )</span>
<span id="cb6-9"><a href="#cb6-9"></a></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">## finally, update the gene annotation in the ExpressionSet</span></span>
<span id="cb6-11"><a href="#cb6-11"></a>LOAD1 &lt;-<span class="st"> </span>LOAD</span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD1) &lt;-<span class="st"> </span>fdata</span>
<span id="cb6-13"><a href="#cb6-13"></a></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co">## save data</span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="cf">if</span> (overwrite) {</span>
<span id="cb6-16"><a href="#cb6-16"></a>    <span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>(LOAD1, <span class="dt">file=</span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(OMPATH,<span class="st">"data/LOAD1.RDS"</span>))</span>
<span id="cb6-17"><a href="#cb6-17"></a>}</span></code></pre></div>
</div>
<div id="gene-annotation-based-on-biomart" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-annotation-based-on-biomart" class="anchor"></a>Gene Annotation based on <code>biomaRt</code>
</h2>
<p>We will use the <code>EntrezGeneID</code> column in the fData to retrieve gene symbols and descriptions with <code>biomaRt</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co">## check the annotation columns in fData</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD)))</span></code></pre></div>
<pre><code>## [1] "ID"                            "EntrezGeneID"                 
## [3] "ORF"                           "GB_ACC"                       
## [5] "PROBE_DERIVED_FROM_TRANSCRIPT" "SPOT_ID"                      
## [7] "RosettaGeneModelID"</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">## we first restrict to rows w/ non-empty EntrezGeneID annotation</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>LOAD &lt;-<span class="st"> </span>LOAD[<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD)[,<span class="st">"EntrezGeneID"</span>]),]; <span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(LOAD)</span></code></pre></div>
<pre><code>## Features 
##    25929</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co">## notice that there are replicate entries, but we will deal with them later</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>( <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD)[,<span class="st">"EntrezGeneID"</span>])) )</span></code></pre></div>
<pre><code>## [1] 21576</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co">## use biomaRt databased to retrieve the relevant annotation (type</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="co">## `?useMart` and `?getBM` for details)</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>mart &lt;-<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span>(<span class="dt">biomart=</span><span class="st">"ensembl"</span>, <span class="dt">dataset=</span><span class="st">"hsapiens_gene_ensembl"</span>)</span>
<span id="cb13-4"><a href="#cb13-4"></a>martMap &lt;-<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span>(<span class="dt">attributes=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"entrezgene_id"</span>,<span class="st">"hgnc_symbol"</span>,<span class="st">"description"</span>),</span>
<span id="cb13-5"><a href="#cb13-5"></a>                          <span class="dt">filters=</span><span class="st">"entrezgene_id"</span>,</span>
<span id="cb13-6"><a href="#cb13-6"></a>                          <span class="dt">values=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD)[,<span class="st">"EntrezGeneID"</span>],</span>
<span id="cb13-7"><a href="#cb13-7"></a>                          <span class="dt">mart=</span>mart)</span>
<span id="cb13-8"><a href="#cb13-8"></a></span>
<span id="cb13-9"><a href="#cb13-9"></a><span class="co">## remove entries with empty gene symbol</span></span>
<span id="cb13-10"><a href="#cb13-10"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(martMap) <span class="co"># before</span></span></code></pre></div>
<pre><code>## [1] 22183</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>martMap &lt;-<span class="st"> </span>martMap[martMap[,<span class="st">"hgnc_symbol"</span>]<span class="op">!=</span><span class="st">""</span>,]</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(martMap) <span class="co"># after</span></span></code></pre></div>
<pre><code>## [1] 22060</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co">## notice that there are some duplicated entrez IDs and gene symbols</span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(base<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(martMap[,<span class="st">"entrezgene_id"</span>]))     </span></code></pre></div>
<pre><code>## [1] 3920</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(base<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(martMap[,<span class="st">"hgnc_symbol"</span>]))</span></code></pre></div>
<pre><code>## [1] 3920</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co">## we adopt the simple (perhaps too simple) approach of taking the</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="co">## first occurence of each EntrezID</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(martMap)</span></code></pre></div>
<pre><code>## [1] 22060</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>martMap &lt;-<span class="st"> </span>martMap[<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(martMap[,<span class="st">"entrezgene_id"</span>]),martMap[,<span class="st">"entrezgene_id"</span>]),]</span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(martMap)</span></code></pre></div>
<pre><code>## [1] 18140</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co">## now there are no more replicated EntrezID (but there are still replicate gene symbols)</span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(base<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(martMap[,<span class="st">"entrezgene_id"</span>]))     </span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(base<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(martMap[,<span class="st">"hgnc_symbol"</span>]))</span></code></pre></div>
<pre><code>## [1] 27</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="co">## let us now match gene annotation and expression data</span></span>
<span id="cb29-2"><a href="#cb29-2"></a>matchIdx &lt;-<span class="st"> </span><span class="kw">match.nona</span>( martMap[,<span class="st">"entrezgene_id"</span>], <span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD)[,<span class="st">"EntrezGeneID"</span>] )</span>
<span id="cb29-3"><a href="#cb29-3"></a>LOAD2 &lt;-<span class="st"> </span>LOAD[matchIdx,]</span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="cf">if</span> ( <span class="kw"><a href="https://rdrr.io/r/base/any.html">any</a></span>(martMap[,<span class="st">"entrezgene_id"</span>]<span class="op">!=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD2)[,<span class="st">"EntrezGeneID"</span>]) ) <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span>( <span class="st">"row mismatch"</span> )</span>
<span id="cb29-5"><a href="#cb29-5"></a><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(LOAD2) &lt;-<span class="st"> </span>martMap</span>
<span id="cb29-6"><a href="#cb29-6"></a></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co">## save data</span></span>
<span id="cb29-8"><a href="#cb29-8"></a><span class="cf">if</span> (overwrite) {</span>
<span id="cb29-9"><a href="#cb29-9"></a>    <span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>( LOAD2, <span class="dt">file=</span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(OMPATH,<span class="st">"data/LOAD2.RDS"</span>) )</span>
<span id="cb29-10"><a href="#cb29-10"></a>}</span></code></pre></div>
</div>
<div id="merge-replicated-entries" class="section level2">
<h2 class="hasAnchor">
<a href="#merge-replicated-entries" class="anchor"></a>Merge Replicated Entries</h2>
<p>This next step is not necessary, as one might want to keep the multiple chip probes associated with the same gene symbol separate until the end of the analysis. However, if desired, one can merge the rows corresponding to multiple probes.</p>
<p>Here, we define a simple function to collapse multiple rows by median. The function is based on the R functions <code>melt</code> and <code>dcast</code> defined in the package <code>reshape2</code>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(reshape2)</span>
<span id="cb30-2"><a href="#cb30-2"></a>collapseByMedian &lt;-<span class="st"> </span><span class="cf">function</span>(eset, rowid)</span>
<span id="cb30-3"><a href="#cb30-3"></a>{</span>
<span id="cb30-4"><a href="#cb30-4"></a>    <span class="co">## require(reshape2)</span></span>
<span id="cb30-5"><a href="#cb30-5"></a></span>
<span id="cb30-6"><a href="#cb30-6"></a>    <span class="co">## remove unmapped probe sets</span></span>
<span id="cb30-7"><a href="#cb30-7"></a>    genes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(eset)[, rowid]</span>
<span id="cb30-8"><a href="#cb30-8"></a>    rows.mapped &lt;-<span class="st"> </span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(genes) <span class="op">&amp;</span><span class="st"> </span>genes <span class="op">!=</span><span class="st"> ""</span></span>
<span id="cb30-9"><a href="#cb30-9"></a>    eset &lt;-<span class="st"> </span>eset[rows.mapped,]</span>
<span id="cb30-10"><a href="#cb30-10"></a>    genes &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(eset)[, rowid]</span>
<span id="cb30-11"><a href="#cb30-11"></a></span>
<span id="cb30-12"><a href="#cb30-12"></a>    <span class="co">## collapse by median value among duplicate probes</span></span>
<span id="cb30-13"><a href="#cb30-13"></a>    df &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(eset), <span class="dt">genes =</span> genes)</span>
<span id="cb30-14"><a href="#cb30-14"></a>    df.melt &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span>(df, <span class="dt">id.vars =</span> <span class="st">"genes"</span>)</span>
<span id="cb30-15"><a href="#cb30-15"></a>    df.median.collapsed &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/reshape2/man/cast.html">dcast</a></span>(df.melt, genes <span class="op">~</span><span class="st"> </span>variable, median)</span>
<span id="cb30-16"><a href="#cb30-16"></a></span>
<span id="cb30-17"><a href="#cb30-17"></a>    <span class="co">## reassemble collapsed eset</span></span>
<span id="cb30-18"><a href="#cb30-18"></a>    fdat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(eset)</span>
<span id="cb30-19"><a href="#cb30-19"></a>    <span class="cf">if</span> ( <span class="kw"><a href="https://rdrr.io/r/base/any.html">any</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(roworder &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(df.median.collapsed[,<span class="st">'genes'</span>],fdat[,rowid]))) )</span>
<span id="cb30-20"><a href="#cb30-20"></a>        <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span>( <span class="st">"something wrong"</span> )</span>
<span id="cb30-21"><a href="#cb30-21"></a>    fdat.collapsed &lt;-<span class="st"> </span>fdat[roworder,]</span>
<span id="cb30-22"><a href="#cb30-22"></a>    eset &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html">ExpressionSet</a></span>(<span class="dt">assayData=</span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(df.median.collapsed[, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(eset)]),</span>
<span id="cb30-23"><a href="#cb30-23"></a>                          <span class="dt">phenoData=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/class.AnnotatedDataFrame.html">AnnotatedDataFrame</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span>(eset)),</span>
<span id="cb30-24"><a href="#cb30-24"></a>                          <span class="dt">featureData=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/class.AnnotatedDataFrame.html">AnnotatedDataFrame</a></span>(fdat.collapsed))</span>
<span id="cb30-25"><a href="#cb30-25"></a>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(eset)</span>
<span id="cb30-26"><a href="#cb30-26"></a>}</span>
<span id="cb30-27"><a href="#cb30-27"></a>collapseByFun &lt;-<span class="st"> </span><span class="cf">function</span>(eset,rowid,<span class="dt">method=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"median"</span>,<span class="st">"mean"</span>))</span>
<span id="cb30-28"><a href="#cb30-28"></a>{</span>
<span id="cb30-29"><a href="#cb30-29"></a>    tbl &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">key=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(eset)[,rowid],</span>
<span id="cb30-30"><a href="#cb30-30"></a>                      <span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span>(eset))</span>
<span id="cb30-31"><a href="#cb30-31"></a>    tbl1 &lt;-<span class="st"> </span>tbl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(key) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize_all</span>(median)</span>
<span id="cb30-32"><a href="#cb30-32"></a></span>
<span id="cb30-33"><a href="#cb30-33"></a>    dat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(tbl1[,<span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="st">"key"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(tbl1))])</span>
<span id="cb30-34"><a href="#cb30-34"></a>    <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(dat) &lt;-<span class="st"> </span>tbl1[,<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="st">"key"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(tbl1))]</span>
<span id="cb30-35"><a href="#cb30-35"></a></span>
<span id="cb30-36"><a href="#cb30-36"></a>    <span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html">ExpressionSet</a></span>(<span class="dt">assayData=</span>dat,</span>
<span id="cb30-37"><a href="#cb30-37"></a>                  <span class="dt">phenoData=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span>(eset)[,<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(dat),<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html">sampleNames</a></span>(eset))],</span>
<span id="cb30-38"><a href="#cb30-38"></a>                  <span class="dt">featureData=</span><span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(eset)[<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(dat),<span class="kw"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html">fData</a></span>(eset)[,rowid]),])</span>
<span id="cb30-39"><a href="#cb30-39"></a>    </span>
<span id="cb30-40"><a href="#cb30-40"></a>}</span></code></pre></div>
<p>We now apply <code>collapseByMedian</code> to both LOAD1 (n=19756 duplicates), and LOAD2 (n=27 duplicates). Notice that running it on a 600+ sample size is somewhat slow.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(LOAD1)</span>
<span id="cb31-2"><a href="#cb31-2"></a>LOAD1.collapsed &lt;-<span class="st"> </span><span class="kw">collapseByMedian</span>(LOAD1, <span class="dt">rowid=</span><span class="st">"gene_symbol"</span>)</span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(LOAD1.collapsed)</span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="cf">if</span> (overwrite) {</span>
<span id="cb31-5"><a href="#cb31-5"></a>    <span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>(LOAD1.collapsed,<span class="dt">file=</span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(OMPATH,<span class="st">"data/LOAD1.collapsed.RDS"</span>))</span>
<span id="cb31-6"><a href="#cb31-6"></a>}</span>
<span id="cb31-7"><a href="#cb31-7"></a>tbl &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">key=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"a"</span>,<span class="st">"b"</span>,<span class="st">"a"</span>,<span class="st">"b"</span>),<span class="dt">col1=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>,<span class="dt">col2=</span><span class="dv">5</span><span class="op">:</span><span class="dv">8</span>)</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(LOAD2)</span>
<span id="cb32-2"><a href="#cb32-2"></a>LOAD2.collapsed &lt;-<span class="st"> </span><span class="kw">collapseByMedian</span>(LOAD2, <span class="dt">rowid=</span><span class="st">"hgnc_symbol"</span>)</span>
<span id="cb32-3"><a href="#cb32-3"></a><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(LOAD2.collapsed)</span>
<span id="cb32-4"><a href="#cb32-4"></a><span class="cf">if</span> (overwrite) {</span>
<span id="cb32-5"><a href="#cb32-5"></a>    <span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span>(LOAD2.collapsed,<span class="dt">file=</span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(OMPATH,<span class="st">"data/LOAD2.collapsed.RDS"</span>))</span>
<span id="cb32-6"><a href="#cb32-6"></a>}</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#download-data-from-geo">Download data from GEO</a><ul class="nav nav-pills nav-stacked">
<li><a href="#use-pre-existing-gene-annotation">Use pre-existing Gene Annotation</a></li>
      <li><a href="#gene-annotation-based-on-biomart">Gene Annotation based on <code>biomaRt</code></a></li>
      <li><a href="#merge-replicated-entries">Merge Replicated Entries</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/lia978">Amy Li</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/anfederico">Anthony Federico</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
