<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sample Size and Statistical Significance • BS831</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Sample Size and Statistical Significance">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Sample Size and Statistical Significance</h1>
                        <h4 class="author">Stefano Monti</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/BS831/blob/master/vignettes/docs/SampleSize.Rmd"><code>vignettes/docs/SampleSize.Rmd</code></a></small>
      <div class="hidden name"><code>SampleSize.Rmd</code></div>

    </div>

    
    
<p>In this module, we show how testing for multiple hypotheses (genes) can increase the chance of false positives, especially for small sample sizes.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(heatmap.plus)</span></code></pre></div>
<div id="sample-sizes-effect-on-heatmap-visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-sizes-effect-on-heatmap-visualization" class="anchor"></a>Sample Size’s effect on heatmap visualization</h1>
<p>Here we show the scenario presented in class (slide “Gene markers selection: better than chance?”). In the examples below, we show heatmaps corresponding to random noise, and we show that, if enough hypotheses are tested (in this case, 10,000), and the sample size is sufficiently small (e.g., n=6), we can easily identify ‘genes’ whose expression pattern <em>seems</em> to be strongly associated with the phenotype (in this case, a random head/tail), as suggested by the heatmap with a clear blue-to-red pattern. As the sample size increases (e.g., n=100), it is more difficult to be ‘fooled’, as the corresponding heatmap shows a less clear blue-to-red pattern.</p>
<p>We start by generating a large [10,000x100] matrix filled with random values drawn from a Gaussian distribution with mean=0 and stdev=0.5.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">123</span>) <span class="co"># for reproducible results</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>Nrow &lt;-<span class="st"> </span><span class="dv">10000</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>Ncol &lt;-<span class="st"> </span><span class="dv">100</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>DAT &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(Ncol<span class="op">*</span>Nrow,<span class="dt">mean=</span><span class="dv">0</span>,<span class="dt">sd=</span><span class="fl">0.5</span>),<span class="dt">nrow=</span>Nrow,<span class="dt">ncol=</span>Ncol)</span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(DAT)</span></code></pre></div>
<p><img src="SampleSize_files/figure-html/random_matrix-1.png" width="700"></p>
<p>We then pick a small subset of columns from this matrix and randomly assign them a binary (0-1) phenotype. We then pick the top 25 markers associated to 0 and 1, and plot the corresponding heatmaps.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">## creating a black-to-red palette for heatmap display</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>ramp.br &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRamp</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>( <span class="st">"blue"</span>,<span class="st">"white"</span>,<span class="st">"red"</span>))</span>
<span id="cb3-3"><a href="#cb3-3"></a>palette.blue2red &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/grDevices/rgb.html">rgb</a></span>( <span class="kw">ramp.br</span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dt">length =</span> <span class="dv">14</span>)), <span class="dt">max =</span> <span class="dv">255</span>)</span>
<span id="cb3-4"><a href="#cb3-4"></a></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">## wrapper function extracting t-statistic and p-value from a call to function t.test</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>tscore &lt;-<span class="st"> </span><span class="cf">function</span>(y,x) { </span>
<span id="cb3-7"><a href="#cb3-7"></a>  tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span>(y<span class="op">~</span>x)</span>
<span id="cb3-8"><a href="#cb3-8"></a>  <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">score=</span>tmp<span class="op">$</span>statistic,<span class="dt">pval=</span>tmp<span class="op">$</span>p.value)</span>
<span id="cb3-9"><a href="#cb3-9"></a>}</span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">## carry out multiple differential analysis on progressively larger sample sizes</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="cf">for</span> (ndraw <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">6</span>,<span class="dv">14</span>,<span class="dv">30</span>,<span class="dv">100</span>) ) </span>
<span id="cb3-12"><a href="#cb3-12"></a>{ </span>
<span id="cb3-13"><a href="#cb3-13"></a>  <span class="co">## randomly select Ncol columns from the full matrix</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>  DATi &lt;-<span class="st"> </span>DAT[,colDraw &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(Ncol,<span class="dt">size=</span>ndraw)]</span>
<span id="cb3-15"><a href="#cb3-15"></a>  <span class="co">## generate a (head/tail) phenotype of proper size</span></span>
<span id="cb3-16"><a href="#cb3-16"></a>  pheno &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'head'</span>,<span class="st">'tail'</span>),<span class="dt">each=</span>ndraw<span class="op">/</span><span class="dv">2</span>))</span>
<span id="cb3-17"><a href="#cb3-17"></a>  <span class="co">## perform t.test on each data row with respect to the random phenotype</span></span>
<span id="cb3-18"><a href="#cb3-18"></a>  DIFi &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(DATi,<span class="dv">1</span>,tscore,<span class="dt">x=</span>pheno))</span>
<span id="cb3-19"><a href="#cb3-19"></a>  </span>
<span id="cb3-20"><a href="#cb3-20"></a>  <span class="co">## pick top 25 markers in each direction</span></span>
<span id="cb3-21"><a href="#cb3-21"></a>  topMarkers &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(DIFi[,<span class="dv">1</span>],<span class="dt">decreasing=</span><span class="ot">FALSE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>],<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(DIFi[,<span class="dv">1</span>],<span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>])</span>
<span id="cb3-22"><a href="#cb3-22"></a>  <span class="co">## visualize the corresponding heatmap of 50 markers</span></span>
<span id="cb3-23"><a href="#cb3-23"></a>  CSC &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'green'</span>,<span class="st">'orange'</span>)[pheno] <span class="co"># green=head, orange=tail</span></span>
<span id="cb3-24"><a href="#cb3-24"></a>  <span class="kw"><a href="https://rdrr.io/pkg/heatmap.plus/man/heatmap.plus.package.html">heatmap.plus</a></span>(DATi[topMarkers,],<span class="dt">col=</span>palette.blue2red,<span class="dt">Colv=</span><span class="ot">NA</span>,<span class="dt">Rowv=</span><span class="ot">NA</span>,<span class="dt">ColSideColors=</span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(CSC,CSC),</span>
<span id="cb3-25"><a href="#cb3-25"></a>               <span class="dt">main=</span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"sample size ="</span>,ndraw))</span>
<span id="cb3-26"><a href="#cb3-26"></a>  </span>
<span id="cb3-27"><a href="#cb3-27"></a>  <span class="co">## show the top markers (by p-value)</span></span>
<span id="cb3-28"><a href="#cb3-28"></a>  <span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(DIFi,<span class="dt">FDR=</span><span class="kw"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span>(DIFi[,<span class="dv">2</span>],<span class="dt">method=</span><span class="st">"BH"</span>))[topMarkers,]))</span>
<span id="cb3-29"><a href="#cb3-29"></a>}  </span></code></pre></div>
<p><img src="SampleSize_files/figure-html/sample.size.heatmaps-1.png" width="700"></p>
<pre><code>##         score.t         pval       FDR
## [1,] -11.347225 0.0005607121 0.9976306
## [2,] -11.141542 0.0003839618 0.9976306
## [3,]  -9.973578 0.0017062988 0.9976306
## [4,]  -9.125245 0.0051549978 0.9976306
## [5,]  -8.658030 0.0033664075 0.9976306
## [6,]  -8.649186 0.0038582228 0.9976306</code></pre>
<p><img src="SampleSize_files/figure-html/sample.size.heatmaps-2.png" width="700"></p>
<pre><code>##        score.t         pval       FDR
## [1,] -5.826310 9.253512e-05 0.8228090
## [2,] -5.422426 1.645618e-04 0.8228090
## [3,] -4.659613 6.096780e-04 0.9748352
## [4,] -4.592077 1.588525e-03 0.9748352
## [5,] -4.340584 1.216714e-03 0.9748352
## [6,] -4.025298 1.808353e-03 0.9748352</code></pre>
<p><img src="SampleSize_files/figure-html/sample.size.heatmaps-3.png" width="700"></p>
<pre><code>##        score.t         pval       FDR
## [1,] -5.287057 0.0000138538 0.1385380
## [2,] -4.298420 0.0001897173 0.7915671
## [3,] -4.027061 0.0004073712 0.7915671
## [4,] -4.009203 0.0004556804 0.7915671
## [5,] -3.824368 0.0007124104 0.7915671
## [6,] -3.639743 0.0011916003 0.7934205</code></pre>
<p><img src="SampleSize_files/figure-html/sample.size.heatmaps-4.png" width="700"></p>
<pre><code>##        score.t         pval      FDR
## [1,] -4.021171 0.0001140090 0.602178
## [2,] -3.896476 0.0001879315 0.602178
## [3,] -3.797355 0.0002536329 0.602178
## [4,] -3.669176 0.0003959175 0.602178
## [5,] -3.615838 0.0004827385 0.602178
## [6,] -3.591183 0.0005167335 0.602178</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#sample-sizes-effect-on-heatmap-visualization">Sample Size’s effect on heatmap visualization</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/tetomonti">Stefano Monti</a>, <a href="https://github.com/lia978">Amy Li</a>, <a href="https://github.com/ericreed">Eric Reed</a>, <a href="https://github.com/anfederico">Anthony Federico</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
