---
title: "Gene Expression Differential Analysis with Microarrays"
author: "Amy Li, Eric Reed, Stefano Monti"
output:
  html_document:
    theme: united
    toc: yes
    code_folding: show
    css: "../style/BS831.css" 
---

```{r global, echo=FALSE,message=FALSE, warning=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE)
devtools::load_all(".")
require(Biobase)
require(limma)
require(VennDiagram)
```

```{eval=FALSE}
require(BS831)
require(Biobase) 
require(limma)
require(VennDiagram)
```

Here, we show few examples of gene expression (GE) differential
analysis based on the use of the functions `t.test` and
<code>lm</code>, as well as the package `limma`, which
implements a "moderated" t-test with pooled variance (see <a
href="http://bioconductor.case.edu/bioconductor/2.10/bioc/vignettes/limma/inst/doc/usersguide.pdf">documentation</a>).

[ttest](figures/diffanal.ttest.png")

![ttest](../figs/diffanal.ttest.png)

We also recall how to perform differential analysis by fitting a
linear model, and the relationship of this approach to use of the t-test.

![lm](../figs/diffanal.lm.png)
### Differential expression analysis using <code>t-test</code>

We start by uploading a Breast Cancer dataset (in
`ExpressionSet` format), and by showing the use of the
function `t.test`. Below, we just carry out some "data
cleaning" to simplify data handling: we simplify the sampe annotation
(`pData`), and we extract the two disease states with
larger number of samples.

```{r dataprep}
## load the ExpressionSet object
## see Rmodule_HandsOn_01_microarray_processing.Rmd for how this file was generated
data(AEDAT.collapsed.mad4k)

eset <- AEDAT.collapsed.mad4k

## cleaning up/simplifying the annotation file (the original is a bit messy)
pData(eset) <- pData(eset)[,c("Characteristics.Individual","Characteristics.DiseaseState")]
colnames(pData(eset)) <- c("individual","diseaseState")

## show the data, the annotation, and the stratification by diseaseState
dim(eset)
head(pData(eset))
table(eset$diseaseState)

## Let us limit the analysis to the two most numerous diseaseState's
samples.keep <- eset$diseaseState %in% c("non-basal-like","sporadic basal-like")
cancerSet <- eset[,samples.keep]

## let's order samples by the phenotype
cancerSet <- cancerSet[,order(cancerSet$diseaseState)]
table(cancerSet$diseaseState)
```

We have extracted the data of interest into the object
<code>cancerSet</code>. We are now ready to use the
<code>t.test</code> function. We first show its use applied to a
single gene.

```{r ttest}

## split the data into the two diseaseState groups
group1 <- exprs(cancerSet)[, cancerSet$diseaseState=="non-basal-like"]
group2 <- exprs(cancerSet)[, cancerSet$diseaseState=="sporadic basal-like"]
dim(group1)                   # show the size of group1
dim(group2)                   # show the size of group2
table(cancerSet$diseaseState) # show the size concordance with the phenotype annotation

## for ease of use let's define the variable pheno (although this can be error prone)
pheno <- as.factor(cancerSet$diseaseState)

## use gene symbols to index the rows (rather than the entrez IDs)
rownames(group1) <- fData(cancerSet)$hgnc_symbol
rownames(group2) <- fData(cancerSet)$hgnc_symbol

## show few entries of group1 (the fist 5 genes in the first 5 samples)
group1[1:5,1:5]

## let us show the use of t.test on a single gene (the 1st)
T1 <- t.test(x=group1[1,],y=group2[1,],alternative="two.sided")
T1

## the default is to perform t-test w/ unequal variance. Let's try w/ equal variance
T2 <- t.test(x=group1[1,],y=group2[1,],alternative="two.sided",var.equal=TRUE)
T2
```

We then show how to apply it to all the genes in the dataset.

```{r multi.ttest1}
## apply the t.test to each gene and save the output in a data.frame
## note: this is equivalent to (i.e., no more efficient than) a for loop

ttestRes <- data.frame(t(sapply(1:nrow(group1),
     function(i){
          res <- t.test(x = group1[i, ], y = group2[i,], alternative ="two.sided")
          res.list <-  c(t.score=res$statistic,t.pvalue = res$p.value)
          return(res.list)
      })))
## use the gene names to index the rows (for interpretability)
rownames(ttestRes) <- rownames(group1)
```

In the application above, we made use of the
<code>t.test(x,y,...)</code> version of the command. However, the use
of the <code>t.test(formula,...)</code> version of the command turns
out to be simpler and more elegant, as it does not require to split
the dataset into two groups.

```{r multi.ttest2}

## application to a single gene but using the formula
T3 <- t.test(exprs(cancerSet)[1,] ~ pheno)
print(T3) # same results as before
T3$statistic==T1$statistic

## application to all genes (coerce output into data.frame for easier handling)
ttestRes1 <- as.data.frame(
  t(apply(exprs(cancerSet),1,
          function(y) {
            out <- t.test(y~pheno,var.equal=TRUE)
            c(t.score=out$statistic,t.pvalue=out$p.value)
            })))
## use the gene names to index the rows (for interpretability)
rownames(ttestRes1) <- fData(cancerSet)$hgnc_symbol

## let us add to the output data.frame an extra column reporting the FDR
## .. (i.e., the MHT-corrected p-value)
ttestRes1$t.fdr <- p.adjust(ttestRes1$t.pvalue, method = "BH")

## show few entries
head(ttestRes1)
```

#### Heatmap visualization
We now show how to visualize the top markers for each class by means
of the <code>heatmap.plus</code> function. We are using our own
<code>my.heatmap</code> function (defined in `code/heatmap.R`), which is a simple wrapper adding few
extra features (including the change of the default color palette).

```{r heatmap}
## let us sort the output by t-score
ttestOrd <- order(ttestRes1[,'t.score.t'],decreasing=TRUE)
head(ttestRes1[ttestOrd,])

## let us visualize the top 50 and bottom 50 genes
hiIdx <- ttestOrd[1:50]
loIdx <- ttestOrd[nrow(ttestRes1):(nrow(ttestRes1)-49)]
datOut <- exprs(cancerSet)[c(hiIdx,loIdx),]

## create a color bar to show sample labels (green=non-basal, orange=basal)
CSC <- rep(c("green","orange"),times=c(ncol(group1),ncol(group2)))
CSC <- cbind(CSC,CSC) # ColSideColors needs to be a matrix
my.heatmap(datOut,Colv=NA,Rowv=NA,ColSideColors=CSC)
```

### Differential expression analysis using <code>lm</code>

We now show the use of the function <code>lm</code> (for linear model)
to perform the same analysis. As discussed in class, we can regress
the expression of a gene on the phenotype variable (in this case, a
binary variable). Below, we apply it to a single gene first, and show
that the test result is the same as for the `t.test` with equal
variance. We then apply it to all the genes in the dataset.

```{r lmdiffanal}
## application to a single gene
LM1 <- lm(exprs(cancerSet)[1,] ~ pheno)
summary(LM1)

## same p-value as for T2 above
all.equal(T2$p.value,summary(LM1)$coefficients[2,"Pr(>|t|)"])

## application to all genes
ttestRes2 <- as.data.frame(t(apply(exprs(cancerSet),1,
                                   function(y) {
                                       out <- summary(lm(y~pheno))$coefficients
                                       c(t.score=out[2,"t value"],t.pvalue=out[2,"Pr(>|t|)"])
                                   })))
## use the gene names to index the rows (for interpretability)
rownames(ttestRes2) <- fData(cancerSet)$hgnc_symbol

## let us add to the output data.frame an extra column reportding the FDR
## .. (i.e., the MHT-corrected p-value)
ttestRes2$t.fdr <- p.adjust(ttestRes2$t.pvalue, method = "BH")

## the scores are the same (modulo the sign, which is arbitrary)
plot(ttestRes1$t.score.t,ttestRes2$t.score,pch=20,xlab="t.test scores",ylab="lm scores")
all.equal(-ttestRes1$t.score.t,ttestRes2$t.score)
```

Finally, let us use the package <code>limma</code> for the same purpose.

### Differential expression analysis using the <code>limma</code> package

With this package, we are performing differential analysis taking the
"linear regression" approach (i.e., by regressing each gene's
expression on the phenotype variable (and possibly, on other
covariates). The main difference is in the estimation of the variance,
which is here 'pooled' across multiple genes with similar expression
profiles. This pooling is particularly useful with small sample size
datasets, where the variance estimates of individual genes can be
extremely noisy, and pooling multiple genes allows for "borrowing" of
information.

```{r limma}
#differential expression analysis using limma

design <- model.matrix(~0 + factor(pheno))
colnames(design)
colnames(design) <- c("nonbasallike", "sporadicbasallike")
head(design)

contrast.matrix <- makeContrasts(nonbasallike-sporadicbasallike, levels = design)
fit <- lmFit(cancerSet, design)
fit <- contrasts.fit(fit,contrast.matrix)
fit <- eBayes(fit)
head(fit$coefficients)
#get full differential expression output table, sorted by p-value
limmaRes <- topTable(fit, adjust.method = "BH", n = Inf, sort.by = "P")
#subset to genes with adjusted p-value cutoff
adjpcutoff <- 0.001
limmaRes.sig <- subset(limmaRes, adj.P.Val < adjpcutoff)
nrow(limmaRes.sig)
head(limmaRes.sig)

topGenes <- limmaRes.sig$hgnc_symbol
topGenes.up <- subset(limmaRes.sig, t<0)$hgnc_symbol
topGenes.down <- subset(limmaRes.sig, t>=0)$hgnc_symbol

#out.dir <- paste(OMPATH,"data/results",sep="")
#saveRDS(topGenes, paste(out.dir, "/", "topGenes.RDS", sep = ""))
#saveRDS(topGenes.up, paste(out.dir, "/", "topGenes.up.RDS", sep = ""))
#saveRDS(topGenes.down, paste(out.dir, "/", "topGenes.down.RDS", sep = ""))
```

**Comparing t-test and limma results**

```{r compare DE}
#comparing t-test results to limma results
combinedRes <- cbind(limmaRes, ttestRes1[match(limmaRes$hgnc_symbol, rownames(ttestRes1)),])
plot(combinedRes$t, combinedRes$t.score.t,
 xlab = "limma t-statistic", ylab = "t-test t-statistic", pch = 20, cex = 0.5)
plot(combinedRes$P.Value, combinedRes$t.pvalue,
 xlab = "limma p-value", ylab = "t-test pvalue", pch = 20, cex = 0.5)
plot(combinedRes$adj.P.Val, combinedRes$t.fdr,
 xlab = "limma fdr", ylab = "t-test fdr", pch = 20, cex =0.5)

#limma performs eBayes shrinkage of variance estimates, resulting in moderated t-statistics
empS <- apply(exprs(cancerSet), 1, var)
par(mar=c(5,5,2,2))
n <- length(empS)
plot(1,1,xlim=c(0,14),ylim=c(0,1),type="n",
     xlab="variance estimates",ylab="",yaxt="n")
axis(2,at=c(0.9,0.1),c("shrunken \n variance","sample \n variance"),las=2)
segments(fit$s2.post, rep(.9, n),empS,rep(.1,n))

#what is the overlap between t-ttest and limma derived significant genes? (top1k)
top1k.ttest <- rownames(combinedRes)[order(combinedRes$t.fdr, decreasing = FALSE)[1:1000]]
top1k.limma <- rownames(combinedRes)[order(combinedRes$adj.P.Val, decreasing = FALSE)[1:1000]]
top1k <- list(top1k.ttest = top1k.ttest, top1k.limma = top1k.limma)

p <- venn.diagram(x = top1k, filename = NULL)
grid.newpage()
grid.draw(p)
```


**Modelling with covariates**

```{r newESET}
data(breast_loi_133p2)

## load the ExpressionSet object
BC <- breast_loi_133p2
pData(BC)[1:5,1:6] # show some data annotation

## select top 5000 genes by MAD
MAD <- apply(exprs(BC),1,mad)
BC5K <- BC[order(MAD,decreasing=TRUE)[1:5000],]
dim(BC5K)

## to reuse the same code below, just assign the new dataset to BC
BC <- BC5K
  
# Reformat LN variable and subset
BC$LN_status <- c("negative","positive")[BC$LN_status+1]
pData(BC) <- pData(BC)[, c("LN_status", "age")]
```

**Model without age covariate**
```{r limmaNEW}
#Next, we'll add age as a covariate

design <- model.matrix(~ 0 + factor(LN_status), data = pData(BC))
colnames(design)
colnames(design) <- c("negative", "positive")
head(design)

contrast.matrix <- makeContrasts(positive-negative, levels = design)
fit <- lmFit(BC, design)
fit <- contrasts.fit(fit,contrast.matrix)
fit <- eBayes(fit)
head(fit$coefficients)

## get full differential expression output table, sorted by p-value
limmaRes <- topTable(fit, adjust.method="BH", n=Inf, sort.by="P")
head(limmaRes)
```

**Model with age covariate**
```{r limmaCOV}
#Next, we'll add age as a covariate

designBC <- model.matrix(~ 0 + factor(LN_status) + age, data = pData(BC))
colnames(designBC)
colnames(designBC) <- c("negative", "positive", "age")
head(designBC)

contrast.matrixBC <- makeContrasts(positive-negative, levels = designBC)
fitBC <- lmFit(BC, designBC)

# Difference in expression between LN_status = positive and LN_status = negative individuals
fitC <- contrasts.fit(fitBC,contrast.matrixBC)
fitC <- eBayes(fitC)
head(fitC$coefficients)
#get full differential expression output table, sorted by p-value
limmaResC <- topTable(fitC, adjust.method = "BH", n = Inf, sort.by = "P")
head(limmaResC)

# Effect of age
fitA <- eBayes(fitBC)
head(fitA$coefficients)
#get full differential expression output table, sorted by p-value
limmaResA <- topTable(fitA, adjust.method = "BH", n = Inf, sort.by = "P", coef = "age")
head(limmaResA)
```

**Model with age:LN_status interaction**
```{r limmaInter}
#Next, we'll model the interaction between age and LN_status
designI <- model.matrix(~ 0 + factor(LN_status) + age + factor(LN_status):age, data = pData(BC))
colnames(designI)
colnames(designI) <- c("negative", "positive", "age", "positiveIage")
head(designI)

contrast.matrixI<- makeContrasts(positive-negative, levels = designI)
fitI <- lmFit(BC, designI)

# Difference in expression between LN_status = positive and LN_status = negative individuals
fitIC <- contrasts.fit(fitI,contrast.matrixI)
fitIC <- eBayes(fitIC)
head(fitIC$coefficients)
#get full differential expression output table, sorted by p-value
limmaResIC <- topTable(fitIC, adjust.method = "BH", n = Inf, sort.by = "P")
head(limmaResIC)

# Effect of age on LN_status = negative individuals
fitIA <- eBayes(fitI)
head(fitI$coefficients)
#get full differential expression output table, sorted by p-value
limmaResIA <- topTable(fitIA, adjust.method = "BH", n = Inf, sort.by = "P", coef = "age")
head(limmaResIA)

# Difference in effect of Age between LN_status = positive and LN_status = negative participants
limmaResIP <- topTable(fitIA, adjust.method = "BH", n = Inf, sort.by = "P", coef = "positiveIage")
head(limmaResIP)
```
