<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>ggheat.continuous.group ggheatmap function for plotting groups of esets, within-group column clustering — ggheat.continuous.group • BS831</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  <script src="../extra.js"></script>

<meta property="og:title" content="ggheat.continuous.group ggheatmap function for plotting groups of esets, within-group column clustering — ggheat.continuous.group" />
<meta property="og:description" content="ggheat.continuous.group ggheatmap function for plotting groups of esets, within-group column clustering" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>ggheat.continuous.group</code> ggheatmap function for plotting groups of esets, within-group column clustering</h1>
    <small class="dont-index">Source: <a href='https://github.com/montilab/BS831/blob/master/R/ggheat.continuous.R'><code>R/ggheat.continuous.R</code></a></small>
    <div class="hidden name"><code>ggheat.continuous.group.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>ggheat.continuous.group</code> ggheatmap function for plotting groups of esets, within-group column clustering</p>
    </div>

    <pre class="usage"><span class='fu'>ggheat.continuous.group</span><span class='op'>(</span>
  <span class='va'>esetlist</span>,
  <span class='va'>hclist</span>,
  <span class='va'>hrlist</span>,
  <span class='va'>hmcolors</span>,
  hmtitle <span class='op'>=</span> <span class='st'>"expression"</span>,
  <span class='va'>col_lab</span>,
  <span class='va'>col_legend</span>,
  ylabstr <span class='op'>=</span> <span class='st'>""</span>,
  <span class='va'>fout</span>,
  p.heights <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1.5</span>, <span class='fl'>0.5</span>, <span class='fl'>5</span><span class='op'>)</span>,
  xsize <span class='op'>=</span> <span class='fl'>4</span>,
  ysize <span class='op'>=</span> <span class='fl'>4</span>,
  ysizelab <span class='op'>=</span> <span class='fl'>7</span>,
  xleft <span class='op'>=</span> <span class='fl'>0.15</span>,
  xright <span class='op'>=</span> <span class='fl'>0.24</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>esetlist</th>
      <td><p>list of expression sets, see ?heatmap.make.groups for preprocessing</p></td>
    </tr>
    <tr>
      <th>hclist</th>
      <td><p>list of column clustering from hclust or hcopt</p></td>
    </tr>
    <tr>
      <th>hrlist</th>
      <td><p>list of row clustering from hclust or hcopt</p></td>
    </tr>
    <tr>
      <th>hmcolors</th>
      <td><p>function for heatmap color gradient, default:
 hmcolors&lt;-function(... ) scale_fill_gradient2(low = "blue",
mid = "white", high = "red", midpoint = 0, limits=c(-3,3), oob=squish, ...)</p></td>
    </tr>
    <tr>
      <th>hmtitle</th>
      <td><p>title of heatmap color label, default: "expression"</p></td>
    </tr>
    <tr>
      <th>col_lab</th>
      <td><p>column color label to display, one or more of columns of pData(eset)</p></td>
    </tr>
    <tr>
      <th>col_legend</th>
      <td><p>named list of column color legends, names correspond to col_lab, see format:
col_legend&lt;-list(COL1 = list(col_breaks = COL1levels,
                            col_values = brewer.pal(length(COL1levels),"Set1"),
                            col_labels = COL1levels),
    COL2 = list(col_breaks = COL2levels,
                            col_values = brewer.pal(length(COL2levels),"Set2"),
                            col_labels = COL2levels))</p></td>
    </tr>
    <tr>
      <th>ylabstr</th>
      <td><p>y axis title string,</p></td>
    </tr>
    <tr>
      <th>fout</th>
      <td><p>filename to save plot, default NA, save as object, to display to device, use grid.arrange</p></td>
    </tr>
    <tr>
      <th>p.heights</th>
      <td><p>panel heights default c(1.5, 0.5, 5),</p></td>
    </tr>
    <tr>
      <th>xsize</th>
      <td><p>default 4 x axis labels (uses colnames), set to 0 to turn off display</p></td>
    </tr>
    <tr>
      <th>ysize</th>
      <td><p>default 4, y axis labels (uses rownames), set to 0 to turn off display</p></td>
    </tr>
    <tr>
      <th>ysizelab</th>
      <td><p>default 7, text size for y axis title (specified in ylabstr)</p></td>
    </tr>
    <tr>
      <th>xleft</th>
      <td><p>default 0.15, displayed width for left panel + margin as fraction of whole plot width</p></td>
    </tr>
    <tr>
      <th>xright</th>
      <td><p>default 0.24, displayed width for right margin as fraction of whole plot width</p></td>
    </tr>
    </table>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>tcga.subset.400g.200s</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: data set ‘tcga.subset.400g.200s’ not found</span></div><div class='input'><span class='va'>dat</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'dat' not found</span></div><div class='input'>
<span class='va'>subtypelevels</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>dat</span><span class='op'>$</span><span class='va'>subtype</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in levels(dat$subtype): object 'dat' not found</span></div><div class='input'>
<span class='va'>grps</span><span class='op'>&lt;-</span><span class='fu'><a href='ggheat.make.groups.html'>ggheat.make.groups</a></span><span class='op'>(</span>eset <span class='op'>=</span> <span class='va'>dat</span>,
  labelcol <span class='op'>=</span> <span class='st'>"subtype"</span>,  <span class='co'>#column name for grouping in pData(eset)</span>
  labelvals <span class='op'>=</span> <span class='va'>subtypelevels</span>, <span class='co'>#values to group on (e.g. factor levels of pData(eset)[, labelcol])</span>
  clustFUN <span class='op'>=</span> <span class='va'>clust_eset</span> <span class='co'>#clustering function for eset</span>
  <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in is.data.frame(x): object 'dat' not found</span></div><div class='input'>
<span class='va'>col_legend</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>subtype <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>col_breaks <span class='op'>=</span> <span class='va'>subtypelevels</span>,
    col_values <span class='op'>=</span> <span class='fu'>brewer.pal</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>subtypelevels</span><span class='op'>)</span>,<span class='st'>"Set1"</span><span class='op'>)</span>,
    col_labels <span class='op'>=</span> <span class='va'>subtypelevels</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'subtypelevels' not found</span></div><div class='input'>
<span class='va'>esetlist</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>grps</span><span class='op'>$</span><span class='va'>esetlist</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>i</span><span class='op'>)</span> <span class='fu'>scale_row</span><span class='op'>(</span><span class='va'>i</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in lapply(grps$esetlist, function(i) scale_row(i)): object 'grps' not found</span></div><div class='input'>
<span class='va'>hmcolors</span><span class='op'>&lt;-</span><span class='kw'>function</span><span class='op'>(</span><span class='va'>...</span> <span class='op'>)</span> <span class='fu'>scale_fill_gradient2</span><span class='op'>(</span>low <span class='op'>=</span> <span class='st'>"blue"</span>, mid <span class='op'>=</span> <span class='st'>"white"</span>,
       high <span class='op'>=</span> <span class='st'>"red"</span>, midpoint <span class='op'>=</span> <span class='fl'>0</span>, limits<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>3</span>,<span class='fl'>3</span><span class='op'>)</span>, oob<span class='op'>=</span><span class='va'>squish</span>, <span class='va'>...</span><span class='op'>)</span>

<span class='va'>p</span><span class='op'>&lt;-</span><span class='fu'>ggheat.continuous.group</span><span class='op'>(</span><span class='va'>esetlist</span>,
  <span class='va'>grps</span><span class='op'>$</span><span class='va'>hclist</span>,
  <span class='va'>grps</span><span class='op'>$</span><span class='va'>hrlist</span>,
  <span class='va'>hmcolors</span>,
  hmtitle <span class='op'>=</span> <span class='st'>"row-zscore GE"</span>,
  col_lab <span class='op'>=</span> <span class='st'>"subtype"</span>,
  col_legend <span class='op'>=</span> <span class='va'>col_legend</span>,
  ylabstr <span class='op'>=</span> <span class='st'>""</span>,
  fout  <span class='op'>=</span> <span class='cn'>NA</span>,
  p.heights <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1.5</span>, <span class='fl'>0.5</span>, <span class='fl'>5</span><span class='op'>)</span>,
  xsize <span class='op'>=</span> <span class='fl'>0</span>,
  ysize <span class='op'>=</span> <span class='fl'>0</span>,
  ysizelab <span class='op'>=</span> <span class='fl'>7</span>,
  xleft <span class='op'>=</span> <span class='fl'>0.10</span>,
  xright <span class='op'>=</span> <span class='fl'>0.24</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in ggheat.continuous.group(esetlist, grps$hclist, grps$hrlist, hmcolors,     hmtitle = "row-zscore GE", col_lab = "subtype", col_legend = col_legend,     ylabstr = "", fout = NA, p.heights = c(1.5, 0.5, 5), xsize = 0,     ysize = 0, ysizelab = 7, xleft = 0.1, xright = 0.24): object 'esetlist' not found</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/tetomonti'>Stefano Monti</a>, <a href='https://github.com/anfederico'>Anthony Federico</a>, <a href='https://github.com/ericreed'>Eric Reed</a>, <a href='https://github.com/lia978'>Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


