<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>heatmap.ggplot2 — heatmap.ggplot2 • BS831</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  <script src="../extra.js"></script>

<meta property="og:title" content="heatmap.ggplot2 — heatmap.ggplot2" />
<meta property="og:description" content="heatmap.ggplot2 is the main function to draw and/or save the heatmap from an 
expression set object" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BS831</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Markdowns
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Introduction to Genomics Analysis</li>
    <li>
      <a href="../articles/docs/matrixOp.html">Efficient Computation in R</a>
    </li>
    <li>
      <a href="../articles/docs/ExpressionSet.html">The ExpressionSet Data Object</a>
    </li>
    <li>
      <a href="../articles/docs/Distributions.html">Probability Distributions</a>
    </li>
    <li>
      <a href="../articles/docs/Mix_Normals.html">Mixture of Normals</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Preprocessing and Quality Control</li>
    <li>
      <a href="../articles/docs/Microarray_Processing.html">Data Acquisition and ExpressionSet Basics</a>
    </li>
    <li>
      <a href="../articles/docs/microarrayQC.html">Quality Control</a>
    </li>
    <li>
      <a href="../articles/docs/Dimensionality_Reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/docs/Heatmaps.html">Heatmaps</a>
    </li>
    <li>
      <a href="../articles/docs/hclust.html">Hierarchical Clustering</a>
    </li>
    <li>
      <a href="../articles/docs/quantileNormalization.html">Quantile Normalization</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Comparative Experiments I: Microarrays and Linear Models</li>
    <li>
      <a href="../articles/docs/SampleSize.html">Sample Size and Statistical Significance</a>
    </li>
    <li>
      <a href="../articles/docs/MHTcorrection.html">Multiple-Hypothesis Correction</a>
    </li>
    <li>
      <a href="../articles/docs/Linear_Regression.html">Linear Regression</a>
    </li>
    <li>
      <a href="../articles/docs/diffanalLM.html">Differential Analysis as Linear Regression</a>
    </li>
    <li>
      <a href="../articles/docs/Diffanalysis.html">Gene Expression Differential Analysis with Microarrays</a>
    </li>
    <li>
      <a href="../articles/docs/Diffanal_LimmaBC.html">Gene Expression Differential Analysis based on Limma</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Comparative Experiments II: RNA-seq and Generalized Lineary Models</li>
    <li>
      <a href="../articles/docs/RNAseq_ScalNorm.html">RNA-Seq Data Scaling and Normalization</a>
    </li>
    <li>
      <a href="../articles/docs/DiffanalysisRNAseqComparison.html">RNA-seq Differential Expression Analysis with DEseq2, edgeR and limma</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Comparative Experiments III: Differential Enrichment Analysis</li>
    <li>
      <a href="../articles/docs/HyperEnrichment.html">HyperGeometric Enrichment</a>
    </li>
    <li>
      <a href="../articles/docs/KS_Enrichment.html">KS-based Enrichment</a>
    </li>
    <li>
      <a href="../articles/docs/GenesetMapping4KS.html">Mapping genesets for KS testing</a>
    </li>
    <li>
      <a href="../articles/docs/GeneSetProjectionGSVA.html">Gene Set Projection based on KS</a>
    </li>
    <li>
      <a href="../articles/docs/GeneSetProjectionASSIGN.html">Geneset projection based on ASSIGN</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Classification</li>
    <li>
      <a href="../articles/docs/featureRedundancy.html">Feature Redundancy: the Good and the Bad</a>
    </li>
    <li>
      <a href="../articles/docs/Classification_Caret.html">Classification Using Caret</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Clustering</li>
    <li>
      <a href="../articles/docs/Clustering.html">Clustering and Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/montilab/BS831">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>heatmap.ggplot2</h1>
    <small class="dont-index">Source: <a href='https://github.com/montilab/BS831/blob/master/R/heatmap.ggplot.R'><code>R/heatmap.ggplot.R</code></a></small>
    <div class="hidden name"><code>heatmap.ggplot2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>heatmap.ggplot2</code> is the main function to draw and/or save the heatmap from an 
expression set object</p>
    </div>

    <pre class="usage"><span class='fu'>heatmap.ggplot2</span><span class='op'>(</span>
  <span class='va'>eSet</span>,
  brewer.pal.name <span class='op'>=</span> <span class='st'>"RdBu"</span>,
  brewer.pal.rev <span class='op'>=</span> <span class='cn'>TRUE</span>,
  brewer.numColors <span class='op'>=</span> <span class='fl'>11</span>,
  col.legend.brewer <span class='op'>=</span> <span class='st'>""</span>,
  row.legend.brewer <span class='op'>=</span> <span class='st'>""</span>,
  col.clust <span class='op'>=</span> <span class='cn'>TRUE</span>,
  row.clust <span class='op'>=</span> <span class='cn'>TRUE</span>,
  col.clust.hc <span class='op'>=</span> <span class='cn'>NA</span>,
  row.clust.hc <span class='op'>=</span> <span class='cn'>NA</span>,
  col.lab <span class='op'>=</span> <span class='st'>""</span>,
  row.lab <span class='op'>=</span> <span class='st'>""</span>,
  heatmap.y.text <span class='op'>=</span> <span class='cn'>FALSE</span>,
  heatmap.x.text <span class='op'>=</span> <span class='cn'>FALSE</span>,
  heatmap.y.text.size <span class='op'>=</span> <span class='fl'>7</span>,
  heatmap.x.text.size <span class='op'>=</span> <span class='fl'>6</span>,
  heatmap.colorlegend.name <span class='op'>=</span> <span class='st'>"val"</span>,
  title.text <span class='op'>=</span> <span class='st'>""</span>,
  col.legend.name <span class='op'>=</span> <span class='st'>""</span>,
  row.legend.name <span class='op'>=</span> <span class='st'>""</span>,
  legend.lab.max.char <span class='op'>=</span> <span class='fl'>15</span>,
  row.scaling <span class='op'>=</span> <span class='st'>"none"</span>,
  z.norm <span class='op'>=</span> <span class='cn'>FALSE</span>,
  cuttree.col <span class='op'>=</span> <span class='fl'>0</span>,
  cuttree.row <span class='op'>=</span> <span class='fl'>0</span>,
  verbose <span class='op'>=</span> <span class='cn'>FALSE</span>,
  grid.heights <span class='op'>=</span> <span class='cn'>NA</span>,
  grid.widths <span class='op'>=</span> <span class='cn'>NA</span>,
  show <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>eSet</th>
      <td><p>expression set object to plot</p></td>
    </tr>
    <tr>
      <th>brewer.pal.name</th>
      <td><p>name of colorbrewer, see options at RColorBrewer::display.brewer.all()</p></td>
    </tr>
    <tr>
      <th>brewer.pal.rev</th>
      <td><p>reverse colorbrewer (TRUE or FALSE)</p></td>
    </tr>
    <tr>
      <th>brewer.numColors</th>
      <td><p>number of colors for heatmap palette default 11</p></td>
    </tr>
    <tr>
      <th>col.legend.brewer</th>
      <td><p>string vector of hexdecimal values for column labels, see example p3 and p4 for how to determine annotation labels and assigning colors</p></td>
    </tr>
    <tr>
      <th>row.legend.brewer</th>
      <td><p>string vector of hexdecimal values for row labels</p></td>
    </tr>
    <tr>
      <th>col.clust</th>
      <td><p>perform column-wise hierarchical clustering (TRUE or FALSE)</p></td>
    </tr>
    <tr>
      <th>row.clust</th>
      <td><p>perform row-wise hierarchical clustering (TRUE or FALSE)</p></td>
    </tr>
    <tr>
      <th>col.clust.hc</th>
      <td><p>hc object to be passed to as.dendrogram, available when col.clust = TRUE, default NA: hc.col = hcopt(stats::dist(t(x)), method="ward.D")</p></td>
    </tr>
    <tr>
      <th>row.clust.hc</th>
      <td><p>hc object to be passed to as.dendrogram, available when row.clust = TRUE, default NA: hc.row = hcopt(stats::as.dist(1-cor(t(x))),method="ward.D")</p></td>
    </tr>
    <tr>
      <th>col.lab</th>
      <td><p>column labels to include: subset of pData colnames character vector</p></td>
    </tr>
    <tr>
      <th>row.lab</th>
      <td><p>row labels to include: subset of fData colnames character vector</p></td>
    </tr>
    <tr>
      <th>heatmap.y.text</th>
      <td><p>include y axis labels for heatmap, uses rownames(eSet) (TRUE or FALSE)</p></td>
    </tr>
    <tr>
      <th>heatmap.x.text</th>
      <td><p>include x axis labels for heatmap, uses colnames(eSet)) (TRUE or FALSE)</p></td>
    </tr>
    <tr>
      <th>heatmap.y.text.size</th>
      <td><p>text size for y axis labels</p></td>
    </tr>
    <tr>
      <th>heatmap.x.text.size</th>
      <td><p>text size for x axis labels</p></td>
    </tr>
    <tr>
      <th>heatmap.colorlegend.name</th>
      <td><p>name for heatmap color legend</p></td>
    </tr>
    <tr>
      <th>title.text</th>
      <td><p>main title for the plot</p></td>
    </tr>
    <tr>
      <th>col.legend.name</th>
      <td><p>character vector for subset of col.lab to include in color legend</p></td>
    </tr>
    <tr>
      <th>row.legend.name</th>
      <td><p>character vector for subset of row.lab to include in color legend</p></td>
    </tr>
    <tr>
      <th>legend.lab.max.char</th>
      <td><p>number of characters limit for legend labels, default = 15</p></td>
    </tr>
    <tr>
      <th>row.scaling</th>
      <td><p>how should rows be scaled ("none", "quantile", "z-score.all", or "z-score.capped")</p></td>
    </tr>
    <tr>
      <th>z.norm</th>
      <td><p>heatmap colors reflect z-scores rather than original values, can be TRUE if row.scaling is "none"</p></td>
    </tr>
    <tr>
      <th>cuttree.col</th>
      <td><p>number of clusters for columns, default 0: do not show cluster assignment</p></td>
    </tr>
    <tr>
      <th>cuttree.row</th>
      <td><p>number of clusters for rows, default 0: do not show cluster assignment</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>return additional clustred annotation information as well as heatmap</p></td>
    </tr>
    <tr>
      <th>grid.heights</th>
      <td><p>numeric vector of length 7 representing heights of subpanels, leave NA for default</p></td>
    </tr>
    <tr>
      <th>grid.widths</th>
      <td><p>numeric vector of length 4 represeting weights of subpanels, leave NA for default</p></td>
    </tr>
    <tr>
      <th>show</th>
      <td><p>prints the heatmap within execution of the function (TRUE or FALSE)</p></td>
    </tr>
    </table>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>#Use example data #1, for data set information: ?eSet1</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>eSet.brca.100</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: data set ‘eSet.brca.100’ not found</span></div><div class='input'><span class='va'>eSet1</span><span class='op'>&lt;-</span><span class='va'>eSet.brca.100</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'eSet.brca.100' not found</span></div><div class='input'><span class='va'>eSet1</span><span class='op'>&lt;-</span><span class='va'>eSet1</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>,<span class='fl'>1</span><span class='op'>:</span><span class='fl'>25</span><span class='op'>]</span>
</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'eSet1' not found</span></div><div class='input'>
<span class='va'>p1</span><span class='op'>&lt;-</span><span class='fu'>heatmap.ggplot2</span><span class='op'>(</span>eSet<span class='op'>=</span><span class='va'>eSet1</span>, col.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, row.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, 
col.clust.hc <span class='op'>=</span> <span class='cn'>NA</span>, row.clust.hc <span class='op'>=</span> <span class='cn'>NA</span>,
col.lab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, row.lab <span class='op'>=</span> <span class='st'>""</span>, 
heatmap.y.text <span class='op'>=</span> <span class='cn'>TRUE</span>, heatmap.x.text <span class='op'>=</span> <span class='cn'>TRUE</span>,
heatmap.colorlegend.name <span class='op'>=</span> <span class='st'>"RNASeq_expression"</span>,
title.text <span class='op'>=</span> <span class='st'>"TCGA BRCA log2 RNA-seq expression, z-score row normalized"</span>,
col.legend.name <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, 
row.legend.name <span class='op'>=</span> <span class='st'>""</span>, 
row.scaling <span class='op'>=</span> <span class='st'>"z-score.capped"</span>, 
z.norm <span class='op'>=</span> <span class='cn'>FALSE</span>, 
cuttree.col <span class='op'>=</span> <span class='fl'>4</span>, cuttree.row <span class='op'>=</span> <span class='fl'>3</span>,
verbose <span class='op'>=</span> <span class='cn'>FALSE</span>, show <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: `axis.ticks.margin` is deprecated. Please set `margin` property of `axis.text` instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `panel.margin` is deprecated. Please use `panel.spacing` property instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `legend.margin` must be specified using `margin()`. For the old behavior use legend.spacing</span></div><div class='output co'>#&gt; <span class='error'>Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'exprs': object 'eSet1' not found</span></div><div class='input'><span class='fu'>grid.arrange</span><span class='op'>(</span><span class='va'>p1</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.arrange(p1): could not find function "grid.arrange"</span></div><div class='input'>

<span class='va'>x</span><span class='op'>&lt;-</span><span class='fu'>exprs</span><span class='op'>(</span><span class='va'>eSet1</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in exprs(eSet1): could not find function "exprs"</span></div><div class='input'><span class='va'>hc.row</span><span class='op'>&lt;-</span><span class='fu'>hcopt</span><span class='op'>(</span><span class='fu'>stats</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/stats/dist.html'>as.dist</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/stats/cor.html'>cor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,method<span class='op'>=</span><span class='st'>"ward.D"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in t(x): object 'x' not found</span></div><div class='input'><span class='va'>hc.col</span> <span class='op'>&lt;-</span> <span class='fu'>hcopt</span><span class='op'>(</span><span class='fu'>stats</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/stats/dist.html'>dist</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, method <span class='op'>=</span> <span class='st'>"euclidean"</span><span class='op'>)</span>, method<span class='op'>=</span><span class='st'>"ward.D"</span><span class='op'>)</span> 
</div><div class='output co'>#&gt; <span class='error'>Error in t(x): object 'x' not found</span></div><div class='input'>
<span class='co'>#Adding custom hclust object in col.clust.hc and row.clust.hc</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>heatmap.ggplot2</span><span class='op'>(</span>eSet<span class='op'>=</span><span class='va'>eSet1</span>, col.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, row.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, 
col.clust.hc <span class='op'>=</span> <span class='va'>hc.col</span>, row.clust.hc <span class='op'>=</span> <span class='va'>hc.row</span>,
col.lab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, row.lab <span class='op'>=</span> <span class='st'>""</span>, 
heatmap.y.text <span class='op'>=</span> <span class='cn'>TRUE</span>, heatmap.x.text <span class='op'>=</span> <span class='cn'>TRUE</span>,
heatmap.colorlegend.name <span class='op'>=</span> <span class='st'>"RNASeq_expression"</span>,
title.text <span class='op'>=</span> <span class='st'>"TCGA BRCA log2 RNA-seq expression, z-score row normalized"</span>,
col.legend.name <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, 
row.legend.name <span class='op'>=</span> <span class='st'>""</span>, 
row.scaling <span class='op'>=</span> <span class='st'>"z-score.capped"</span>, 
z.norm <span class='op'>=</span> <span class='cn'>FALSE</span>, 
cuttree.col <span class='op'>=</span> <span class='fl'>4</span>, cuttree.row <span class='op'>=</span> <span class='fl'>3</span>,
verbose <span class='op'>=</span> <span class='cn'>FALSE</span>, show <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: `axis.ticks.margin` is deprecated. Please set `margin` property of `axis.text` instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `panel.margin` is deprecated. Please use `panel.spacing` property instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `legend.margin` must be specified using `margin()`. For the old behavior use legend.spacing</span></div><div class='output co'>#&gt; <span class='error'>Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'exprs': object 'eSet1' not found</span></div><div class='input'><span class='fu'>grid.newpage</span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.newpage(): could not find function "grid.newpage"</span></div><div class='input'><span class='fu'>grid.arrange</span><span class='op'>(</span><span class='va'>p2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.arrange(p2): could not find function "grid.arrange"</span></div><div class='input'>

<span class='co'>#Saving plot in verbose format</span>
<span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'>heatmap.ggplot2</span><span class='op'>(</span>eSet<span class='op'>=</span><span class='va'>eSet1</span>, col.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, row.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, 
col.clust.hc <span class='op'>=</span> <span class='va'>hc.col</span>, row.clust.hc <span class='op'>=</span> <span class='va'>hc.row</span>,
col.lab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, row.lab <span class='op'>=</span> <span class='st'>""</span>, 
heatmap.y.text <span class='op'>=</span> <span class='cn'>TRUE</span>, heatmap.x.text <span class='op'>=</span> <span class='cn'>TRUE</span>,
heatmap.colorlegend.name <span class='op'>=</span> <span class='st'>"RNASeq_expression"</span>,
title.text <span class='op'>=</span> <span class='st'>"TCGA BRCA log2 RNA-seq expression, z-score row normalized"</span>,
col.legend.name <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, 
row.legend.name <span class='op'>=</span> <span class='st'>""</span>, 
row.scaling <span class='op'>=</span> <span class='st'>"z-score.capped"</span>, 
z.norm <span class='op'>=</span> <span class='cn'>FALSE</span>, 
cuttree.col <span class='op'>=</span> <span class='fl'>4</span>, cuttree.row <span class='op'>=</span> <span class='fl'>3</span>,
verbose <span class='op'>=</span> <span class='cn'>TRUE</span>, show <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: `axis.ticks.margin` is deprecated. Please set `margin` property of `axis.text` instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `panel.margin` is deprecated. Please use `panel.spacing` property instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `legend.margin` must be specified using `margin()`. For the old behavior use legend.spacing</span></div><div class='output co'>#&gt; <span class='error'>Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'exprs': object 'eSet1' not found</span></div><div class='input'><span class='fu'>grid.newpage</span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.newpage(): could not find function "grid.newpage"</span></div><div class='input'><span class='fu'>grid.arrange</span><span class='op'>(</span><span class='va'>p3</span><span class='op'>$</span><span class='va'>heatmap</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.arrange(p3$heatmap): could not find function "grid.arrange"</span></div><div class='input'>

<span class='co'>#Adding custom colors to column and row annotation labels</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>p3</span><span class='op'>$</span><span class='va'>meta.c</span><span class='op'>$</span><span class='va'>id</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in print(p3$meta.c$id): object 'p3' not found</span></div><div class='input'><span class='va'>meta.c.color.string</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"yellow"</span>, <span class='st'>"khaki3"</span>, <span class='st'>"gold"</span>, <span class='st'>"chocolate"</span>, <span class='st'>"darkred"</span>, <span class='st'>"cyan"</span><span class='op'>)</span>
<span class='va'>meta.c.color</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>meta.c.color.string</span>, <span class='va'>to.hex</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>meta.c.color</span><span class='op'>)</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Negative"</span>, <span class='st'>"Positive"</span>, <span class='st'>"1"</span>, <span class='st'>"2"</span>, <span class='st'>"3"</span>, <span class='st'>"4"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>p3</span><span class='op'>$</span><span class='va'>meta.r</span><span class='op'>$</span><span class='va'>id</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in print(p3$meta.r$id): object 'p3' not found</span></div><div class='input'><span class='va'>meta.r.color.string</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"pink"</span>, <span class='st'>"azure"</span>, <span class='st'>"green"</span><span class='op'>)</span>
<span class='va'>meta.r.color</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='va'>meta.r.color.string</span>, <span class='va'>to.hex</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>meta.r.color</span><span class='op'>)</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"1"</span>, <span class='st'>"2"</span>, <span class='st'>"3"</span><span class='op'>)</span>

<span class='va'>p4</span><span class='op'>&lt;-</span><span class='fu'>heatmap.ggplot2</span><span class='op'>(</span>eSet<span class='op'>=</span><span class='va'>eSet1</span>, 
col.legend.brewer <span class='op'>=</span> <span class='va'>meta.c.color</span>,
row.legend.brewer <span class='op'>=</span> <span class='va'>meta.r.color</span>,
col.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, row.clust <span class='op'>=</span> <span class='cn'>TRUE</span>, 
col.clust.hc <span class='op'>=</span> <span class='va'>hc.col</span>, row.clust.hc <span class='op'>=</span> <span class='va'>hc.row</span>,
col.lab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span><span class='op'>)</span>, row.lab <span class='op'>=</span> <span class='st'>"cluster.row"</span>, 
heatmap.y.text <span class='op'>=</span> <span class='cn'>TRUE</span>, heatmap.x.text <span class='op'>=</span> <span class='cn'>TRUE</span>,
heatmap.colorlegend.name <span class='op'>=</span> <span class='st'>"RNASeq_expression"</span>,
title.text <span class='op'>=</span> <span class='st'>"TCGA BRCA log2 RNA-seq expression, z-score row normalized"</span>,
col.legend.name <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HER2_status"</span>, <span class='st'>"ER_status"</span>, <span class='st'>"PR_status"</span>, <span class='st'>"TN_status"</span>, <span class='st'>"cluster.col"</span><span class='op'>)</span>, 
row.legend.name <span class='op'>=</span> <span class='st'>"cluster.row"</span>, 
row.scaling <span class='op'>=</span> <span class='st'>"z-score.capped"</span>, 
z.norm <span class='op'>=</span> <span class='cn'>FALSE</span>, 
cuttree.col <span class='op'>=</span> <span class='fl'>4</span>, cuttree.row <span class='op'>=</span> <span class='fl'>3</span>,
verbose <span class='op'>=</span> <span class='cn'>FALSE</span>, show <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: `axis.ticks.margin` is deprecated. Please set `margin` property of `axis.text` instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `panel.margin` is deprecated. Please use `panel.spacing` property instead</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `legend.margin` must be specified using `margin()`. For the old behavior use legend.spacing</span></div><div class='output co'>#&gt; <span class='error'>Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'exprs': object 'eSet1' not found</span></div><div class='input'><span class='fu'>grid.newpage</span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.newpage(): could not find function "grid.newpage"</span></div><div class='input'><span class='fu'>grid.arrange</span><span class='op'>(</span><span class='va'>p4</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in grid.arrange(p4): could not find function "grid.arrange"</span></div><div class='input'>



</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/tetomonti'>Stefano Monti</a>, <a href='https://github.com/anfederico'>Anthony Federico</a>, <a href='https://github.com/ericreed'>Eric Reed</a>, <a href='https://github.com/lia978'>Amy Li</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


